<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="description" content="Test script with all steps for QC pipe">

<title>QC Pipeline</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script><script src="qc_overview_files/libs/clipboard/clipboard.min.js"></script>
<script src="qc_overview_files/libs/quarto-html/quarto.js"></script>
<script src="qc_overview_files/libs/quarto-html/popper.min.js"></script>
<script src="qc_overview_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="qc_overview_files/libs/quarto-html/anchor.min.js"></script>
<link href="qc_overview_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="qc_overview_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="qc_overview_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="qc_overview_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="qc_overview_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script src="qc_overview_files/libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="qc_overview_files/libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="qc_overview_files/libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active" data-toc-expanded="1">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#setup" id="toc-setup" class="nav-link active" data-scroll-target="#setup">Setup</a></li>
  <li><a href="#load-data-and-cell-calling" id="toc-load-data-and-cell-calling" class="nav-link" data-scroll-target="#load-data-and-cell-calling">Load data and cell calling</a>
  <ul class="collapse">
  <li><a href="#run-emptydrops" id="toc-run-emptydrops" class="nav-link" data-scroll-target="#run-emptydrops">Run EmptyDrops</a></li>
  <li><a href="#read-in-cellranger-filtering" id="toc-read-in-cellranger-filtering" class="nav-link" data-scroll-target="#read-in-cellranger-filtering">Read in cellranger filtering</a></li>
  <li><a href="#barcode-ranks" id="toc-barcode-ranks" class="nav-link" data-scroll-target="#barcode-ranks">Barcode ranks</a></li>
  <li><a href="#load-count-matrices" id="toc-load-count-matrices" class="nav-link" data-scroll-target="#load-count-matrices">Load count matrices</a></li>
  <li><a href="#celltyping" id="toc-celltyping" class="nav-link" data-scroll-target="#celltyping">Celltyping</a></li>
  <li><a href="#merge-object" id="toc-merge-object" class="nav-link" data-scroll-target="#merge-object">Merge object</a></li>
  </ul></li>
  <li><a href="#gene-annotations" id="toc-gene-annotations" class="nav-link" data-scroll-target="#gene-annotations">Gene annotations</a></li>
  <li><a href="#calculate-qc" id="toc-calculate-qc" class="nav-link" data-scroll-target="#calculate-qc">Calculate QC</a></li>
  <li><a href="#qc-violins" id="toc-qc-violins" class="nav-link" data-scroll-target="#qc-violins">QC violins</a></li>
  <li><a href="#filter-cells" id="toc-filter-cells" class="nav-link" data-scroll-target="#filter-cells">Filter cells</a></li>
  <li><a href="#top-genes" id="toc-top-genes" class="nav-link" data-scroll-target="#top-genes">Top genes</a></li>
  <li><a href="#filter-genes" id="toc-filter-genes" class="nav-link" data-scroll-target="#filter-genes">Filter genes</a></li>
  <li><a href="#celltypes" id="toc-celltypes" class="nav-link" data-scroll-target="#celltypes">Celltypes</a></li>
  <li><a href="#cell-sex" id="toc-cell-sex" class="nav-link" data-scroll-target="#cell-sex">Cell sex</a></li>
  <li><a href="#cellcycle" id="toc-cellcycle" class="nav-link" data-scroll-target="#cellcycle">Cellcycle</a></li>
  <li><a href="#scrublet---doublet-prediction" id="toc-scrublet---doublet-prediction" class="nav-link" data-scroll-target="#scrublet---doublet-prediction">Scrublet - Doublet prediction</a></li>
  <li><a href="#umap-and-clustering" id="toc-umap-and-clustering" class="nav-link" data-scroll-target="#umap-and-clustering">Umap and clustering</a></li>
  <li><a href="#save-file" id="toc-save-file" class="nav-link" data-scroll-target="#save-file">Save file</a></li>
  </ul>
</nav>
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">QC Pipeline</h1>
</div>

<div>
  <div class="description">
    Test script with all steps for QC pipe
  </div>
</div>


<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="setup" class="level2">
<h2 class="anchored" data-anchor-id="setup">Setup</h2>
<section id="read-in-the-yaml-file" class="level4">
<h4 class="anchored" data-anchor-id="read-in-the-yaml-file">Read in the yaml file:</h4>
<div id="yaml-load" class="cell" data-execution_count="1">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> yaml</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> <span class="bu">open</span>(<span class="st">'./settings.yml'</span>, <span class="st">'r'</span>) <span class="im">as</span> f:</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>    settings <span class="op">=</span> yaml.load(f, Loader<span class="op">=</span>yaml.SafeLoader)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Visualize all metadata:</p>
<div id="meta-sankey" class="cell" data-fig.heigth="8" data-fig.width="12" data-execution_count="2">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> plotly.graph_objects <span class="im">as</span> go</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>allmeta <span class="op">=</span> [settings[<span class="st">'samples'</span>][sample][<span class="st">'meta'</span>] <span class="cf">for</span> sample <span class="kw">in</span> settings[<span class="st">'samples'</span>].keys()]</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>meta <span class="op">=</span> pd.DataFrame.from_dict(allmeta) </span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>meta.index <span class="op">=</span> settings[<span class="st">'samples'</span>].keys()</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>meta[<span class="st">'sample'</span>] <span class="op">=</span> meta.index</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>meta[<span class="st">'idx'</span>] <span class="op">=</span> <span class="bu">range</span>(<span class="dv">0</span>,meta.shape[<span class="dv">0</span>])</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>dims <span class="op">=</span> []</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> var <span class="kw">in</span> meta.columns[:<span class="op">-</span><span class="dv">2</span>]:</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>    dims.append(go.parcats.Dimension(values<span class="op">=</span>meta[var], label<span class="op">=</span>var, categoryorder<span class="op">=</span><span class="st">'category ascending'</span>))</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>fig <span class="op">=</span> go.Figure(data <span class="op">=</span> [go.Parcats(dimensions<span class="op">=</span>[x <span class="cf">for</span> x <span class="kw">in</span> dims],</span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a>        line<span class="op">=</span>{<span class="st">'color'</span>: meta.idx})])</span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a><span class="co"># could not get the plotly to display, instead save as static image.</span></span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="kw">not</span> os.path.exists(<span class="st">"images"</span>):</span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a>    os.mkdir(<span class="st">"images"</span>)</span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a>fig.write_image(<span class="st">"images/metadata.png"</span>)            </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="images/metadata.png" class="lightbox" data-gallery="quarto-lightbox-gallery-1" title="image"><img src="images/metadata.png" class="img-fluid figure-img" alt="image"></a></p>
<figcaption>image</figcaption>
</figure>
</div>
</section>
<section id="paths-to-files" class="level4">
<h4 class="anchored" data-anchor-id="paths-to-files">Paths to files</h4>
<div id="paths" class="cell" data-execution_count="3">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>resdir <span class="op">=</span> settings[<span class="st">'outdir'</span>]</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="kw">not</span> os.path.exists(resdir):</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>    os.makedirs(resdir)</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="co">#dropsdir = "/Users/asabjor/projects/sc-devop/scQC/data/mouse/output/emptydrops/"</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="co">#singlerdir = "/Users/asabjor/projects/sc-devop/scQC/data/mouse/output/singler/"</span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>samples <span class="op">=</span> <span class="bu">list</span>(settings[<span class="st">'samples'</span>].keys())</span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a><span class="co">#["GSM6757771_rep1","GSM6757772_rep2","GSM6757773_rep3","GSM6757774_nuc2","GSM6757775_nuc3"]</span></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> sample <span class="kw">in</span> samples:</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>    sdir <span class="op">=</span> os.path.join(resdir,sample)</span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="kw">not</span> os.path.exists(sdir):</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>        os.makedirs(sdir)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="packages" class="level4">
<h4 class="anchored" data-anchor-id="packages">Packages</h4>
<div id="libraries" class="cell" data-execution_count="4">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> gc</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> warnings</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.patches <span class="im">as</span> mpatches</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> subprocess</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> anndata</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> IPython.display <span class="im">import</span> display, Markdown</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>warnings.simplefilter(action<span class="op">=</span><span class="st">'ignore'</span>, category<span class="op">=</span><span class="pp">Warning</span>)</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a><span class="co"># verbosity: errors (0), warnings (1), info (2), hints (3)</span></span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a>sc.settings.verbosity <span class="op">=</span> <span class="dv">3</span></span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a>sc.settings.set_figure_params(dpi<span class="op">=</span><span class="dv">80</span>)</span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a><span class="op">%</span>matplotlib inline</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="functions" class="level4">
<h4 class="anchored" data-anchor-id="functions">Functions</h4>
<div id="functions" class="cell" data-execution_count="5">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> rotate_xticks(ax, size<span class="op">=</span><span class="dv">8</span>, rotation<span class="op">=</span><span class="dv">45</span>):</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>    ax.set_xticklabels(ax.get_xticklabels(), rotation <span class="op">=</span> rotation, size<span class="op">=</span>size, ha<span class="op">=</span><span class="st">'right'</span>)    </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
</section>
<section id="load-data-and-cell-calling" class="level2">
<h2 class="anchored" data-anchor-id="load-data-and-cell-calling">Load data and cell calling</h2>
<p>Run emptydrops for cell calling and ambient RNA estimation. Load the count matrices and also run SingleR for celltype assignment.</p>
<section id="run-emptydrops" class="level3">
<h3 class="anchored" data-anchor-id="run-emptydrops">Run EmptyDrops</h3>
<p>First run <code>emptyDrops</code> for each of the samples via an Rscript. Then load the csv files with cell predictions for plotting barcode rank plots and ambient signal.</p>
<div id="read-dropdata" class="cell" data-execution_count="6">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>Rscript<span class="op">=</span> <span class="st">"/Users/asabjor/miniconda3/envs/rtools_scqc/bin/Rscript"</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>os.environ[<span class="st">"R_HOME"</span>] <span class="op">=</span> <span class="st">"/Users/asabjor/miniconda3/envs/rtools_scqc/lib/R"</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>script <span class="op">=</span> <span class="st">"/Users/asabjor/projects/sc-devop/scQC/source/all_qc/run_emptydrops.R"</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>dropletdata <span class="op">=</span> {}</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>ambientgenes <span class="op">=</span> {}</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> sample <span class="kw">in</span> samples:</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>    infile <span class="op">=</span> settings[<span class="st">'samples'</span>][sample][<span class="st">'raw_file'</span>]</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="kw">not</span> os.path.exists(infile):</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="st">"No such file: "</span>, infile)</span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>        os.sys.exit()</span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>    odir <span class="op">=</span> os.path.join(resdir,sample,<span class="st">"emptydrops"</span>)</span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="kw">not</span> os.path.exists(odir):</span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>       os.makedirs(odir)</span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a>    gene_file <span class="op">=</span> os.path.join(odir,<span class="st">'gene_stats.csv'</span>)</span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a>    cell_file <span class="op">=</span> os.path.join(odir,<span class="st">'cell_stats.csv'</span>)</span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (os.path.exists(gene_file) <span class="kw">and</span> os.path.exists(cell_file)) :</span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="st">"EmptyDrops done for "</span><span class="op">+</span>sample)</span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:   </span>
<span id="cb6-21"><a href="#cb6-21" aria-hidden="true" tabindex="-1"></a>        cmd <span class="op">=</span> [Rscript, script, <span class="st">"-i"</span>, infile, <span class="st">"-o"</span>, odir]</span>
<span id="cb6-22"><a href="#cb6-22" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(cmd)</span>
<span id="cb6-23"><a href="#cb6-23" aria-hidden="true" tabindex="-1"></a>        logfile <span class="op">=</span> os.path.join(odir,<span class="st">"emptydrops.log"</span>)</span>
<span id="cb6-24"><a href="#cb6-24" aria-hidden="true" tabindex="-1"></a>        <span class="cf">with</span> <span class="bu">open</span>(logfile,<span class="st">"w"</span>) <span class="im">as</span> f:    </span>
<span id="cb6-25"><a href="#cb6-25" aria-hidden="true" tabindex="-1"></a>            subprocess.call(cmd, stdout<span class="op">=</span>f) </span>
<span id="cb6-26"><a href="#cb6-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-27"><a href="#cb6-27" aria-hidden="true" tabindex="-1"></a>    stats <span class="op">=</span> pd.read_csv(cell_file, index_col <span class="op">=</span> <span class="dv">0</span>)</span>
<span id="cb6-28"><a href="#cb6-28" aria-hidden="true" tabindex="-1"></a>    stats <span class="op">=</span> stats.loc[stats[<span class="st">'Total'</span>]<span class="op">&gt;</span><span class="dv">0</span>,:]</span>
<span id="cb6-29"><a href="#cb6-29" aria-hidden="true" tabindex="-1"></a>    dropletdata[sample] <span class="op">=</span> stats</span>
<span id="cb6-30"><a href="#cb6-30" aria-hidden="true" tabindex="-1"></a>    ambientgenes[sample] <span class="op">=</span> pd.read_csv(gene_file, index_col <span class="op">=</span> <span class="dv">0</span>)</span>
<span id="cb6-31"><a href="#cb6-31" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(sample<span class="op">+</span> <span class="st">" with "</span><span class="op">+</span> <span class="bu">str</span>(ambientgenes[sample].shape[<span class="dv">0</span>]) <span class="op">+</span> <span class="st">" genes and "</span> <span class="op">+</span> <span class="bu">str</span>(stats.shape[<span class="dv">0</span>]) <span class="op">+</span> <span class="st">" cells</span><span class="ch">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>EmptyDrops done for GSM6757771_rep1
GSM6757771_rep1 with 20607 genes and 688611 cells

EmptyDrops done for GSM6757772_rep2
GSM6757772_rep2 with 20701 genes and 880265 cells

EmptyDrops done for GSM6757773_rep3
GSM6757773_rep3 with 19321 genes and 324663 cells

EmptyDrops done for GSM6757774_nuc2
GSM6757774_nuc2 with 22906 genes and 577560 cells

EmptyDrops done for GSM6757775_nuc3
GSM6757775_nuc3 with 21841 genes and 604486 cells

EmptyDrops done for GSE261852_CI82_spl1
GSE261852_CI82_spl1 with 23996 genes and 729492 cells

EmptyDrops done for GSE261852_CI82_spl2
GSE261852_CI82_spl2 with 23469 genes and 594136 cells

EmptyDrops done for GSM6594872_SAM24349928
GSM6594872_SAM24349928 with 22684 genes and 118937 cells

EmptyDrops done for GSM6594883_SAM24374037
GSM6594883_SAM24374037 with 26367 genes and 184388 cells
</code></pre>
</div>
</div>
</section>
<section id="read-in-cellranger-filtering" class="level3">
<h3 class="anchored" data-anchor-id="read-in-cellranger-filtering">Read in cellranger filtering</h3>
<p>If it is available, read in the filtered files from cellranger and store the list of barcodes.</p>
<div id="load-filtered" class="cell" data-execution_count="7">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>cellranger_barcodes <span class="op">=</span> {}</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> sample <span class="kw">in</span> samples:</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="st">'filt_file'</span> <span class="kw">in</span> settings[<span class="st">'samples'</span>][sample].keys() <span class="kw">and</span> settings[<span class="st">'samples'</span>][sample][<span class="st">'filt_file'</span>] <span class="op">!=</span> <span class="va">None</span> : </span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>        infile <span class="op">=</span> settings[<span class="st">'samples'</span>][sample][<span class="st">'filt_file'</span>]</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> os.path.exists(infile):</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>            tmp <span class="op">=</span> sc.read_10x_h5(infile)</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>            cellranger_barcodes[sample] <span class="op">=</span> tmp.obs_names</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>        <span class="cf">else</span>: </span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>            <span class="bu">print</span>()</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>            cellranger_barcodes[sample]<span class="op">=</span>[]</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>        cellranger_barcodes[sample]<span class="op">=</span>[]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>reading /Users/asabjor/projects/sc-devop/scQC/data/mouse/GSE218853/GSM6757771_rep1_filtered_feature_bc_matrix.h5
 (0:00:01)
reading /Users/asabjor/projects/sc-devop/scQC/data/mouse/GSE218853/GSM6757772_rep2_filtered_feature_bc_matrix.h5
 (0:00:01)
reading /Users/asabjor/projects/sc-devop/scQC/data/mouse/GSE218853/GSM6757773_rep3_filtered_feature_bc_matrix.h5
 (0:00:00)
reading /Users/asabjor/projects/sc-devop/scQC/data/mouse/GSE218853/GSM6757774_nuc2_filtered_feature_bc_matrix.h5
 (0:00:01)
reading /Users/asabjor/projects/sc-devop/scQC/data/mouse/GSE218853/GSM6757775_nuc3_filtered_feature_bc_matrix.h5
 (0:00:00)
reading /Users/asabjor/projects/sc-devop/scQC/data/mouse/GSE213825/GSM6594872_SAM24349928_filtered_feature_bc_matrix.h5
 (0:00:00)
reading /Users/asabjor/projects/sc-devop/scQC/data/mouse/GSE213825/GSM6594883_SAM24374037_filtered_feature_bc_matrix.h5
 (0:00:00)</code></pre>
</div>
</div>
</section>
<section id="barcode-ranks" class="level3">
<h3 class="anchored" data-anchor-id="barcode-ranks">Barcode ranks</h3>
<p>For each sample, plot the barcode rank plots with predicted cells according to EmptyDrops and CellRanger (if available). Second plot shows the predicted ambient signal genes from emptyDrops. For each sample separately.</p>
<div id="barcode-rank" class="cell" data-execution_count="8">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>fdr_cut <span class="op">=</span> <span class="fl">0.05</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> plot_barcode_rank(sample):</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>    stats <span class="op">=</span> dropletdata[sample]</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>    stats[<span class="st">"rank"</span>] <span class="op">=</span> stats[<span class="st">'Total'</span>].rank(method<span class="op">=</span><span class="st">'first'</span>, ascending <span class="op">=</span> <span class="va">False</span>)</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>    stats[<span class="st">"isCell"</span>] <span class="op">=</span> stats[<span class="st">"FDR"</span>] <span class="op">&lt;=</span> fdr_cut</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>    stats[<span class="st">"isCellCR"</span>] <span class="op">=</span> dropletdata[sample].index.isin(cellranger_barcodes[sample])</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">"EmptyDrops cell prediction: "</span>, <span class="bu">str</span>(stats[<span class="st">"isCell"</span>].<span class="bu">sum</span>()))</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">" lowest nUMI: "</span>, <span class="bu">str</span>(stats.Total[stats.isCell].<span class="bu">min</span>()), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">"Cellranger cell prediction: "</span> <span class="op">+</span> <span class="bu">str</span>(stats[<span class="st">"isCellCR"</span>].<span class="bu">sum</span>()))</span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">" lowest nUMI: "</span>, <span class="bu">str</span>(stats.Total[stats.isCellCR].<span class="bu">min</span>()), <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a>    stats[<span class="st">"color"</span>] <span class="op">=</span> <span class="st">"Grey"</span></span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a>    stats[<span class="st">"color"</span>][stats.isCell] <span class="op">=</span> <span class="st">"Blue"</span></span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true" tabindex="-1"></a>    stats[<span class="st">"color"</span>][stats.isCellCR] <span class="op">=</span> <span class="st">"Red"</span></span>
<span id="cb10-18"><a href="#cb10-18" aria-hidden="true" tabindex="-1"></a>    stats[<span class="st">"color"</span>][stats.isCell <span class="op">&amp;</span> stats.isCellCR] <span class="op">=</span> <span class="st">"Purple"</span></span>
<span id="cb10-19"><a href="#cb10-19" aria-hidden="true" tabindex="-1"></a>    clegend <span class="op">=</span> [mpatches.Patch(color<span class="op">=</span><span class="st">'Purple'</span>, label<span class="op">=</span><span class="st">'Both'</span>),</span>
<span id="cb10-20"><a href="#cb10-20" aria-hidden="true" tabindex="-1"></a>        mpatches.Patch(color<span class="op">=</span><span class="st">'Blue'</span>, label<span class="op">=</span><span class="st">'EmptyDrops'</span>),</span>
<span id="cb10-21"><a href="#cb10-21" aria-hidden="true" tabindex="-1"></a>        mpatches.Patch(color<span class="op">=</span><span class="st">'Red'</span>, label<span class="op">=</span><span class="st">'CellRanger'</span>),</span>
<span id="cb10-22"><a href="#cb10-22" aria-hidden="true" tabindex="-1"></a>        mpatches.Patch(color<span class="op">=</span><span class="st">'Grey'</span>, label<span class="op">=</span><span class="st">'None'</span>)</span>
<span id="cb10-23"><a href="#cb10-23" aria-hidden="true" tabindex="-1"></a>        ]</span>
<span id="cb10-24"><a href="#cb10-24" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">"Intersection cell prediction:"</span>, <span class="bu">str</span>((stats[<span class="st">'color'</span>] <span class="op">==</span><span class="st">"Purple"</span>).<span class="bu">sum</span>()))</span>
<span id="cb10-25"><a href="#cb10-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-26"><a href="#cb10-26" aria-hidden="true" tabindex="-1"></a>    fig, axs <span class="op">=</span> plt.subplots(<span class="dv">1</span>, <span class="dv">2</span>, figsize<span class="op">=</span>(<span class="dv">10</span>,<span class="dv">6</span>),constrained_layout<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb10-27"><a href="#cb10-27" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">0</span>].scatter(stats[<span class="st">"rank"</span>], stats[<span class="st">'Total'</span>], c<span class="op">=</span>stats[<span class="st">'color'</span>])</span>
<span id="cb10-28"><a href="#cb10-28" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">0</span>].set_xscale(<span class="st">"log"</span>)</span>
<span id="cb10-29"><a href="#cb10-29" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">0</span>].set_yscale(<span class="st">"log"</span>)</span>
<span id="cb10-30"><a href="#cb10-30" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">0</span>].set_xlabel(<span class="st">"rank"</span>)</span>
<span id="cb10-31"><a href="#cb10-31" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">0</span>].set_ylabel(<span class="st">"nUMI"</span>)</span>
<span id="cb10-32"><a href="#cb10-32" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">0</span>].title.set_text(sample)</span>
<span id="cb10-33"><a href="#cb10-33" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">0</span>].legend(handles<span class="op">=</span>clegend,</span>
<span id="cb10-34"><a href="#cb10-34" aria-hidden="true" tabindex="-1"></a>          loc<span class="op">=</span><span class="st">'lower left'</span>)</span>
<span id="cb10-35"><a href="#cb10-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-36"><a href="#cb10-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-37"><a href="#cb10-37" aria-hidden="true" tabindex="-1"></a>    gstats <span class="op">=</span> ambientgenes[sample]</span>
<span id="cb10-38"><a href="#cb10-38" aria-hidden="true" tabindex="-1"></a>    topA <span class="op">=</span> gstats[<span class="st">'ambient'</span>].sort_values(ascending<span class="op">=</span><span class="va">False</span>).head(<span class="dv">20</span>).index</span>
<span id="cb10-39"><a href="#cb10-39" aria-hidden="true" tabindex="-1"></a>    gstats[<span class="st">'ambient'</span>][topA].plot.bar(ax<span class="op">=</span>axs[<span class="dv">1</span>])</span>
<span id="cb10-40"><a href="#cb10-40" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">1</span>].title.set_text(<span class="st">"EmptyDrops ambient genes"</span>)</span>
<span id="cb10-41"><a href="#cb10-41" aria-hidden="true" tabindex="-1"></a>    rotate_xticks(axs[<span class="dv">1</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>FIX! Order of barplot for ambient is different to the violins.</p>
<div class="tabset-margin-container"></div><div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-1-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-1" role="tab" aria-controls="tabset-1-1" aria-selected="true" href="">GSM6757771_rep1</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-2" role="tab" aria-controls="tabset-1-2" aria-selected="false" href="">GSM6757772_rep2</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-3" role="tab" aria-controls="tabset-1-3" aria-selected="false" href="">GSM6757773_rep3</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-4-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-4" role="tab" aria-controls="tabset-1-4" aria-selected="false" href="">GSM6757774_nuc2</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-5-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-5" role="tab" aria-controls="tabset-1-5" aria-selected="false" href="">GSM6757775_nuc3</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-6-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-6" role="tab" aria-controls="tabset-1-6" aria-selected="false" href="">GSE261852_CI82_spl1</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-7-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-7" role="tab" aria-controls="tabset-1-7" aria-selected="false" href="">GSE261852_CI82_spl2</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-8-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-8" role="tab" aria-controls="tabset-1-8" aria-selected="false" href="">GSM6594872_SAM24349928</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-9-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-9" role="tab" aria-controls="tabset-1-9" aria-selected="false" href="">GSM6594883_SAM24374037</a></li></ul>
<div class="tab-content">
<div id="tabset-1-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-1-1-tab">
<p>EmptyDrops cell prediction: 73360 lowest nUMI: 101</p>
<p>Cellranger cell prediction: 21819 lowest nUMI: 528</p>
<p>Intersection cell prediction: 21819 <img src="qc_overview_files/figure-html/cell-10-output-3.png" width="809" height="489"></p>
</div>
<div id="tabset-1-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-2-tab">
<p>EmptyDrops cell prediction: 87078 lowest nUMI: 101</p>
<p>Cellranger cell prediction: 28187 lowest nUMI: 1090</p>
<p>Intersection cell prediction: 28187 <img src="qc_overview_files/figure-html/cell-10-output-7.png" width="809" height="489"></p>
</div>
<div id="tabset-1-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-3-tab">
<p>EmptyDrops cell prediction: 74869 lowest nUMI: 101</p>
<p>Cellranger cell prediction: 6546 lowest nUMI: 500</p>
<p>Intersection cell prediction: 6546 <img src="qc_overview_files/figure-html/cell-10-output-11.png" width="809" height="489"></p>
</div>
<div id="tabset-1-4" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-4-tab">
<p>EmptyDrops cell prediction: 58172 lowest nUMI: 101</p>
<p>Cellranger cell prediction: 59249 lowest nUMI: 500</p>
<p>Intersection cell prediction: 55909 <img src="qc_overview_files/figure-html/cell-10-output-15.png" width="809" height="489"></p>
</div>
<div id="tabset-1-5" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-5-tab">
<p>EmptyDrops cell prediction: 79352 lowest nUMI: 101</p>
<p>Cellranger cell prediction: 20129 lowest nUMI: 569</p>
<p>Intersection cell prediction: 20129 <img src="qc_overview_files/figure-html/cell-10-output-19.png" width="809" height="489"></p>
</div>
<div id="tabset-1-6" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-6-tab">
<p>EmptyDrops cell prediction: 32903 lowest nUMI: 101</p>
<p>Cellranger cell prediction: 0 lowest nUMI: nan</p>
<p>Intersection cell prediction: 0 <img src="qc_overview_files/figure-html/cell-10-output-23.png" width="809" height="489"></p>
</div>
<div id="tabset-1-7" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-7-tab">
<p>EmptyDrops cell prediction: 12063 lowest nUMI: 101</p>
<p>Cellranger cell prediction: 0 lowest nUMI: nan</p>
<p>Intersection cell prediction: 0 <img src="qc_overview_files/figure-html/cell-10-output-27.png" width="809" height="489"></p>
</div>
<div id="tabset-1-8" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-8-tab">
<p>EmptyDrops cell prediction: 1178 lowest nUMI: 101</p>
<p>Cellranger cell prediction: 1220 lowest nUMI: 500</p>
<p>Intersection cell prediction: 971 <img src="qc_overview_files/figure-html/cell-10-output-31.png" width="809" height="489"></p>
</div>
<div id="tabset-1-9" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-9-tab">
<p>EmptyDrops cell prediction: 7084 lowest nUMI: 101</p>
<p>Cellranger cell prediction: 7495 lowest nUMI: 500</p>
<p>Intersection cell prediction: 6018 <img src="qc_overview_files/figure-html/cell-10-output-35.png" width="809" height="489"></p>
</div>
</div>
</div>
</section>
<section id="load-count-matrices" class="level3">
<h3 class="anchored" data-anchor-id="load-count-matrices">Load count matrices</h3>
<p>Use cutoff for the emptyDrops predictions to filter the cells in the raw matrices and create one merged dataset with all samples.</p>
<div id="read-data" class="cell" data-execution_count="10">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>metadata_all <span class="op">=</span> {}</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>alldata <span class="op">=</span> {}</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> sample <span class="kw">in</span> samples:</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>    infile <span class="op">=</span> settings[<span class="st">'samples'</span>][sample][<span class="st">'raw_file'</span>]</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>    adata_raw <span class="op">=</span> sc.read_10x_h5(infile)</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>    adata_raw.var_names_make_unique()</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(sample<span class="op">+</span> <span class="st">" with "</span><span class="op">+</span> <span class="bu">str</span>(adata_raw.n_vars) <span class="op">+</span> <span class="st">" genes and "</span> <span class="op">+</span> <span class="bu">str</span>(adata_raw.n_obs) <span class="op">+</span> <span class="st">" cells</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>    adata_raw.obs[<span class="st">"sample"</span>] <span class="op">=</span> sample</span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (k,v) <span class="kw">in</span> settings[<span class="st">'samples'</span>][sample][<span class="st">'meta'</span>].items(): </span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>        adata_raw.obs[k] <span class="op">=</span> v</span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a>        metadata_all[k] <span class="op">=</span> <span class="dv">1</span></span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a><span class="co">#    adata_raw.obs["type"] = sample.split("_")[1]</span></span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true" tabindex="-1"></a>    <span class="co"># filter for predicted cells</span></span>
<span id="cb11-16"><a href="#cb11-16" aria-hidden="true" tabindex="-1"></a>    select <span class="op">=</span> dropletdata[sample][<span class="st">'isCell'</span>]</span>
<span id="cb11-17"><a href="#cb11-17" aria-hidden="true" tabindex="-1"></a>    cell_idx <span class="op">=</span> select.index[select]</span>
<span id="cb11-18"><a href="#cb11-18" aria-hidden="true" tabindex="-1"></a>    adata_raw  <span class="op">=</span> adata_raw[cell_idx,:]</span>
<span id="cb11-19"><a href="#cb11-19" aria-hidden="true" tabindex="-1"></a>    <span class="co"># save gene info from ambient in .var</span></span>
<span id="cb11-20"><a href="#cb11-20" aria-hidden="true" tabindex="-1"></a>    gdata <span class="op">=</span> ambientgenes[sample]</span>
<span id="cb11-21"><a href="#cb11-21" aria-hidden="true" tabindex="-1"></a>    adata_raw.var[<span class="st">'ambient'</span>] <span class="op">=</span> np.nan</span>
<span id="cb11-22"><a href="#cb11-22" aria-hidden="true" tabindex="-1"></a>    adata_raw.var[<span class="st">'ambient'</span>][adata_raw.var.index.isin(gdata.index)] <span class="op">=</span> gdata[<span class="st">'ambient'</span>]</span>
<span id="cb11-23"><a href="#cb11-23" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">"After filtering "</span> <span class="op">+</span> <span class="bu">str</span>(adata_raw.n_obs) <span class="op">+</span> <span class="st">" cells</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb11-24"><a href="#cb11-24" aria-hidden="true" tabindex="-1"></a>    alldata[sample]<span class="op">=</span>adata_raw</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>reading /Users/asabjor/projects/sc-devop/scQC/data/mouse/GSE218853/GSM6757771_rep1_raw_feature_bc_matrix.h5
 (0:00:06)
GSM6757771_rep1 with 28692 genes and 6794880 cells

After filtering 73360 cells

reading /Users/asabjor/projects/sc-devop/scQC/data/mouse/GSE218853/GSM6757772_rep2_raw_feature_bc_matrix.h5
 (0:00:06)
GSM6757772_rep2 with 28692 genes and 6794880 cells

After filtering 87078 cells

reading /Users/asabjor/projects/sc-devop/scQC/data/mouse/GSE218853/GSM6757773_rep3_raw_feature_bc_matrix.h5
 (0:00:05)
GSM6757773_rep3 with 28692 genes and 6794880 cells

After filtering 74869 cells

reading /Users/asabjor/projects/sc-devop/scQC/data/mouse/GSE218853/GSM6757774_nuc2_raw_feature_bc_matrix.h5
 (0:00:06)
GSM6757774_nuc2 with 28692 genes and 6794880 cells

After filtering 58172 cells

reading /Users/asabjor/projects/sc-devop/scQC/data/mouse/GSE218853/GSM6757775_nuc3_raw_feature_bc_matrix.h5
 (0:00:06)
GSM6757775_nuc3 with 28692 genes and 6794880 cells

After filtering 79352 cells

reading /Users/asabjor/projects/sc-devop/scQC/data/mouse/GSE261852/GSE261852_CI82_spl1_raw_feature_bc_matrix.h5
 (0:00:02)
GSE261852_CI82_spl1 with 32285 genes and 2106361 cells

After filtering 32903 cells

reading /Users/asabjor/projects/sc-devop/scQC/data/mouse/GSE261852/GSE261852_CI82_spl2_raw_feature_bc_matrix.h5
 (0:00:02)
GSE261852_CI82_spl2 with 32285 genes and 1871599 cells

After filtering 12063 cells

reading /Users/asabjor/projects/sc-devop/scQC/data/mouse/GSE213825/GSM6594872_SAM24349928_raw_feature_bc_matrix.h5
 (0:00:00)
GSM6594872_SAM24349928 with 52636 genes and 737280 cells

After filtering 1178 cells

reading /Users/asabjor/projects/sc-devop/scQC/data/mouse/GSE213825/GSM6594883_SAM24374037_raw_feature_bc_matrix.h5
 (0:00:06)
GSM6594883_SAM24374037 with 52636 genes and 6794880 cells

After filtering 7084 cells
</code></pre>
</div>
</div>
</section>
<section id="celltyping" class="level3">
<h3 class="anchored" data-anchor-id="celltyping">Celltyping</h3>
<p>Run for one sample at a time using a braod reference. Only found such ref for singleR, so run with an Rscript.</p>
<p>Add the predictions to <code>adata.var</code></p>
<div id="singler" class="cell" data-execution_count="11">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>Rscript<span class="op">=</span> <span class="st">"/Users/asabjor/miniconda3/envs/rtools_scqc/bin/Rscript"</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>os.environ[<span class="st">"R_HOME"</span>] <span class="op">=</span> <span class="st">"/Users/asabjor/miniconda3/envs/rtools_scqc/lib/R"</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>script <span class="op">=</span> <span class="st">"/Users/asabjor/projects/sc-devop/scQC/source/all_qc/run_singleR.R"</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> sample <span class="kw">in</span> samples:</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>    infile <span class="op">=</span> settings[<span class="st">'samples'</span>][sample][<span class="st">'raw_file'</span>]</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>    dropsfile <span class="op">=</span> os.path.join(resdir,sample,<span class="st">'emptydrops'</span>,<span class="st">'cell_stats.csv'</span>)</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>    outfile <span class="op">=</span> os.path.join(resdir,sample, sample <span class="op">+</span> <span class="st">'_singler.csv'</span>)</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (os.path.exists(outfile)) :</span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="st">"SingleR done for "</span><span class="op">+</span>sample)</span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:   </span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>        cmd <span class="op">=</span> [Rscript, script, <span class="st">"-i"</span>, infile, <span class="st">"-o"</span>, outfile, <span class="st">"-d"</span>, dropsfile, <span class="st">"-s"</span>, <span class="st">"mouse"</span>]</span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(cmd)</span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a>        logfile <span class="op">=</span> os.path.join(resdir,sample, sample <span class="op">+</span> <span class="st">"_singler.log"</span>)</span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a>        <span class="cf">with</span> <span class="bu">open</span>(logfile,<span class="st">"w"</span>) <span class="im">as</span> f:    </span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a>            subprocess.call(cmd, stdout<span class="op">=</span>f) </span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true" tabindex="-1"></a>    stats <span class="op">=</span> pd.read_csv(outfile, index_col <span class="op">=</span> <span class="dv">0</span>)</span>
<span id="cb13-19"><a href="#cb13-19" aria-hidden="true" tabindex="-1"></a>    alldata[sample].obs <span class="op">=</span> pd.concat([alldata[sample].obs,stats], axis<span class="op">=</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>SingleR done for GSM6757771_rep1
SingleR done for GSM6757772_rep2
SingleR done for GSM6757773_rep3
SingleR done for GSM6757774_nuc2
SingleR done for GSM6757775_nuc3
SingleR done for GSE261852_CI82_spl1
SingleR done for GSE261852_CI82_spl2
SingleR done for GSM6594872_SAM24349928
SingleR done for GSM6594883_SAM24374037</code></pre>
</div>
</div>
<div id="celltype-print1" class="cell" data-execution_count="12">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Stats on proportions per sample. List top 10 celltypes.</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>ct_stats <span class="op">=</span> {}</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> sample <span class="kw">in</span> samples:</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>    ct_stats[sample] <span class="op">=</span> alldata[sample].obs[<span class="st">'pruned.labels'</span>].value_counts(normalize<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>ct_stats <span class="op">=</span> pd.concat(ct_stats.values(), axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>ct_stats.columns <span class="op">=</span> samples</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a><span class="co"># order by mean</span></span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>ct_stats[<span class="st">'mean'</span>] <span class="op">=</span> ct_stats.mean(axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>ct_stats <span class="op">=</span> ct_stats.sort_values(by<span class="op">=</span><span class="st">"mean"</span>, ascending<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(ct_stats)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>                   GSM6757771_rep1  GSM6757772_rep2  GSM6757773_rep3  \
Hepatocytes               0.198260         0.245940         0.220474   
Adipocytes                0.291781         0.290569         0.235835   
T cells                   0.006476         0.001964         0.007841   
Epithelial cells          0.315597         0.333509         0.325755   
Neurons                   0.006885         0.001688         0.010138   
Endothelial cells         0.086786         0.081804         0.080784   
Fibroblasts               0.012556         0.009371         0.011541   
NK cells                  0.000286         0.000459         0.000094   
Oligodendrocytes          0.000982         0.000172         0.003206   
Monocytes                 0.042711         0.020144         0.055152   
Cardiomyocytes            0.014287         0.003859         0.022520   
Astrocytes                     NaN         0.000011         0.000053   
Macrophages               0.008071         0.005788         0.006679   
Erythrocytes              0.012542         0.003204         0.017137   
Granulocytes              0.000286         0.000057         0.000321   
B cells                   0.002208         0.001436         0.001883   
Dendritic cells           0.000082         0.000023         0.000094   
Microglia                 0.000204              NaN         0.000494   

                   GSM6757774_nuc2  GSM6757775_nuc3  GSE261852_CI82_spl1  \
Hepatocytes               0.232832         0.375154                  NaN   
Adipocytes                0.571112         0.377070                  NaN   
T cells                   0.000052         0.000517             0.831834   
Epithelial cells          0.117568         0.162756                  NaN   
Neurons                   0.003456         0.002231             0.000030   
Endothelial cells         0.034112         0.023668             0.000304   
Fibroblasts               0.005897         0.007121             0.000030   
NK cells                  0.000034         0.000013             0.163486   
Oligodendrocytes          0.005192         0.005268                  NaN   
Monocytes                 0.000378         0.004499                  NaN   
Cardiomyocytes            0.012947         0.030197                  NaN   
Astrocytes                0.014838         0.002634                  NaN   
Macrophages               0.000946         0.003277             0.000061   
Erythrocytes              0.000344         0.004802             0.000061   
Granulocytes                   NaN         0.000076                  NaN   
B cells                   0.000052         0.000429             0.001641   
Dendritic cells           0.000052         0.000063             0.002523   
Microglia                 0.000189         0.000227             0.000030   

                   GSE261852_CI82_spl2  GSM6594872_SAM24349928  \
Hepatocytes                        NaN                     NaN   
Adipocytes                         NaN                0.008496   
T cells                       0.971704                0.004248   
Epithelial cells                   NaN                0.048428   
Neurons                       0.000090                0.613424   
Endothelial cells                  NaN                0.086661   
Fibroblasts                   0.000180                0.105353   
NK cells                      0.024151                0.001699   
Oligodendrocytes                   NaN                0.062022   
Monocytes                     0.000090                0.010195   
Cardiomyocytes                     NaN                0.000850   
Astrocytes                         NaN                0.028887   
Macrophages                   0.000180                0.022940   
Erythrocytes                  0.000180                0.001699   
Granulocytes                       NaN                0.001699   
B cells                       0.002703                0.001699   
Dendritic cells               0.000631                0.000850   
Microglia                     0.000090                0.000850   

                   GSM6594883_SAM24374037      mean  
Hepatocytes                           NaN  0.254532  
Adipocytes                       0.002542  0.253915  
T cells                          0.000706  0.202816  
Epithelial cells                 0.027401  0.190145  
Neurons                          0.673870  0.145757  
Endothelial cells                0.138559  0.066585  
Fibroblasts                      0.051130  0.022575  
NK cells                         0.000989  0.021246  
Oligodendrocytes                 0.046610  0.017636  
Monocytes                        0.002542  0.016964  
Cardiomyocytes                        NaN  0.014110  
Astrocytes                       0.034605  0.013505  
Macrophages                      0.010028  0.006441  
Erythrocytes                     0.001836  0.004645  
Granulocytes                     0.007203  0.001607  
B cells                          0.001977  0.001559  
Dendritic cells                       NaN  0.000539  
Microglia                             NaN  0.000298  </code></pre>
</div>
</div>
<p>Barplot for top 10 categories</p>
<div id="cell-celltype-plot1" class="cell" data-execution_count="13">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>ct_stats.head(<span class="dv">10</span>).T.plot.bar(stacked<span class="op">=</span><span class="va">True</span>).legend(loc<span class="op">=</span><span class="st">'center left'</span>,bbox_to_anchor<span class="op">=</span>(<span class="fl">1.0</span>, <span class="fl">0.5</span>))<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="qc_overview_files/figure-html/celltype-plot1-output-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-2"><img src="qc_overview_files/figure-html/celltype-plot1-output-1.png" id="celltype-plot1" width="438" height="474" class="figure-img"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="merge-object" class="level3">
<h3 class="anchored" data-anchor-id="merge-object">Merge object</h3>
<p>Create one adata object from all samples</p>
<div id="cell-merge" class="cell" data-execution_count="14">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># anndata.concat works to merge the data, But selects the intersection of genes across the datasets.</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> anndata.concat(alldata)</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>adata.var <span class="op">=</span> alldata[samples[<span class="dv">0</span>]].var[alldata[samples[<span class="dv">0</span>]].var_names.isin(adata.var_names)]</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>adata.obs_names_make_unique() <span class="co"># in case of same barcode in 2 samples</span></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a><span class="kw">del</span> alldata</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a><span class="kw">del</span> ambientgenes</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a><span class="kw">del</span> dropletdata</span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a><span class="kw">del</span> cellranger_barcodes</span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>gc.collect(generation<span class="op">=</span><span class="dv">2</span>)</span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(adata.shape)</span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">'sample'</span>].value_counts()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>(426059, 26634)</code></pre>
</div>
<div id="merge" class="cell-output cell-output-display" data-execution_count="20">
<pre><code>GSM6757772_rep2           87078
GSM6757775_nuc3           79352
GSM6757773_rep3           74869
GSM6757771_rep1           73360
GSM6757774_nuc2           58172
GSE261852_CI82_spl1       32903
GSE261852_CI82_spl2       12063
GSM6594883_SAM24374037     7084
GSM6594872_SAM24349928     1178
Name: sample, dtype: int64</code></pre>
</div>
</div>
</section>
</section>
<section id="gene-annotations" class="level2">
<h2 class="anchored" data-anchor-id="gene-annotations">Gene annotations</h2>
<p>First, read in the annotation file for biotypes.</p>
<p>Make broader groupings:</p>
<ul>
<li>Mito genes - both PC and nc.</li>
<li>PC genes - includes also IG/TR genes</li>
<li>lncRNA</li>
<li>other - all other categories, also genes with not translation in biomart.</li>
</ul>
<div id="read-annot" class="cell" data-execution_count="15">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>indir <span class="op">=</span> <span class="st">"/Users/asabjor/projects/sc-devop/scQC/data/mouse"</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>annot_file <span class="op">=</span> os.path.join(resdir, <span class="st">"mouse_annotation.csv"</span>)</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> os.path.exists(annot_file):</span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>    annot <span class="op">=</span> pd.read_csv(annot_file, index_col<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span>:   </span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>    annot <span class="op">=</span> sc.queries.biomart_annotations(<span class="st">"mmusculus"</span>, [<span class="st">"ensembl_gene_id"</span>, <span class="st">"external_gene_name"</span>, <span class="st">"start_position"</span>, <span class="st">"end_position"</span>, <span class="st">"chromosome_name"</span>, <span class="st">"gene_biotype"</span>] )</span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>    annot.to_csv(annot_file)</span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a><span class="co"># make slim biotype</span></span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a>annot[<span class="st">"gene_class"</span>]<span class="op">=</span> <span class="st">"other"</span></span>
<span id="cb21-13"><a href="#cb21-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-14"><a href="#cb21-14" aria-hidden="true" tabindex="-1"></a>pc_genes <span class="op">=</span> [<span class="st">"protein_coding"</span>,<span class="st">"IG_V_gene"</span>,<span class="st">"IG_D_gene"</span>,<span class="st">"IG_J_gene"</span>,<span class="st">"IG_C_gene"</span>,<span class="st">"IG_LV_gene"</span>, <span class="st">"TR_V_gene"</span>,<span class="st">"TR_J_gene"</span>,<span class="st">"TR_C_gene"</span>,<span class="st">"TR_D_gene"</span> ]</span>
<span id="cb21-15"><a href="#cb21-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-16"><a href="#cb21-16" aria-hidden="true" tabindex="-1"></a>annot[<span class="st">"gene_class"</span>][annot.gene_biotype.isin(pc_genes)] <span class="op">=</span> <span class="st">"pc"</span></span>
<span id="cb21-17"><a href="#cb21-17" aria-hidden="true" tabindex="-1"></a>annot[<span class="st">"gene_class"</span>][annot.gene_biotype <span class="op">==</span> <span class="st">"lncRNA"</span>] <span class="op">=</span> <span class="st">"lncRNA"</span></span>
<span id="cb21-18"><a href="#cb21-18" aria-hidden="true" tabindex="-1"></a>annot[<span class="st">"gene_class"</span>][annot.external_gene_name.<span class="bu">str</span>.startswith(<span class="st">'mt-'</span>).fillna(<span class="va">False</span>)] <span class="op">=</span> <span class="st">"mt"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="merge-annot" class="cell" data-execution_count="16">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co">## add into .var of the object</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>t <span class="op">=</span> adata.var.merge(annot, left_on<span class="op">=</span><span class="st">"gene_ids"</span>, right_on<span class="op">=</span><span class="st">"ensembl_gene_id"</span>, how<span class="op">=</span><span class="st">'left'</span>)</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>t.index <span class="op">=</span> adata.var.index</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>t.gene_class[t.gene_class.isna()] <span class="op">=</span> <span class="st">"other"</span></span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>adata.var <span class="op">=</span> t</span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a><span class="co"># add in the biotypes as own columns in adata.var</span></span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a>gene_classes <span class="op">=</span> [<span class="st">'pc'</span>,<span class="st">'lncRNA'</span>,<span class="st">'mt'</span>,<span class="st">'other'</span>]</span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> cl <span class="kw">in</span> gene_classes:</span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a>    adata.var[cl]<span class="op">=</span>adata.var[<span class="st">'gene_class'</span>] <span class="op">==</span> cl</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="calculate-qc" class="level2">
<h2 class="anchored" data-anchor-id="calculate-qc">Calculate QC</h2>
<p>First calculate percantage mito, then remove mito genes and calculate the rest.</p>
<div id="qc-genes" class="cell" data-execution_count="17">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Add in also some more gene groups. Ribo, HB </span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="co"># ribosomal genes</span></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>adata.var[<span class="st">'ribo'</span>] <span class="op">=</span> adata.var_names.<span class="bu">str</span>.startswith((<span class="st">"Rps"</span>,<span class="st">"Rpl"</span>))</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a><span class="co"># hemoglobin genes.</span></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>adata.var[<span class="st">'hb'</span>] <span class="op">=</span> adata.var_names.<span class="bu">str</span>.contains((<span class="st">"^Hb[^(p|e|s)]"</span>))</span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>sc.pp.calculate_qc_metrics(adata, qc_vars<span class="op">=</span>[<span class="st">'mt'</span>], percent_top<span class="op">=</span><span class="va">None</span>, log1p<span class="op">=</span><span class="va">False</span>, inplace<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> adata[:,<span class="op">~</span>adata.var[<span class="st">'mt'</span>]]</span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a>sc.pp.calculate_qc_metrics(adata, qc_vars<span class="op">=</span>[<span class="st">'ribo'</span>,<span class="st">'hb'</span>,<span class="st">'pc'</span>,<span class="st">'lncRNA'</span>,<span class="st">'ribo'</span>,<span class="st">'other'</span>], percent_top<span class="op">=</span><span class="va">None</span>, log1p<span class="op">=</span><span class="va">False</span>, inplace<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="qc-violins" class="level2">
<h2 class="anchored" data-anchor-id="qc-violins">QC violins</h2>
<p>Violin plots with different metadata groupings, always include sample and top celltypes, but also add in any other metadata found in the metadata field of the settings.yml.</p>
<div id="qc-meta" class="cell" data-execution_count="18">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># fix top10 celltype and "other" annotation</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>topC <span class="op">=</span> ct_stats.head(<span class="dv">10</span>).index.tolist()</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">"topCelltype"</span>] <span class="op">=</span> adata.obs[<span class="st">'pruned.labels'</span>].copy()</span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">'topCelltype'</span>][<span class="op">~</span>adata.obs[<span class="st">'pruned.labels'</span>].isin(topC)] <span class="op">=</span> <span class="st">"Other"</span></span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>plot_vln_meta <span class="op">=</span>  [<span class="st">"sample"</span>,<span class="st">"topCelltype"</span>]  <span class="op">+</span>  <span class="bu">list</span>(metadata_all.keys())</span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a>plot_stats <span class="op">=</span> [<span class="st">'n_genes_by_counts'</span>, <span class="st">'total_counts'</span>, <span class="st">'pct_counts_mt'</span>, <span class="st">'pct_counts_ribo'</span>, <span class="st">'pct_counts_hb'</span>, <span class="st">'pct_counts_pc'</span>, <span class="st">'pct_counts_lncRNA'</span>, <span class="st">'pct_counts_other'</span>]</span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> plot_qc_vln(meta_slot):</span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true" tabindex="-1"></a>    fig, axs <span class="op">=</span> plt.subplots(<span class="dv">3</span>, <span class="dv">3</span>, figsize<span class="op">=</span>(<span class="dv">10</span>,<span class="dv">10</span>),constrained_layout<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb24-15"><a href="#cb24-15" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> stat, ax <span class="kw">in</span> <span class="bu">zip</span>(plot_stats, axs.flatten()):</span>
<span id="cb24-16"><a href="#cb24-16" aria-hidden="true" tabindex="-1"></a>        sc.pl.violin(adata, stat, groupby <span class="op">=</span> meta_slot, jitter<span class="op">=</span><span class="fl">0.4</span>,  size <span class="op">=</span> <span class="fl">0.1</span>, ax<span class="op">=</span>ax, show<span class="op">=</span><span class="va">False</span> )</span>
<span id="cb24-17"><a href="#cb24-17" aria-hidden="true" tabindex="-1"></a>        rotate_xticks(ax)</span>
<span id="cb24-18"><a href="#cb24-18" aria-hidden="true" tabindex="-1"></a>        ax.title.set_text(stat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="tabset-margin-container"></div><div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-2-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-1" role="tab" aria-controls="tabset-2-1" aria-selected="true" href="">sample</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-2-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-2" role="tab" aria-controls="tabset-2-2" aria-selected="false" href="">topCelltype</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-2-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-3" role="tab" aria-controls="tabset-2-3" aria-selected="false" href="">libtype</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-2-4-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-4" role="tab" aria-controls="tabset-2-4" aria-selected="false" href="">sample_id</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-2-5-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-5" role="tab" aria-controls="tabset-2-5" aria-selected="false" href="">tissue</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-2-6-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-6" role="tab" aria-controls="tabset-2-6" aria-selected="false" href="">study</a></li></ul>
<div class="tab-content">
<div id="tabset-2-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-2-1-tab">
<p><a href="qc_overview_files/figure-html/cell-20-output-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-3"><img src="qc_overview_files/figure-html/cell-20-output-2.png" width="809" height="809"></a></p>
</div>
<div id="tabset-2-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-2-2-tab">
<p><a href="qc_overview_files/figure-html/cell-20-output-5.png" class="lightbox" data-gallery="quarto-lightbox-gallery-4"><img src="qc_overview_files/figure-html/cell-20-output-5.png" width="809" height="809"></a></p>
</div>
<div id="tabset-2-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-2-3-tab">
<p><a href="qc_overview_files/figure-html/cell-20-output-8.png" class="lightbox" data-gallery="quarto-lightbox-gallery-5"><img src="qc_overview_files/figure-html/cell-20-output-8.png" width="809" height="809"></a></p>
</div>
<div id="tabset-2-4" class="tab-pane" role="tabpanel" aria-labelledby="tabset-2-4-tab">
<p><a href="qc_overview_files/figure-html/cell-20-output-11.png" class="lightbox" data-gallery="quarto-lightbox-gallery-6"><img src="qc_overview_files/figure-html/cell-20-output-11.png" width="809" height="809"></a></p>
</div>
<div id="tabset-2-5" class="tab-pane" role="tabpanel" aria-labelledby="tabset-2-5-tab">
<p><a href="qc_overview_files/figure-html/cell-20-output-14.png" class="lightbox" data-gallery="quarto-lightbox-gallery-7"><img src="qc_overview_files/figure-html/cell-20-output-14.png" width="809" height="809"></a></p>
</div>
<div id="tabset-2-6" class="tab-pane" role="tabpanel" aria-labelledby="tabset-2-6-tab">
<p><a href="qc_overview_files/figure-html/cell-20-output-17.png" class="lightbox" data-gallery="quarto-lightbox-gallery-8"><img src="qc_overview_files/figure-html/cell-20-output-17.png" width="809" height="809"></a></p>
</div>
</div>
</div>
<p>Scatter plots</p>
<div id="cell-qc-scatter" class="cell" data-execution_count="20">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>fig, axs <span class="op">=</span> plt.subplots(<span class="dv">1</span>, <span class="dv">2</span>, figsize<span class="op">=</span>(<span class="dv">10</span>,<span class="dv">5</span>),constrained_layout<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>sc.pl.scatter(adata, x<span class="op">=</span><span class="st">'total_counts'</span>, y<span class="op">=</span><span class="st">'pct_counts_mt'</span>, color<span class="op">=</span><span class="st">'sample'</span>, ax<span class="op">=</span>axs[<span class="dv">0</span>],show<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>axs[<span class="dv">0</span>].set_xscale(<span class="st">"log"</span>)</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>sc.pl.scatter(adata, x<span class="op">=</span><span class="st">'total_counts'</span>, y<span class="op">=</span><span class="st">'n_genes_by_counts'</span>, color<span class="op">=</span><span class="st">'sample'</span>, ax<span class="op">=</span>axs[<span class="dv">1</span>],show<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>axs[<span class="dv">1</span>].set_xscale(<span class="st">"log"</span>)</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>axs[<span class="dv">1</span>].set_yscale(<span class="st">"log"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="qc_overview_files/figure-html/qc-scatter-output-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-9"><img src="qc_overview_files/figure-html/qc-scatter-output-1.png" id="qc-scatter" width="809" height="409" class="figure-img"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="filter-cells" class="level2">
<h2 class="anchored" data-anchor-id="filter-cells">Filter cells</h2>
<p>Filter for now on:</p>
<ul>
<li>nFeatures</li>
<li>pct_counts_mt</li>
<li>pct_counts_pc</li>
</ul>
<p>Later, make it adatptive for multiple different qc stats selected.</p>
<p>For now, have options:</p>
<ul>
<li>Fixed/Adaptive - same cutoff for all samples or adaptive cutoff.</li>
<li>setting - either the value for fixed, or the number of MADs (default 2)</li>
<li>below/above - remove the cells with low/high values on the stat.</li>
</ul>
<p>Have as default setting for now:</p>
<pre><code>filtering:
    n_genes_by_counts: ['fixed',200,'below']
    pct_counts_mt': ['adaptive',2,'above']
    pct_counts_pc': ['adaptive',2,'below']</code></pre>
<p>Plot violins for each stat with:</p>
<ul>
<li>black solid line at median</li>
<li>black dotted line at median * 1,2,3 MAD in both directions</li>
<li>red line at decided cutoff</li>
</ul>
<div id="filter-cutoffs" class="cell" data-execution_count="21">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>filter_batch <span class="op">=</span> <span class="st">'sample'</span>    </span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="st">'filter_batch'</span> <span class="kw">in</span> settings[<span class="st">'filtering'</span>].keys():</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>    filter_batch <span class="op">=</span> settings[<span class="st">'filtering'</span>][<span class="st">'filter_batch'</span>]</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a><span class="co"># default settings for filtering if not provided in the </span></span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>filter_settings <span class="op">=</span> settings[<span class="st">'filtering'</span>][<span class="st">'cutoffs'</span>]</span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a><span class="co"># group the stats by selected bathch and qc-stats.</span></span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a>sel <span class="op">=</span> <span class="bu">list</span>(<span class="bu">set</span>([x[<span class="dv">0</span>] <span class="cf">for</span> x <span class="kw">in</span> filter_settings]))</span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a>sel.append(filter_batch)</span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a>sample_stats <span class="op">=</span> adata.obs.loc[:,sel].groupby(filter_batch)</span>
<span id="cb27-12"><a href="#cb27-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-13"><a href="#cb27-13" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate mean/median per sample</span></span>
<span id="cb27-14"><a href="#cb27-14" aria-hidden="true" tabindex="-1"></a>median <span class="op">=</span> sample_stats.median()</span>
<span id="cb27-15"><a href="#cb27-15" aria-hidden="true" tabindex="-1"></a>mad <span class="op">=</span> sample_stats.mad() </span>
<span id="cb27-16"><a href="#cb27-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-17"><a href="#cb27-17" aria-hidden="true" tabindex="-1"></a>cells_remove <span class="op">=</span> []</span>
<span id="cb27-18"><a href="#cb27-18" aria-hidden="true" tabindex="-1"></a>filtered <span class="op">=</span> {}</span>
<span id="cb27-19"><a href="#cb27-19" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(filter_settings)):</span>
<span id="cb27-20"><a href="#cb27-20" aria-hidden="true" tabindex="-1"></a>    [st, kind, cut, direction] <span class="op">=</span> filter_settings[i]</span>
<span id="cb27-21"><a href="#cb27-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-22"><a href="#cb27-22" aria-hidden="true" tabindex="-1"></a>    <span class="co"># define the cutoffs</span></span>
<span id="cb27-23"><a href="#cb27-23" aria-hidden="true" tabindex="-1"></a>    cutoff <span class="op">=</span> {}</span>
<span id="cb27-24"><a href="#cb27-24" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> kind <span class="op">==</span> <span class="st">'fixed'</span>:</span>
<span id="cb27-25"><a href="#cb27-25" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> sample <span class="kw">in</span> samples: </span>
<span id="cb27-26"><a href="#cb27-26" aria-hidden="true" tabindex="-1"></a>            cutoff[sample] <span class="op">=</span> cut</span>
<span id="cb27-27"><a href="#cb27-27" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="st">"Filtering "</span> <span class="op">+</span> st <span class="op">+</span> <span class="st">" for all cells "</span> <span class="op">+</span> direction <span class="op">+</span> <span class="st">" cutoff "</span> <span class="op">+</span> <span class="bu">str</span>(cut))</span>
<span id="cb27-28"><a href="#cb27-28" aria-hidden="true" tabindex="-1"></a>    <span class="cf">elif</span> kind <span class="op">==</span> <span class="st">'adaptive'</span>:</span>
<span id="cb27-29"><a href="#cb27-29" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> direction <span class="op">==</span> <span class="st">'below'</span>:</span>
<span id="cb27-30"><a href="#cb27-30" aria-hidden="true" tabindex="-1"></a>            <span class="cf">for</span> sample <span class="kw">in</span> samples: </span>
<span id="cb27-31"><a href="#cb27-31" aria-hidden="true" tabindex="-1"></a>                cutoff[sample] <span class="op">=</span> median.loc[sample,st] <span class="op">-</span> mad.loc[sample,st] <span class="op">*</span> cut</span>
<span id="cb27-32"><a href="#cb27-32" aria-hidden="true" tabindex="-1"></a>                <span class="bu">print</span>(<span class="st">"Filtering "</span> <span class="op">+</span> st <span class="op">+</span> <span class="st">" for "</span><span class="op">+</span> sample <span class="op">+</span> <span class="st">" below cutoff "</span> <span class="op">+</span> <span class="bu">str</span>(cutoff[sample]))</span>
<span id="cb27-33"><a href="#cb27-33" aria-hidden="true" tabindex="-1"></a>        <span class="cf">elif</span> direction <span class="op">==</span> <span class="st">'above'</span>:</span>
<span id="cb27-34"><a href="#cb27-34" aria-hidden="true" tabindex="-1"></a>            <span class="cf">for</span> sample <span class="kw">in</span> samples: </span>
<span id="cb27-35"><a href="#cb27-35" aria-hidden="true" tabindex="-1"></a>                cutoff[sample] <span class="op">=</span> median.loc[sample,st] <span class="op">+</span> mad.loc[sample,st] <span class="op">*</span> cut</span>
<span id="cb27-36"><a href="#cb27-36" aria-hidden="true" tabindex="-1"></a>                <span class="bu">print</span>(<span class="st">"Filtering "</span> <span class="op">+</span> st <span class="op">+</span> <span class="st">" for "</span><span class="op">+</span> sample <span class="op">+</span> <span class="st">" above cutoff "</span> <span class="op">+</span> <span class="bu">str</span>(cutoff[sample]))</span>
<span id="cb27-37"><a href="#cb27-37" aria-hidden="true" tabindex="-1"></a>        <span class="cf">else</span>:</span>
<span id="cb27-38"><a href="#cb27-38" aria-hidden="true" tabindex="-1"></a>            <span class="bu">print</span>(<span class="st">"unknown direction type "</span> <span class="op">+</span> direction <span class="op">+</span> <span class="st">"skipping</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb27-39"><a href="#cb27-39" aria-hidden="true" tabindex="-1"></a>            <span class="cf">continue</span></span>
<span id="cb27-40"><a href="#cb27-40" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>: </span>
<span id="cb27-41"><a href="#cb27-41" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="st">"unknown filtering type "</span> <span class="op">+</span> kind <span class="op">+</span> <span class="st">"skipping</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb27-42"><a href="#cb27-42" aria-hidden="true" tabindex="-1"></a>        <span class="cf">continue</span> </span>
<span id="cb27-43"><a href="#cb27-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-44"><a href="#cb27-44" aria-hidden="true" tabindex="-1"></a>    <span class="co"># plot</span></span>
<span id="cb27-45"><a href="#cb27-45" aria-hidden="true" tabindex="-1"></a>    fig, axs <span class="op">=</span> plt.subplots(<span class="dv">1</span>, <span class="dv">2</span>, figsize<span class="op">=</span>(<span class="dv">10</span>,<span class="dv">6</span>),constrained_layout<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb27-46"><a href="#cb27-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-47"><a href="#cb27-47" aria-hidden="true" tabindex="-1"></a>    sc.pl.violin(adata, st, groupby <span class="op">=</span> <span class="st">'sample'</span>, jitter<span class="op">=</span><span class="fl">0.4</span>,  rotation<span class="op">=</span> <span class="dv">45</span>, size <span class="op">=</span> <span class="fl">0.1</span>, show <span class="op">=</span> <span class="va">False</span>, ax <span class="op">=</span> axs[<span class="dv">0</span>]) </span>
<span id="cb27-48"><a href="#cb27-48" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (idx,sample) <span class="kw">in</span> <span class="bu">enumerate</span>(adata.obs[<span class="st">'sample'</span>].cat.categories):</span>
<span id="cb27-49"><a href="#cb27-49" aria-hidden="true" tabindex="-1"></a>        me <span class="op">=</span> median.loc[sample,st]</span>
<span id="cb27-50"><a href="#cb27-50" aria-hidden="true" tabindex="-1"></a>        ma <span class="op">=</span> mad.loc[sample,st]</span>
<span id="cb27-51"><a href="#cb27-51" aria-hidden="true" tabindex="-1"></a>        axs[<span class="dv">0</span>].hlines(me,idx<span class="op">-</span><span class="fl">0.5</span>,idx<span class="op">+</span><span class="fl">0.5</span>, linewidth<span class="op">=</span><span class="dv">2</span>, color <span class="op">=</span> <span class="st">'black'</span>)</span>
<span id="cb27-52"><a href="#cb27-52" aria-hidden="true" tabindex="-1"></a>        axs[<span class="dv">0</span>].hlines(me<span class="op">+</span>ma<span class="op">*</span><span class="dv">1</span>,idx<span class="op">-</span><span class="fl">0.5</span>,idx<span class="op">+</span><span class="fl">0.5</span>, linewidth<span class="op">=</span><span class="dv">2</span>, linestyles <span class="op">=</span> <span class="st">'dotted'</span>, color <span class="op">=</span> <span class="st">'black'</span>)</span>
<span id="cb27-53"><a href="#cb27-53" aria-hidden="true" tabindex="-1"></a>        axs[<span class="dv">0</span>].hlines(me<span class="op">-</span>ma<span class="op">*</span><span class="dv">1</span>,idx<span class="op">-</span><span class="fl">0.5</span>,idx<span class="op">+</span><span class="fl">0.5</span>, linewidth<span class="op">=</span><span class="dv">2</span>, linestyles <span class="op">=</span> <span class="st">'dotted'</span>, color <span class="op">=</span> <span class="st">'black'</span>)</span>
<span id="cb27-54"><a href="#cb27-54" aria-hidden="true" tabindex="-1"></a>        axs[<span class="dv">0</span>].hlines(me<span class="op">+</span>ma<span class="op">*</span><span class="dv">2</span>,idx<span class="op">-</span><span class="fl">0.5</span>,idx<span class="op">+</span><span class="fl">0.5</span>, linewidth<span class="op">=</span><span class="dv">2</span>, linestyles <span class="op">=</span> <span class="st">'dotted'</span>, color <span class="op">=</span> <span class="st">'black'</span>)</span>
<span id="cb27-55"><a href="#cb27-55" aria-hidden="true" tabindex="-1"></a>        axs[<span class="dv">0</span>].hlines(me<span class="op">-</span>ma<span class="op">*</span><span class="dv">2</span>,idx<span class="op">-</span><span class="fl">0.5</span>,idx<span class="op">+</span><span class="fl">0.5</span>, linewidth<span class="op">=</span><span class="dv">2</span>, linestyles <span class="op">=</span> <span class="st">'dotted'</span>, color <span class="op">=</span> <span class="st">'black'</span>)</span>
<span id="cb27-56"><a href="#cb27-56" aria-hidden="true" tabindex="-1"></a>        axs[<span class="dv">0</span>].hlines(me<span class="op">+</span>ma<span class="op">*</span><span class="dv">3</span>,idx<span class="op">-</span><span class="fl">0.5</span>,idx<span class="op">+</span><span class="fl">0.5</span>, linewidth<span class="op">=</span><span class="dv">2</span>, linestyles <span class="op">=</span> <span class="st">'dotted'</span>, color <span class="op">=</span> <span class="st">'black'</span>)</span>
<span id="cb27-57"><a href="#cb27-57" aria-hidden="true" tabindex="-1"></a>        axs[<span class="dv">0</span>].hlines(me<span class="op">-</span>ma<span class="op">*</span><span class="dv">3</span>,idx<span class="op">-</span><span class="fl">0.5</span>,idx<span class="op">+</span><span class="fl">0.5</span>, linewidth<span class="op">=</span><span class="dv">2</span>, linestyles <span class="op">=</span> <span class="st">'dotted'</span>, color <span class="op">=</span> <span class="st">'black'</span>)    </span>
<span id="cb27-58"><a href="#cb27-58" aria-hidden="true" tabindex="-1"></a>        axs[<span class="dv">0</span>].hlines(cutoff[sample],idx<span class="op">-</span><span class="fl">0.5</span>,idx<span class="op">+</span><span class="fl">0.5</span>, linewidth<span class="op">=</span><span class="dv">4</span>, color <span class="op">=</span> <span class="st">'red'</span>)</span>
<span id="cb27-59"><a href="#cb27-59" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">0</span>].title.set_text(st)</span>
<span id="cb27-60"><a href="#cb27-60" aria-hidden="true" tabindex="-1"></a>    rotate_xticks(axs[<span class="dv">0</span>])</span>
<span id="cb27-61"><a href="#cb27-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-62"><a href="#cb27-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-63"><a href="#cb27-63" aria-hidden="true" tabindex="-1"></a>    <span class="co"># save the filtered cells to a list      </span></span>
<span id="cb27-64"><a href="#cb27-64" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> direction <span class="op">==</span> <span class="st">'below'</span>:</span>
<span id="cb27-65"><a href="#cb27-65" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> sample <span class="kw">in</span> samples: </span>
<span id="cb27-66"><a href="#cb27-66" aria-hidden="true" tabindex="-1"></a>            filtered[sample] <span class="op">=</span>   (adata.obs[st] <span class="op">&lt;=</span> cutoff[sample])[adata.obs[<span class="st">'sample'</span>]<span class="op">==</span>sample]</span>
<span id="cb27-67"><a href="#cb27-67" aria-hidden="true" tabindex="-1"></a>    <span class="cf">elif</span> direction <span class="op">==</span> <span class="st">'above'</span>:</span>
<span id="cb27-68"><a href="#cb27-68" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> sample <span class="kw">in</span> samples: </span>
<span id="cb27-69"><a href="#cb27-69" aria-hidden="true" tabindex="-1"></a>            filtered[sample] <span class="op">=</span>   (adata.obs[st] <span class="op">&gt;=</span> cutoff[sample])[adata.obs[<span class="st">'sample'</span>]<span class="op">==</span>sample]</span>
<span id="cb27-70"><a href="#cb27-70" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb27-71"><a href="#cb27-71" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> sample <span class="kw">in</span> samples:</span>
<span id="cb27-72"><a href="#cb27-72" aria-hidden="true" tabindex="-1"></a>        cells_remove.extend(<span class="bu">list</span>(filtered[sample].index[filtered[sample]]) )     </span>
<span id="cb27-73"><a href="#cb27-73" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-74"><a href="#cb27-74" aria-hidden="true" tabindex="-1"></a>    fstats <span class="op">=</span> pd.Series([filtered[sample].<span class="bu">sum</span>()<span class="op">/</span>filtered[sample].shape[<span class="dv">0</span>]<span class="op">*</span><span class="dv">100</span> <span class="cf">for</span> sample <span class="kw">in</span> adata.obs[<span class="st">'sample'</span>].cat.categories], index<span class="op">=</span> adata.obs[<span class="st">'sample'</span>].cat.categories)</span>
<span id="cb27-75"><a href="#cb27-75" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-76"><a href="#cb27-76" aria-hidden="true" tabindex="-1"></a>    <span class="co">#fstats = pd.Series([ filtered['sample'].value_counts(normalize=True)[1] * 100 for sample in samples], index = samples)</span></span>
<span id="cb27-77"><a href="#cb27-77" aria-hidden="true" tabindex="-1"></a>    <span class="co"># OBS! Value counts does not work when all are False!!</span></span>
<span id="cb27-78"><a href="#cb27-78" aria-hidden="true" tabindex="-1"></a>    fstats.plot.bar(ax<span class="op">=</span>axs[<span class="dv">1</span>])</span>
<span id="cb27-79"><a href="#cb27-79" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">1</span>].title.set_text(<span class="st">"Percent filtered cells"</span>)   </span>
<span id="cb27-80"><a href="#cb27-80" aria-hidden="true" tabindex="-1"></a>    rotate_xticks(axs[<span class="dv">1</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Filtering n_genes_by_counts for all cells below cutoff 200
Filtering n_genes_by_counts for GSM6757771_rep1 above cutoff 1821.0341322933225
Filtering n_genes_by_counts for GSM6757772_rep2 above cutoff 1890.2704776371947
Filtering n_genes_by_counts for GSM6757773_rep3 above cutoff 922.7472199564627
Filtering n_genes_by_counts for GSM6757774_nuc2 above cutoff 1385.1673130234149
Filtering n_genes_by_counts for GSM6757775_nuc3 above cutoff 974.9002867074753
Filtering n_genes_by_counts for GSE261852_CI82_spl1 above cutoff 3978.409733126997
Filtering n_genes_by_counts for GSE261852_CI82_spl2 above cutoff 5338.786504524462
Filtering n_genes_by_counts for GSM6594872_SAM24349928 above cutoff 6387.447513987334
Filtering n_genes_by_counts for GSM6594883_SAM24374037 above cutoff 5569.229489571141
Filtering pct_counts_mt for GSM6757771_rep1 above cutoff 80.62651050403437
Filtering pct_counts_mt for GSM6757772_rep2 above cutoff 70.7941143695952
Filtering pct_counts_mt for GSM6757773_rep3 above cutoff 80.6863892830159
Filtering pct_counts_mt for GSM6757774_nuc2 above cutoff 40.23891029658003
Filtering pct_counts_mt for GSM6757775_nuc3 above cutoff 53.4054513591073
Filtering pct_counts_mt for GSE261852_CI82_spl1 above cutoff 5.203391537226236
Filtering pct_counts_mt for GSE261852_CI82_spl2 above cutoff 8.117819659778089
Filtering pct_counts_mt for GSM6594872_SAM24349928 above cutoff 7.973089946316893
Filtering pct_counts_mt for GSM6594883_SAM24374037 above cutoff 5.653369847161591
Filtering pct_counts_pc for GSM6757771_rep1 below cutoff 96.1730951334164
Filtering pct_counts_pc for GSM6757772_rep2 below cutoff 97.28505772989486
Filtering pct_counts_pc for GSM6757773_rep3 below cutoff 96.62621415370113
Filtering pct_counts_pc for GSM6757774_nuc2 below cutoff 94.75338305153512
Filtering pct_counts_pc for GSM6757775_nuc3 below cutoff 97.14835136798928
Filtering pct_counts_pc for GSE261852_CI82_spl1 below cutoff 64.79189417193597
Filtering pct_counts_pc for GSE261852_CI82_spl2 below cutoff 92.71003399576742
Filtering pct_counts_pc for GSM6594872_SAM24349928 below cutoff 87.50377133665748
Filtering pct_counts_pc for GSM6594883_SAM24374037 below cutoff 94.1454663160896</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="qc_overview_files/figure-html/filter-cutoffs-output-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-10"><img src="qc_overview_files/figure-html/filter-cutoffs-output-2.png" id="filter-cutoffs-1" width="809" height="489" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="qc_overview_files/figure-html/filter-cutoffs-output-3.png" class="lightbox" data-gallery="quarto-lightbox-gallery-11"><img src="qc_overview_files/figure-html/filter-cutoffs-output-3.png" id="filter-cutoffs-2" width="809" height="489" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="qc_overview_files/figure-html/filter-cutoffs-output-4.png" class="lightbox" data-gallery="quarto-lightbox-gallery-12"><img src="qc_overview_files/figure-html/filter-cutoffs-output-4.png" id="filter-cutoffs-3" width="809" height="489" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="qc_overview_files/figure-html/filter-cutoffs-output-5.png" class="lightbox" data-gallery="quarto-lightbox-gallery-13"><img src="qc_overview_files/figure-html/filter-cutoffs-output-5.png" id="filter-cutoffs-4" width="809" height="489" class="figure-img"></a></p>
</figure>
</div>
</div>
</div>
<div id="e7612131" class="cell" data-execution_count="22">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># do the final filtering</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>before <span class="op">=</span> adata.obs[<span class="st">'sample'</span>].value_counts()</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> adata[<span class="op">~</span>adata.obs_names.isin(cells_remove),:]</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>after <span class="op">=</span> adata.obs[<span class="st">'sample'</span>].value_counts()</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>stats <span class="op">=</span> pd.concat([before,after], axis<span class="op">=</span><span class="dv">1</span> )</span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>stats.columns <span class="op">=</span> [<span class="st">'Before'</span>,<span class="st">'After'</span>]</span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a>stats[<span class="st">'Percentage'</span>] <span class="op">=</span> (stats.Before<span class="op">-</span>stats.After)<span class="op">/</span>stats.Before <span class="op">*</span> <span class="dv">100</span></span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a>gc.collect()</span>
<span id="cb29-10"><a href="#cb29-10" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(stats)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>                        Before  After  Percentage
GSM6757772_rep2          87078  66260   23.907301
GSM6757775_nuc3          79352  46087   41.920809
GSM6757773_rep3          74869   3700   95.058035
GSM6757771_rep1          73360  16252   77.846238
GSM6757774_nuc2          58172  48674   16.327443
GSE261852_CI82_spl1      32903  11635   64.638483
GSE261852_CI82_spl2      12063  10303   14.590069
GSM6594883_SAM24374037    7084   5551   21.640316
GSM6594872_SAM24349928    1178    793   32.682513</code></pre>
</div>
</div>
<p>FIX! Plot each filtering as tabset!</p>
<p>FIX! Order of the samples in violin and barplot should be the same.</p>
</section>
<section id="top-genes" class="level2">
<h2 class="anchored" data-anchor-id="top-genes">Top genes</h2>
<div id="cell-top-genes" class="cell" data-execution_count="23">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>fig, axs <span class="op">=</span> plt.subplots(<span class="dv">3</span>, <span class="dv">3</span>, figsize<span class="op">=</span>(<span class="dv">10</span>,<span class="dv">10</span>),constrained_layout<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> sample, ax <span class="kw">in</span> <span class="bu">zip</span>(samples, axs.flatten()):</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>    sc.pl.highest_expr_genes(adata[adata.obs[<span class="st">'sample'</span>]<span class="op">==</span>sample,:], n_top<span class="op">=</span><span class="dv">20</span>, ax<span class="op">=</span>ax, show<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>    ax.title.set_text(sample) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>normalizing counts per cell
    finished (0:00:00)
normalizing counts per cell
    finished (0:00:00)
normalizing counts per cell
    finished (0:00:00)
normalizing counts per cell
    finished (0:00:00)
normalizing counts per cell
    finished (0:00:00)
normalizing counts per cell
    finished (0:00:00)
normalizing counts per cell
    finished (0:00:00)
normalizing counts per cell
    finished (0:00:00)
normalizing counts per cell
    finished (0:00:00)</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="qc_overview_files/figure-html/top-genes-output-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-14"><img src="qc_overview_files/figure-html/top-genes-output-2.png" id="top-genes" width="818" height="809" class="figure-img"></a></p>
</figure>
</div>
</div>
</div>
<p>FIX! Dynamic number of subpltos each filtering as tabset!</p>
</section>
<section id="filter-genes" class="level2">
<h2 class="anchored" data-anchor-id="filter-genes">Filter genes</h2>
<p>OBS! Now removing mito genes already earlier..</p>
<div id="filter-genes" class="cell" data-execution_count="24">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(adata.n_vars)</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>malat1 <span class="op">=</span> adata.var_names.<span class="bu">str</span>.startswith(<span class="st">'Malat1'</span>)</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a><span class="co"># we need to redefine the mito_genes since they were first </span></span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a><span class="co"># calculated on the full object before removing low expressed genes.</span></span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>mito_genes <span class="op">=</span> adata.var_names.<span class="bu">str</span>.startswith(<span class="st">'mt-'</span>)</span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a>remove <span class="op">=</span> np.add(mito_genes, malat1)</span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a>keep <span class="op">=</span> np.invert(remove)</span>
<span id="cb33-9"><a href="#cb33-9" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> adata[:,keep]</span>
<span id="cb33-10"><a href="#cb33-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-11"><a href="#cb33-11" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(adata.n_vars)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>26621
26620</code></pre>
</div>
</div>
</section>
<section id="celltypes" class="level2">
<h2 class="anchored" data-anchor-id="celltypes">Celltypes</h2>
<p>Plot again the celltypes after filtering</p>
<div id="celltype-print" class="cell" data-execution_count="25">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Stats on proportions per sample. List top 10 celltypes.</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>ct_stats2 <span class="op">=</span> {}</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> sample <span class="kw">in</span> samples:</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>    ct_stats2[sample] <span class="op">=</span> adata[adata.obs[<span class="st">'sample'</span>] <span class="op">==</span> sample,:].obs[<span class="st">'pruned.labels'</span>].value_counts(normalize<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>ct_stats2 <span class="op">=</span> pd.concat(ct_stats2.values(), axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>ct_stats2.columns <span class="op">=</span> samples</span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a><span class="co"># order by mean</span></span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a>ct_stats2[<span class="st">'mean'</span>] <span class="op">=</span> ct_stats2.mean(axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a>ct_stats2 <span class="op">=</span> ct_stats2.sort_values(by<span class="op">=</span><span class="st">"mean"</span>, ascending<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-12"><a href="#cb35-12" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(ct_stats2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>                   GSM6757771_rep1  GSM6757772_rep2  GSM6757773_rep3  \
Adipocytes                0.352677         0.267942         0.260541   
Hepatocytes               0.192862         0.266085         0.210000   
T cells                   0.005169         0.001675         0.002432   
Epithelial cells          0.301846         0.348748         0.389730   
Neurons                   0.002831         0.001479         0.006216   
Endothelial cells         0.077538         0.075132         0.071351   
Fibroblasts               0.014646         0.003803         0.012162   
Astrocytes                     NaN              NaN              NaN   
Oligodendrocytes          0.000431         0.000075              NaN   
Cardiomyocytes            0.007015         0.003290         0.013784   
Monocytes                 0.018523         0.021809         0.014865   
Macrophages               0.017108         0.004739         0.009730   
NK cells                  0.000738         0.000332              NaN   
Erythrocytes              0.004369         0.003502         0.006216   
Granulocytes              0.000615         0.000015         0.001081   
B cells                   0.003569         0.001373         0.001892   
Microglia                      NaN              NaN              NaN   
Dendritic cells           0.000062              NaN              NaN   

                   GSM6757774_nuc2  GSM6757775_nuc3  GSE261852_CI82_spl1  \
Adipocytes                0.571464         0.398346                  NaN   
Hepatocytes               0.244349         0.407482                  NaN   
T cells                   0.000041         0.000130             0.977568   
Epithelial cells          0.111079         0.139900                  NaN   
Neurons                   0.002610         0.001085                  NaN   
Endothelial cells         0.034848         0.016817                  NaN   
Fibroblasts               0.005979         0.003906                  NaN   
Astrocytes                0.008568         0.000521                  NaN   
Oligodendrocytes          0.004603         0.003060                  NaN   
Cardiomyocytes            0.014260         0.022329                  NaN   
Monocytes                 0.000431         0.002018                  NaN   
Macrophages               0.001068         0.001497                  NaN   
NK cells                  0.000041              NaN             0.021401   
Erythrocytes              0.000411         0.002669                  NaN   
Granulocytes                   NaN         0.000022                  NaN   
B cells                   0.000041         0.000130             0.000774   
Microglia                 0.000144         0.000065             0.000086   
Dendritic cells           0.000062         0.000022             0.000172   

                   GSE261852_CI82_spl2  GSM6594872_SAM24349928  \
Adipocytes                         NaN                0.003783   
Hepatocytes                        NaN                     NaN   
T cells                       0.997461                0.005044   
Epithelial cells                   NaN                0.051702   
Neurons                            NaN                0.659521   
Endothelial cells                  NaN                0.094578   
Fibroblasts                   0.000098                0.074401   
Astrocytes                         NaN                0.027743   
Oligodendrocytes                   NaN                0.047919   
Cardiomyocytes                     NaN                     NaN   
Monocytes                          NaN                0.006305   
Macrophages                   0.000098                0.025221   
NK cells                      0.000684                0.001261   
Erythrocytes                  0.000195                     NaN   
Granulocytes                       NaN                0.002522   
B cells                       0.001465                     NaN   
Microglia                          NaN                     NaN   
Dendritic cells                    NaN                     NaN   

                   GSM6594883_SAM24374037      mean  
Adipocytes                       0.002883  0.265376  
Hepatocytes                           NaN  0.264156  
T cells                          0.000541  0.221118  
Epithelial cells                 0.029369  0.196054  
Neurons                          0.646126  0.188553  
Endothelial cells                0.160360  0.075804  
Fibroblasts                      0.051712  0.020838  
Astrocytes                       0.034414  0.017812  
Oligodendrocytes                 0.044865  0.016825  
Cardiomyocytes                        NaN  0.012136  
Monocytes                        0.002703  0.009522  
Macrophages                      0.012252  0.008964  
NK cells                         0.000901  0.003623  
Erythrocytes                     0.002342  0.002815  
Granulocytes                     0.009009  0.002211  
B cells                          0.002523  0.001471  
Microglia                             NaN  0.000098  
Dendritic cells                       NaN  0.000079  </code></pre>
</div>
</div>
<p>Barplot for top 10 categories</p>
<div id="cell-celltype-plot" class="cell" data-fig.width="12" data-execution_count="26">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>ct_stats2.head(<span class="dv">10</span>).T.plot.bar(stacked<span class="op">=</span><span class="va">True</span>).legend(loc<span class="op">=</span><span class="st">'center left'</span>,bbox_to_anchor<span class="op">=</span>(<span class="fl">1.0</span>, <span class="fl">0.5</span>))<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="qc_overview_files/figure-html/celltype-plot-output-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-15"><img src="qc_overview_files/figure-html/celltype-plot-output-1.png" id="celltype-plot" width="438" height="474" class="figure-img"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="cell-sex" class="level2">
<h2 class="anchored" data-anchor-id="cell-sex">Cell sex</h2>
<p>Need to add in specific function per species.</p>
</section>
<section id="cellcycle" class="level2">
<h2 class="anchored" data-anchor-id="cellcycle">Cellcycle</h2>
<p>First read the file with cell cycle genes, from Regev lab and split into S and G2M phase genes. We first download the file.</p>
<div id="fetch-ccgenes" class="cell" data-execution_count="27">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>path_data <span class="op">=</span> <span class="st">"https://nextcloud.dc.scilifelab.se/public.php/webdav"</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>curl_upass <span class="op">=</span> <span class="st">"zbC5fr2LbEZ9rSE:scRNAseq2025"</span></span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>path_cc_file <span class="op">=</span> os.path.join(indir, <span class="st">'regev_lab_cell_cycle_genes.txt'</span>)</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="kw">not</span> os.path.exists(path_cc_file):</span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>    file_url <span class="op">=</span> os.path.join(path_data, <span class="st">"misc/regev_lab_cell_cycle_genes.txt"</span>)</span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a>    subprocess.call([<span class="st">"curl"</span>, <span class="st">"-u"</span>, curl_upass, <span class="st">"-o"</span>, path_cc_file, file_url ])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="prep-ccgenes" class="cell" data-execution_count="28">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>cell_cycle_genes <span class="op">=</span> [x.strip() <span class="cf">for</span> x <span class="kw">in</span> <span class="bu">open</span>(path_cc_file)]</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a><span class="co"># convert to mouse names if needed</span></span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> convert2mousegene(hgenes):</span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>    hgenes <span class="op">=</span> [s.lower() <span class="cf">for</span> s <span class="kw">in</span> hgenes]</span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>    trans <span class="op">=</span> adata.var_names[adata.var_names.<span class="bu">str</span>.lower().isin(hgenes)]</span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">"Translated "</span>, <span class="bu">str</span>(<span class="bu">len</span>(trans)), <span class="st">" out of "</span>, <span class="bu">len</span>(hgenes), <span class="st">"human genes</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span>(trans)</span>
<span id="cb39-9"><a href="#cb39-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-10"><a href="#cb39-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-11"><a href="#cb39-11" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> settings[<span class="st">'species'</span>] <span class="op">==</span> <span class="st">'mouse'</span>:</span>
<span id="cb39-12"><a href="#cb39-12" aria-hidden="true" tabindex="-1"></a>    s_genes <span class="op">=</span> convert2mousegene(cell_cycle_genes[:<span class="dv">43</span>])</span>
<span id="cb39-13"><a href="#cb39-13" aria-hidden="true" tabindex="-1"></a>    g2m_genes <span class="op">=</span> convert2mousegene(cell_cycle_genes[<span class="dv">43</span>:])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Translated  42  out of  43 human genes

Translated  52  out of  54 human genes
</code></pre>
</div>
</div>
<p>WARNING: genes are not in var_names and ignored: Index([‘Mlf1Ip’, ‘Rad51Ap1’, ‘Casp8Ap2’, ‘Chaf1B’, ‘E2F8’], dtype=‘object’) finished: added ‘S_score’, score of gene set (adata.obs). 386 total control genes are used. (0:00:04) computing score ‘G2M_score’ WARNING: genes are not in var_names and ignored: Index([‘Ube2C’, ‘Top2A’, ‘Cks1B’, ‘Fam64A’, ‘Ckap2L’, ‘Anp32E’, ‘Tubb4B’, ‘Kif20B’, ‘Hn1’, ‘Cdc25C’, ‘Kif2C’, ‘G2E3’, ‘Gas2L3’],</p>
<section id="normalize" class="level4">
<h4 class="anchored" data-anchor-id="normalize">Normalize</h4>
<p>Before running cell cycle we have to normalize the data. In the scanpy object, the data slot will be overwritten with the normalized data. So first, save the raw data into the slot <code>raw</code>. Then run normalization, log transformation and scale the data.</p>
<div id="normalize" class="cell" data-execution_count="29">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co"># save raw counts in raw slot.</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>adata.raw <span class="op">=</span> adata</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a><span class="co"># normalize to depth 10 000</span></span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>sc.pp.normalize_total(adata, target_sum<span class="op">=</span><span class="fl">1e4</span>)</span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a><span class="co"># logaritmize</span></span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a>sc.pp.log1p(adata)</span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-10"><a href="#cb41-10" aria-hidden="true" tabindex="-1"></a><span class="co"># want the unscaled data for later. so scale in a temporary object.</span></span>
<span id="cb41-11"><a href="#cb41-11" aria-hidden="true" tabindex="-1"></a>tmp <span class="op">=</span> adata.copy()</span>
<span id="cb41-12"><a href="#cb41-12" aria-hidden="true" tabindex="-1"></a>sc.pp.scale(tmp)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>normalizing counts per cell
    finished (0:00:00)
... as `zero_center=True`, sparse input is densified and may lead to large memory consumption</code></pre>
</div>
</div>
<p>For comparison, run CC prediction with all samples together or each sample separately and compare output.</p>
<div id="cc" class="cell" data-execution_count="30">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>sc.tl.score_genes_cell_cycle(tmp, s_genes<span class="op">=</span>s_genes, g2m_genes<span class="op">=</span>g2m_genes)</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">'S_score'</span>] <span class="op">=</span> tmp.obs[<span class="st">'S_score'</span>]</span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">'G2M_score'</span>] <span class="op">=</span> tmp.obs[<span class="st">'G2M_score'</span>]</span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">'phase'</span>] <span class="op">=</span> tmp.obs[<span class="st">'phase'</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>calculating cell cycle phase
computing score 'S_score'
    finished: added
    'S_score', score of gene set (adata.obs).
    335 total control genes are used. (0:00:01)
computing score 'G2M_score'
    finished: added
    'G2M_score', score of gene set (adata.obs).
    544 total control genes are used. (0:00:01)
--&gt;     'phase', cell cycle phase (adata.obs)</code></pre>
</div>
</div>
<div id="cc-sample" class="cell" data-execution_count="31">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>cc_scores <span class="op">=</span> {}</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> sample <span class="kw">in</span> samples:</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>    cc <span class="op">=</span> sc.tl.score_genes_cell_cycle(tmp[tmp.obs[<span class="st">'sample'</span>]<span class="op">==</span>sample], s_genes<span class="op">=</span>s_genes, g2m_genes<span class="op">=</span>g2m_genes, copy<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>    cc_scores[sample]<span class="op">=</span>cc.obs.iloc[:,<span class="op">-</span><span class="dv">3</span>:]</span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a>cc_scores <span class="op">=</span> pd.concat(cc_scores)</span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a>cc_scores <span class="op">=</span> cc_scores.reset_index(level<span class="op">=</span>[<span class="dv">0</span>])</span>
<span id="cb45-8"><a href="#cb45-8" aria-hidden="true" tabindex="-1"></a><span class="co">#cc_scores.index.get_level_values(1)</span></span>
<span id="cb45-9"><a href="#cb45-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-10"><a href="#cb45-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-11"><a href="#cb45-11" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">'S_score2'</span>] <span class="op">=</span> cc_scores[<span class="st">'S_score'</span>]</span>
<span id="cb45-12"><a href="#cb45-12" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">'G2M_score2'</span>] <span class="op">=</span> cc_scores[<span class="st">'G2M_score'</span>]</span>
<span id="cb45-13"><a href="#cb45-13" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">'phase2'</span>] <span class="op">=</span> cc_scores[<span class="st">'phase'</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>calculating cell cycle phase
computing score 'S_score'
    finished: added
    'S_score', score of gene set (adata.obs).
    462 total control genes are used. (0:00:00)
computing score 'G2M_score'
    finished: added
    'G2M_score', score of gene set (adata.obs).
    585 total control genes are used. (0:00:00)
--&gt;     'phase', cell cycle phase (adata.obs)
calculating cell cycle phase
computing score 'S_score'
    finished: added
    'S_score', score of gene set (adata.obs).
    545 total control genes are used. (0:00:00)
computing score 'G2M_score'
    finished: added
    'G2M_score', score of gene set (adata.obs).
    586 total control genes are used. (0:00:00)
--&gt;     'phase', cell cycle phase (adata.obs)
calculating cell cycle phase
computing score 'S_score'
    finished: added
    'S_score', score of gene set (adata.obs).
    460 total control genes are used. (0:00:00)
computing score 'G2M_score'
    finished: added
    'G2M_score', score of gene set (adata.obs).
    543 total control genes are used. (0:00:00)
--&gt;     'phase', cell cycle phase (adata.obs)
calculating cell cycle phase
computing score 'S_score'
    finished: added
    'S_score', score of gene set (adata.obs).
    458 total control genes are used. (0:00:00)
computing score 'G2M_score'
    finished: added
    'G2M_score', score of gene set (adata.obs).
    587 total control genes are used. (0:00:00)
--&gt;     'phase', cell cycle phase (adata.obs)
calculating cell cycle phase
computing score 'S_score'
    finished: added
    'S_score', score of gene set (adata.obs).
    461 total control genes are used. (0:00:00)
computing score 'G2M_score'
    finished: added
    'G2M_score', score of gene set (adata.obs).
    629 total control genes are used. (0:00:00)
--&gt;     'phase', cell cycle phase (adata.obs)
calculating cell cycle phase
computing score 'S_score'
    finished: added
    'S_score', score of gene set (adata.obs).
    294 total control genes are used. (0:00:00)
computing score 'G2M_score'
    finished: added
    'G2M_score', score of gene set (adata.obs).
    375 total control genes are used. (0:00:00)
--&gt;     'phase', cell cycle phase (adata.obs)
calculating cell cycle phase
computing score 'S_score'
    finished: added
    'S_score', score of gene set (adata.obs).
    294 total control genes are used. (0:00:00)
computing score 'G2M_score'
    finished: added
    'G2M_score', score of gene set (adata.obs).
    334 total control genes are used. (0:00:00)
--&gt;     'phase', cell cycle phase (adata.obs)
calculating cell cycle phase
computing score 'S_score'
    finished: added
    'S_score', score of gene set (adata.obs).
    503 total control genes are used. (0:00:00)
computing score 'G2M_score'
    finished: added
    'G2M_score', score of gene set (adata.obs).
    585 total control genes are used. (0:00:00)
--&gt;     'phase', cell cycle phase (adata.obs)
calculating cell cycle phase
computing score 'S_score'
    finished: added
    'S_score', score of gene set (adata.obs).
    504 total control genes are used. (0:00:00)
computing score 'G2M_score'
    finished: added
    'G2M_score', score of gene set (adata.obs).
    545 total control genes are used. (0:00:00)
--&gt;     'phase', cell cycle phase (adata.obs)</code></pre>
</div>
</div>
<div id="de55f65f" class="cell" data-execution_count="32">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="kw">del</span> tmp</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>gc.collect()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="38">
<pre><code>18923</code></pre>
</div>
</div>
<div id="cc-vln" class="cell" data-execution_count="33">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="op">%</span>matplotlib inline</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>fig, axs <span class="op">=</span> plt.subplots(<span class="dv">1</span>, <span class="dv">3</span>, figsize<span class="op">=</span>(<span class="dv">10</span>,<span class="dv">6</span>),constrained_layout<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a>sc.pl.violin(adata, [<span class="st">'S_score'</span>] , jitter<span class="op">=</span><span class="fl">0.4</span>, groupby <span class="op">=</span> <span class="st">'sample'</span>, ax<span class="op">=</span>axs[<span class="dv">0</span>], show<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a>rotate_xticks(axs[<span class="dv">0</span>])</span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a>sc.pl.violin(adata, [<span class="st">'G2M_score'</span>] , jitter<span class="op">=</span><span class="fl">0.4</span>, groupby <span class="op">=</span> <span class="st">'sample'</span>, ax<span class="op">=</span>axs[<span class="dv">1</span>], show<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb49-8"><a href="#cb49-8" aria-hidden="true" tabindex="-1"></a>rotate_xticks(axs[<span class="dv">1</span>])</span>
<span id="cb49-9"><a href="#cb49-9" aria-hidden="true" tabindex="-1"></a>sc.pl.scatter(adata, x<span class="op">=</span><span class="st">'S_score'</span>, y<span class="op">=</span><span class="st">'G2M_score'</span>, color<span class="op">=</span><span class="st">"phase"</span>, ax <span class="op">=</span> axs[<span class="dv">2</span>])</span>
<span id="cb49-10"><a href="#cb49-10" aria-hidden="true" tabindex="-1"></a>pd.crosstab(adata.obs[<span class="st">'sample'</span>],adata.obs[<span class="st">'phase'</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="qc_overview_files/figure-html/cc-vln-output-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-16"><img src="qc_overview_files/figure-html/cc-vln-output-1.png" id="cc-vln-1" width="809" height="489" class="figure-img"></a></p>
</figure>
</div>
</div>
<div id="cc-vln-2" class="cell-output cell-output-display" data-execution_count="39">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th">phase</th>
<th data-quarto-table-cell-role="th">G1</th>
<th data-quarto-table-cell-role="th">G2M</th>
<th data-quarto-table-cell-role="th">S</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">sample</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">GSE261852_CI82_spl1</td>
<td>2579</td>
<td>3820</td>
<td>5236</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">GSE261852_CI82_spl2</td>
<td>1780</td>
<td>3958</td>
<td>4565</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">GSM6594872_SAM24349928</td>
<td>727</td>
<td>24</td>
<td>42</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">GSM6594883_SAM24374037</td>
<td>4705</td>
<td>433</td>
<td>413</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">GSM6757771_rep1</td>
<td>11492</td>
<td>923</td>
<td>3837</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">GSM6757772_rep2</td>
<td>42285</td>
<td>6686</td>
<td>17289</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">GSM6757773_rep3</td>
<td>2282</td>
<td>413</td>
<td>1005</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">GSM6757774_nuc2</td>
<td>40175</td>
<td>1773</td>
<td>6726</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">GSM6757775_nuc3</td>
<td>34892</td>
<td>2662</td>
<td>8533</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div id="cc-vln2" class="cell" data-execution_count="34">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="op">%</span>matplotlib inline</span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a>fig, axs <span class="op">=</span> plt.subplots(<span class="dv">1</span>, <span class="dv">3</span>, figsize<span class="op">=</span>(<span class="dv">10</span>,<span class="dv">6</span>),constrained_layout<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a>sc.pl.violin(adata, [<span class="st">'S_score2'</span>] , jitter<span class="op">=</span><span class="fl">0.4</span>, groupby <span class="op">=</span> <span class="st">'sample'</span>, ax<span class="op">=</span>axs[<span class="dv">0</span>], show<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb50-5"><a href="#cb50-5" aria-hidden="true" tabindex="-1"></a>rotate_xticks(axs[<span class="dv">0</span>])</span>
<span id="cb50-6"><a href="#cb50-6" aria-hidden="true" tabindex="-1"></a>sc.pl.violin(adata, [<span class="st">'G2M_score2'</span>] , jitter<span class="op">=</span><span class="fl">0.4</span>, groupby <span class="op">=</span> <span class="st">'sample'</span>, ax<span class="op">=</span>axs[<span class="dv">1</span>], show<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb50-7"><a href="#cb50-7" aria-hidden="true" tabindex="-1"></a>rotate_xticks(axs[<span class="dv">1</span>])</span>
<span id="cb50-8"><a href="#cb50-8" aria-hidden="true" tabindex="-1"></a>sc.pl.scatter(adata, x<span class="op">=</span><span class="st">'S_score2'</span>, y<span class="op">=</span><span class="st">'G2M_score2'</span>, color<span class="op">=</span><span class="st">"phase2"</span>, ax <span class="op">=</span> axs[<span class="dv">2</span>])</span>
<span id="cb50-9"><a href="#cb50-9" aria-hidden="true" tabindex="-1"></a>pd.crosstab(adata.obs[<span class="st">'sample'</span>],adata.obs[<span class="st">'phase2'</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="qc_overview_files/figure-html/cc-vln2-output-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-17"><img src="qc_overview_files/figure-html/cc-vln2-output-1.png" id="cc-vln2-1" width="809" height="489" class="figure-img"></a></p>
</figure>
</div>
</div>
<div id="cc-vln2-2" class="cell-output cell-output-display" data-execution_count="40">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th">phase2</th>
<th data-quarto-table-cell-role="th">G1</th>
<th data-quarto-table-cell-role="th">G2M</th>
<th data-quarto-table-cell-role="th">S</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">sample</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">GSE261852_CI82_spl1</td>
<td>5641</td>
<td>3626</td>
<td>2368</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">GSE261852_CI82_spl2</td>
<td>4268</td>
<td>3693</td>
<td>2342</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">GSM6594872_SAM24349928</td>
<td>610</td>
<td>41</td>
<td>142</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">GSM6594883_SAM24374037</td>
<td>3804</td>
<td>614</td>
<td>1133</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">GSM6757771_rep1</td>
<td>9991</td>
<td>1153</td>
<td>5108</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">GSM6757772_rep2</td>
<td>42547</td>
<td>9074</td>
<td>14639</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">GSM6757773_rep3</td>
<td>2521</td>
<td>239</td>
<td>940</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">GSM6757774_nuc2</td>
<td>25087</td>
<td>10526</td>
<td>13061</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">GSM6757775_nuc3</td>
<td>26946</td>
<td>8997</td>
<td>10144</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<p>Clearly different results with all together vs each one separately.</p>
</section>
</section>
<section id="scrublet---doublet-prediction" class="level2">
<h2 class="anchored" data-anchor-id="scrublet---doublet-prediction">Scrublet - Doublet prediction</h2>
<p>For doublet detection, we will use the package <code>Scrublet</code>, so first we need to get the raw counts from <code>adata.raw.X</code> and run scrublet with that matrix. Then we add in the doublet prediction info into our anndata object.</p>
<p>Doublet prediction should be run for each dataset separately, so first we need to split the adata object into 6 separate objects, one per sample and then run scrublet on each of them.</p>
<div id="doublet" class="cell" data-execution_count="35">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scrublet <span class="im">as</span> scr</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a><span class="co"># split per batch into new objects.</span></span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a>batches <span class="op">=</span> adata.obs[<span class="st">'sample'</span>].cat.categories.tolist()</span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a>scrubdata <span class="op">=</span> {}</span>
<span id="cb51-6"><a href="#cb51-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> batch <span class="kw">in</span> batches:</span>
<span id="cb51-7"><a href="#cb51-7" aria-hidden="true" tabindex="-1"></a>    tmp <span class="op">=</span> adata[adata.obs[<span class="st">'sample'</span>] <span class="op">==</span> batch,]</span>
<span id="cb51-8"><a href="#cb51-8" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(batch, <span class="st">":"</span>, tmp.shape[<span class="dv">0</span>], <span class="st">" cells"</span>)</span>
<span id="cb51-9"><a href="#cb51-9" aria-hidden="true" tabindex="-1"></a>    scrub <span class="op">=</span> scr.Scrublet(tmp.raw.X)</span>
<span id="cb51-10"><a href="#cb51-10" aria-hidden="true" tabindex="-1"></a>    out <span class="op">=</span> scrub.scrub_doublets(verbose<span class="op">=</span><span class="va">False</span>, n_prin_comps <span class="op">=</span> <span class="dv">20</span>)</span>
<span id="cb51-11"><a href="#cb51-11" aria-hidden="true" tabindex="-1"></a>    scrubdata[batch] <span class="op">=</span> pd.DataFrame({<span class="st">'doublet_score'</span>:out[<span class="dv">0</span>],<span class="st">'predicted_doublets'</span>:out[<span class="dv">1</span>]},index <span class="op">=</span> tmp.obs.index)</span>
<span id="cb51-12"><a href="#cb51-12" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(scrubdata[batch].predicted_doublets.<span class="bu">sum</span>(), <span class="st">" predicted_doublets"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>GSE261852_CI82_spl1 : 11635  cells
4  predicted_doublets
GSE261852_CI82_spl2 : 10303  cells
669  predicted_doublets
GSM6594872_SAM24349928 : 793  cells
12  predicted_doublets
GSM6594883_SAM24374037 : 5551  cells
119  predicted_doublets
GSM6757771_rep1 : 16252  cells
181  predicted_doublets
GSM6757772_rep2 : 66260  cells
1  predicted_doublets
GSM6757773_rep3 : 3700  cells
5  predicted_doublets
GSM6757774_nuc2 : 48674  cells
12  predicted_doublets
GSM6757775_nuc3 : 46087  cells
0  predicted_doublets</code></pre>
</div>
</div>
<div id="cell-add-doublet" class="cell" data-execution_count="36">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="co"># add predictions to the adata object.</span></span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>scrub_pred <span class="op">=</span> pd.concat(scrubdata.values())</span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">'doublet_scores'</span>] <span class="op">=</span> scrub_pred[<span class="st">'doublet_score'</span>] </span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">'predicted_doublets'</span>] <span class="op">=</span> scrub_pred[<span class="st">'predicted_doublets'</span>] </span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-6"><a href="#cb53-6" aria-hidden="true" tabindex="-1"></a><span class="bu">sum</span>(adata.obs[<span class="st">'predicted_doublets'</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div id="add-doublet" class="cell-output cell-output-display" data-execution_count="42">
<pre><code>1003</code></pre>
</div>
</div>
<p>Plot number of genes per cell split by doublet scores</p>
<div id="cell-plot-doublet" class="cell" data-fig-height="5" data-fig-width="12" data-execution_count="37">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="op">%</span>matplotlib inline</span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">'doublet_info'</span>] <span class="op">=</span> adata.obs[<span class="st">"predicted_doublets"</span>].astype(<span class="bu">str</span>)</span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> sns.violinplot(data<span class="op">=</span>adata.obs, x<span class="op">=</span><span class="st">'sample'</span>, y<span class="op">=</span><span class="st">'n_genes_by_counts'</span>, hue<span class="op">=</span><span class="st">"doublet_info"</span>,  split<span class="op">=</span><span class="va">False</span>, inner<span class="op">=</span><span class="st">"quart"</span>, linewidth<span class="op">=</span><span class="dv">1</span>)  </span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a>rotate_xticks(ax)</span>
<span id="cb55-6"><a href="#cb55-6" aria-hidden="true" tabindex="-1"></a>fig.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="qc_overview_files/figure-html/plot-doublet-output-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-18"><img src="qc_overview_files/figure-html/plot-doublet-output-1.png" id="plot-doublet" width="335" height="376" class="figure-img"></a></p>
</figure>
</div>
</div>
</div>
<p>Filter out the predicted doublets</p>
<div id="cell-doublet-filt" class="cell" data-execution_count="38">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> adata[adata.obs[<span class="st">'doublet_info'</span>] <span class="op">==</span> <span class="st">'False'</span>,:]</span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(adata.shape)</span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">"sample"</span>].value_counts()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>(208252, 26620)</code></pre>
</div>
<div id="doublet-filt" class="cell-output cell-output-display" data-execution_count="44">
<pre><code>GSM6757772_rep2           66259
GSM6757774_nuc2           48662
GSM6757775_nuc3           46087
GSM6757771_rep1           16071
GSE261852_CI82_spl1       11631
GSE261852_CI82_spl2        9634
GSM6594883_SAM24374037     5432
GSM6757773_rep3            3695
GSM6594872_SAM24349928      781
Name: sample, dtype: int64</code></pre>
</div>
</div>
<p>FIX? Run as separate function?</p>
</section>
<section id="umap-and-clustering" class="level2">
<h2 class="anchored" data-anchor-id="umap-and-clustering">Umap and clustering</h2>
<p>Run a quick analysis with default settings in umap and clustering. For now, regression on percent mitochondria an number of genes</p>
<div id="run-umap" class="cell" data-execution_count="39">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>sc.pp.highly_variable_genes(adata, min_mean<span class="op">=</span><span class="fl">0.0125</span>, max_mean<span class="op">=</span><span class="dv">3</span>, min_disp<span class="op">=</span><span class="fl">0.5</span>, batch_key <span class="op">=</span> <span class="st">'sample'</span>)</span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> adata[:, adata.var.highly_variable]</span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a>sc.pp.regress_out(adata, [<span class="st">'n_genes_by_counts'</span>, <span class="st">'pct_counts_mt'</span>])</span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a>sc.pp.scale(adata, max_value<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb59-5"><a href="#cb59-5" aria-hidden="true" tabindex="-1"></a>sc.tl.pca(adata, svd_solver<span class="op">=</span><span class="st">'arpack'</span>)</span>
<span id="cb59-6"><a href="#cb59-6" aria-hidden="true" tabindex="-1"></a>sc.pp.neighbors(adata, n_neighbors<span class="op">=</span><span class="dv">10</span>, n_pcs<span class="op">=</span><span class="dv">40</span>)</span>
<span id="cb59-7"><a href="#cb59-7" aria-hidden="true" tabindex="-1"></a>sc.tl.umap(adata)</span>
<span id="cb59-8"><a href="#cb59-8" aria-hidden="true" tabindex="-1"></a>sc.tl.leiden(adata, resolution <span class="op">=</span> <span class="fl">0.6</span>, key_added <span class="op">=</span> <span class="st">"leiden_0.6"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>extracting highly variable genes
    finished (0:00:11)
--&gt; added
    'highly_variable', boolean vector (adata.var)
    'means', float vector (adata.var)
    'dispersions', float vector (adata.var)
    'dispersions_norm', float vector (adata.var)
regressing out ['n_genes_by_counts', 'pct_counts_mt']
    sparse input is densified and may lead to high memory use
    finished (0:07:22)
computing PCA
    with n_comps=50
    finished (0:00:14)
computing neighbors
    using 'X_pca' with n_pcs = 40
    finished: added to `.uns['neighbors']`
    `.obsp['distances']`, distances for each pair of neighbors
    `.obsp['connectivities']`, weighted adjacency matrix (0:00:37)
computing UMAP
    finished: added
    'X_umap', UMAP coordinates (adata.obsm)
    'umap', UMAP parameters (adata.uns) (0:02:22)
running Leiden clustering
    finished: found 19 clusters and added
    'leiden_0.6', the cluster labels (adata.obs, categorical) (0:00:44)</code></pre>
</div>
</div>
<p>Visualize on umap/clusters:</p>
<ul>
<li>all qc stats</li>
<li>celltypes</li>
<li>Doublet scores</li>
<li>CC scores</li>
</ul>
<div id="plot-umap" class="cell" data-execution_count="40">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="op">%</span>matplotlib inline</span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(adata, color<span class="op">=</span>[<span class="st">'sample'</span>,<span class="st">'leiden_0.6'</span>,<span class="st">'pruned.labels'</span>])</span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a><span class="co"># proprtion of clusters vs samples.</span></span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a>tmp <span class="op">=</span> pd.crosstab(adata.obs[<span class="st">'leiden_0.6'</span>],adata.obs[<span class="st">'sample'</span>], normalize<span class="op">=</span><span class="st">'index'</span>)</span>
<span id="cb61-6"><a href="#cb61-6" aria-hidden="true" tabindex="-1"></a>tmp.plot.bar(stacked<span class="op">=</span><span class="va">True</span>).legend(bbox_to_anchor<span class="op">=</span>(<span class="fl">1.4</span>, <span class="dv">1</span>), loc<span class="op">=</span><span class="st">'upper right'</span>)</span>
<span id="cb61-7"><a href="#cb61-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-8"><a href="#cb61-8" aria-hidden="true" tabindex="-1"></a>tmp <span class="op">=</span> pd.crosstab(adata.obs[<span class="st">'sample'</span>],adata.obs[<span class="st">'leiden_0.6'</span>], normalize<span class="op">=</span><span class="st">'index'</span>)</span>
<span id="cb61-9"><a href="#cb61-9" aria-hidden="true" tabindex="-1"></a>tmp.plot.bar(stacked<span class="op">=</span><span class="va">True</span>).legend(bbox_to_anchor<span class="op">=</span>(<span class="fl">1.4</span>, <span class="dv">1</span>), loc<span class="op">=</span><span class="st">'upper right'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="qc_overview_files/figure-html/plot-umap-output-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-19"><img src="qc_overview_files/figure-html/plot-umap-output-1.png" id="plot-umap-1" width="1222" height="299" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="qc_overview_files/figure-html/plot-umap-output-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-20"><img src="qc_overview_files/figure-html/plot-umap-output-2.png" id="plot-umap-2" width="387" height="303" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="qc_overview_files/figure-html/plot-umap-output-3.png" class="lightbox" data-gallery="quarto-lightbox-gallery-21"><img src="qc_overview_files/figure-html/plot-umap-output-3.png" id="plot-umap-3" width="387" height="492" class="figure-img"></a></p>
</figure>
</div>
</div>
</div>
<p>Plot all metadata</p>
<div id="9afc7d8f" class="cell" data-execution_count="41">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>meta_plot <span class="op">=</span> <span class="bu">list</span>(meta.columns[:<span class="op">-</span><span class="dv">2</span>])</span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(adata, color<span class="op">=</span>meta_plot, ncols<span class="op">=</span><span class="dv">2</span>, cmap <span class="op">=</span> <span class="st">'viridis_r'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="qc_overview_files/figure-html/cell-42-output-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-22"><img src="qc_overview_files/figure-html/cell-42-output-1.png" width="750" height="577" class="figure-img"></a></p>
</figure>
</div>
</div>
</div>
<p>Plot QC stats, doublet scores, CC scores</p>
<div id="d0fc85bb" class="cell" data-execution_count="42">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(adata, color<span class="op">=</span>plot_stats, ncols<span class="op">=</span><span class="dv">3</span>, cmap <span class="op">=</span> <span class="st">'viridis_r'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="qc_overview_files/figure-html/cell-43-output-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-23"><img src="qc_overview_files/figure-html/cell-43-output-1.png" width="978" height="856" class="figure-img"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="save-file" class="level2">
<h2 class="anchored" data-anchor-id="save-file">Save file</h2>
<div id="save" class="cell" data-execution_count="43">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>adata.write_h5ad(os.path.join(resdir,<span class="st">"all_filt.h5ad"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<details>
<summary>
Click here
</summary>
<div id="session" class="cell" data-execution_count="44">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>sc.logging.print_versions()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>-----
anndata     0.10.8
scanpy      1.10.3
-----
CoreFoundation              NA
Foundation                  NA
PIL                         10.3.0
PyObjCTools                 NA
annoy                       NA
anyio                       NA
appnope                     0.1.4
array_api_compat            1.10.0
arrow                       1.3.0
asttokens                   NA
attr                        25.1.0
attrs                       25.1.0
babel                       2.17.0
brotli                      1.1.0
certifi                     2025.01.31
cffi                        1.17.1
charset_normalizer          3.4.1
colorama                    0.4.6
comm                        0.2.2
cycler                      0.12.1
cython_runtime              NA
dateutil                    2.9.0.post0
debugpy                     1.8.12
decorator                   5.2.1
defusedxml                  0.7.1
exceptiongroup              1.2.2
executing                   2.1.0
fastjsonschema              NA
fqdn                        NA
google                      NA
h5py                        3.13.0
idna                        3.10
igraph                      0.11.5
ipykernel                   6.29.5
isoduration                 NA
jedi                        0.19.2
jinja2                      3.1.5
joblib                      1.4.2
json5                       0.10.0
jsonpointer                 3.0.0
jsonschema                  4.23.0
jsonschema_specifications   NA
jupyter_events              0.12.0
jupyter_server              2.15.0
jupyterlab_server           2.27.3
kaleido                     0.2.1
kiwisolver                  1.4.7
lazy_loader                 0.4
legacy_api_wrap             NA
leidenalg                   0.10.2
llvmlite                    0.44.0
markupsafe                  3.0.2
matplotlib                  3.10.0
matplotlib_inline           0.1.7
mpl_toolkits                NA
narwhals                    1.27.1
natsort                     8.4.0
nbformat                    5.10.4
numba                       0.61.0
numpy                       1.26.4
objc                        11.0
overrides                   NA
packaging                   24.2
pandas                      1.5.3
parso                       0.8.4
patsy                       1.0.1
pickleshare                 0.7.5
platformdirs                4.3.6
plotly                      6.0.0
prometheus_client           NA
prompt_toolkit              3.0.50
psutil                      6.1.1
pure_eval                   0.2.3
pycparser                   2.22
pydev_ipython               NA
pydevconsole                NA
pydevd                      3.2.3
pydevd_file_utils           NA
pydevd_plugins              NA
pydevd_tracing              NA
pygments                    2.19.1
pynndescent                 0.5.13
pyparsing                   3.2.1
pythonjsonlogger            NA
pytz                        2025.1
referencing                 NA
requests                    2.32.3
rfc3339_validator           0.1.4
rfc3986_validator           0.1.1
rpds                        NA
scipy                       1.14.1
scrublet                    NA
seaborn                     0.13.2
send2trash                  NA
session_info                1.0.0
six                         1.17.0
skimage                     0.25.0
sklearn                     1.1.3
sniffio                     1.3.1
socks                       1.7.1
sparse                      0.15.5
sphinxcontrib               NA
stack_data                  0.6.3
statsmodels                 0.14.4
texttable                   1.7.0
threadpoolctl               3.5.0
torch                       2.4.0.post101
torchgen                    NA
tornado                     6.4.2
tqdm                        4.67.1
traitlets                   5.14.3
typing_extensions           NA
umap                        0.5.7
uri_template                NA
urllib3                     2.3.0
wcwidth                     0.2.13
webcolors                   NA
websocket                   1.8.0
yaml                        6.0.2
zmq                         26.2.1
zoneinfo                    NA
zstandard                   0.23.0
-----
IPython             8.32.0
jupyter_client      8.6.3
jupyter_core        5.7.2
jupyterlab          4.3.5
notebook            7.3.2
-----
Python 3.10.10 | packaged by conda-forge | (main, Mar 24 2023, 20:17:34) [Clang 14.0.6 ]
macOS-15.3.1-x86_64-i386-64bit
-----
Session information updated at 2025-03-07 13:05</code></pre>
</div>
</div>
</details>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<script>var lightboxQuarto = GLightbox({"openEffect":"zoom","selector":".lightbox","descPosition":"bottom","closeEffect":"zoom","loop":false});
(function() {
  let previousOnload = window.onload;
  window.onload = () => {
    if (previousOnload) {
      previousOnload();
    }
    lightboxQuarto.on('slide_before_load', (data) => {
      const { slideIndex, slideNode, slideConfig, player, trigger } = data;
      const href = trigger.getAttribute('href');
      if (href !== null) {
        const imgEl = window.document.querySelector(`a[href="${href}"] img`);
        if (imgEl !== null) {
          const srcAttr = imgEl.getAttribute("src");
          if (srcAttr && srcAttr.startsWith("data:")) {
            slideConfig.href = srcAttr;
          }
        }
      } 
    });
  
    lightboxQuarto.on('slide_after_load', (data) => {
      const { slideIndex, slideNode, slideConfig, player, trigger } = data;
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(slideNode);
      }
    });
  
  };
  
})();
          </script>




</body></html>