<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="description" content="Visualizing cellbender, soupX and decontX output">

<title>Ambient RNA removal</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="ambient_methods_overview_hsa_parse_files/libs/clipboard/clipboard.min.js"></script>
<script src="ambient_methods_overview_hsa_parse_files/libs/quarto-html/quarto.js"></script>
<script src="ambient_methods_overview_hsa_parse_files/libs/quarto-html/popper.min.js"></script>
<script src="ambient_methods_overview_hsa_parse_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="ambient_methods_overview_hsa_parse_files/libs/quarto-html/anchor.min.js"></script>
<link href="ambient_methods_overview_hsa_parse_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="ambient_methods_overview_hsa_parse_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="ambient_methods_overview_hsa_parse_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="ambient_methods_overview_hsa_parse_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="ambient_methods_overview_hsa_parse_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script src="ambient_methods_overview_hsa_parse_files/libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="ambient_methods_overview_hsa_parse_files/libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="ambient_methods_overview_hsa_parse_files/libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet">


</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active" data-toc-expanded="1">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#read-in-data." id="toc-read-in-data." class="nav-link active" data-scroll-target="#read-in-data.">Read in data.</a></li>
  <li><a href="#test-tabset" id="toc-test-tabset" class="nav-link" data-scroll-target="#test-tabset">Test tabset</a></li>
  <li><a href="#cell-filtering-cellbender" id="toc-cell-filtering-cellbender" class="nav-link" data-scroll-target="#cell-filtering-cellbender">Cell filtering cellbender</a></li>
  <li><a href="#filtering-stats-all-methods" id="toc-filtering-stats-all-methods" class="nav-link" data-scroll-target="#filtering-stats-all-methods">Filtering stats all methods</a></li>
  <li><a href="#gene-stats" id="toc-gene-stats" class="nav-link" data-scroll-target="#gene-stats">Gene stats</a></li>
  <li><a href="#mito-etc." id="toc-mito-etc." class="nav-link" data-scroll-target="#mito-etc.">Mito, etc.</a></li>
  <li><a href="#biotype-proportions" id="toc-biotype-proportions" class="nav-link" data-scroll-target="#biotype-proportions">Biotype proportions</a></li>
  <li><a href="#check-top-genes" id="toc-check-top-genes" class="nav-link" data-scroll-target="#check-top-genes">Check top genes</a></li>
  <li><a href="#parse-data" id="toc-parse-data" class="nav-link" data-scroll-target="#parse-data">Parse data</a>
  <ul class="collapse">
  <li><a href="#cell-filtering" id="toc-cell-filtering" class="nav-link" data-scroll-target="#cell-filtering">Cell filtering</a></li>
  <li><a href="#ambient-profile" id="toc-ambient-profile" class="nav-link" data-scroll-target="#ambient-profile">Ambient profile</a></li>
  </ul></li>
  </ul>
</nav>
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Ambient RNA removal</h1>
</div>

<div>
  <div class="description">
    Visualizing cellbender, soupX and decontX output
  </div>
</div>


<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>Human samples from Gezelius et al.&nbsp;With all the different parse subsamples.</p>
<div id="yaml-load" class="cell" data-execution_count="1">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> yaml</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> <span class="bu">open</span>(<span class="st">'./settings_hsa_parse.yml'</span>, <span class="st">'r'</span>) <span class="im">as</span> f:</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>    settings <span class="op">=</span> yaml.load(f, Loader<span class="op">=</span>yaml.SafeLoader)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="libraries" class="cell" data-execution_count="2">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> sys</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> glob</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> warnings</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="co">#import pyreadr</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> IPython.display <span class="im">import</span> display, Markdown</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>warnings.simplefilter(action<span class="op">=</span><span class="st">'ignore'</span>, category<span class="op">=</span><span class="pp">Warning</span>)</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a><span class="co"># verbosity: errors (0), warnings (1), info (2), hints (3)</span></span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a>sc.settings.verbosity <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>sc.settings.set_figure_params(dpi<span class="op">=</span><span class="dv">80</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<section id="read-in-data." class="level3">
<h3 class="anchored" data-anchor-id="read-in-data.">Read in data.</h3>
<p>First, read in the annotation file for biotypes.</p>
<p>Make broader groupings:</p>
<ul>
<li>Mito genes - both PC and nc.</li>
<li>PC genes - includes also IG/TR genes</li>
<li>lncRNA</li>
<li>other - all other categories.</li>
</ul>
<div id="load-annot" class="cell" data-execution_count="3">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>indir <span class="op">=</span> <span class="st">"/Users/asabjor/projects/sc-devop/scQC/data/"</span> <span class="op">+</span> settings[<span class="st">'species'</span>]</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>cbdir <span class="op">=</span> os.path.join(indir,<span class="st">"output"</span>)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>pdir <span class="op">=</span> os.path.join(indir,<span class="st">"pipeline"</span>)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>annot_file <span class="op">=</span> os.path.join(cbdir, settings[<span class="st">'species'</span>] <span class="op">+</span> <span class="st">"_annotation.csv"</span>)</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> os.path.exists(annot_file):</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>    annot <span class="op">=</span> pd.read_csv(annot_file, index_col<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span>:   </span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> settings[<span class="st">'species'</span>] <span class="op">==</span> <span class="st">"human"</span>:</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>        annot <span class="op">=</span> sc.queries.biomart_annotations(<span class="st">"hsapiens"</span>, [<span class="st">"ensembl_gene_id"</span>, <span class="st">"external_gene_name"</span>, <span class="st">"start_position"</span>, <span class="st">"end_position"</span>, <span class="st">"chromosome_name"</span>, <span class="st">"gene_biotype"</span>] )</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>        annot.to_csv(annot_file)</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>    <span class="cf">elif</span> settings[<span class="st">'species'</span>] <span class="op">==</span> <span class="st">"mouse"</span>:</span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>        annot <span class="op">=</span> sc.queries.biomart_annotations(<span class="st">"mmusculus"</span>, [<span class="st">"ensembl_gene_id"</span>, <span class="st">"external_gene_name"</span>, <span class="st">"start_position"</span>, <span class="st">"end_position"</span>, <span class="st">"chromosome_name"</span>, <span class="st">"gene_biotype"</span>] )</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>        annot.to_csv(annot_file)</span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>: </span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="st">"unknown speceis "</span> <span class="op">+</span> settings[<span class="st">'species'</span>] )</span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a>        sys.exit()</span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a><span class="co"># make slim biotype</span></span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a>annot[<span class="st">"gene_class"</span>]<span class="op">=</span> <span class="st">"other"</span></span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a>pc_genes <span class="op">=</span> [<span class="st">"protein_coding"</span>,<span class="st">"IG_V_gene"</span>,<span class="st">"IG_D_gene"</span>,<span class="st">"IG_J_gene"</span>,<span class="st">"IG_C_gene"</span>,<span class="st">"IG_LV_gene"</span>, <span class="st">"TR_V_gene"</span>,<span class="st">"TR_J_gene"</span>,<span class="st">"TR_C_gene"</span>,<span class="st">"TR_D_gene"</span> ]</span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-27"><a href="#cb3-27" aria-hidden="true" tabindex="-1"></a>annot[<span class="st">"gene_class"</span>][annot.gene_biotype.isin(pc_genes)] <span class="op">=</span> <span class="st">"pc"</span></span>
<span id="cb3-28"><a href="#cb3-28" aria-hidden="true" tabindex="-1"></a>annot[<span class="st">"gene_class"</span>][annot.gene_biotype <span class="op">==</span> <span class="st">"lncRNA"</span>] <span class="op">=</span> <span class="st">"lncRNA"</span></span>
<span id="cb3-29"><a href="#cb3-29" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> settings[<span class="st">'species'</span>] <span class="op">==</span> <span class="st">"human"</span>:</span>
<span id="cb3-30"><a href="#cb3-30" aria-hidden="true" tabindex="-1"></a>    annot[<span class="st">"gene_class"</span>][annot.external_gene_name.<span class="bu">str</span>.startswith(<span class="st">'MT-'</span>).fillna(<span class="va">False</span>)] <span class="op">=</span> <span class="st">"mt"</span></span>
<span id="cb3-31"><a href="#cb3-31" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span>:</span>
<span id="cb3-32"><a href="#cb3-32" aria-hidden="true" tabindex="-1"></a>    annot[<span class="st">"gene_class"</span>][annot.external_gene_name.<span class="bu">str</span>.startswith(<span class="st">'mt-'</span>).fillna(<span class="va">False</span>)] <span class="op">=</span> <span class="st">"mt"</span></span>
<span id="cb3-33"><a href="#cb3-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-34"><a href="#cb3-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-35"><a href="#cb3-35" aria-hidden="true" tabindex="-1"></a><span class="co"># define list of all classes.</span></span>
<span id="cb3-36"><a href="#cb3-36" aria-hidden="true" tabindex="-1"></a>gene_classes <span class="op">=</span> [<span class="st">"mt"</span>,<span class="st">"pc"</span>,<span class="st">"lncRNA"</span>,<span class="st">"other"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Then read in all the raw and cellbender, soupX, decontX files. For soupX select 2 different resolutions (0.2, 0.8).</p>
<p>For soupX and decontX output is not in integer counts, round the values!</p>
<div id="functions" class="cell" data-execution_count="4">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># does not work to read in the cellbender files from GSE244142 with sc.read_10x_h5</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="co"># gives zero genes.</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="co"># minor modifications of the scanpy function, now works to load the data!</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> h5py</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> anndata <span class="im">as</span> ad</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> _collect_datasets(dsets: <span class="bu">dict</span>, group: h5py.Group):</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> k, v <span class="kw">in</span> group.items():</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> <span class="bu">isinstance</span>(v, h5py.Dataset):</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>            dsets[k] <span class="op">=</span> v[()]</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>        <span class="cf">else</span>:</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>            _collect_datasets(dsets, v)</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> _read_v3_10x_h5(filename, <span class="op">*</span>, start<span class="op">=</span><span class="va">None</span>):</span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""</span></span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a><span class="co">    Read hdf5 file from Cell Ranger v3 or later versions.</span></span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a>    <span class="cf">with</span> h5py.File(<span class="bu">str</span>(filename), <span class="st">"r"</span>) <span class="im">as</span> f:</span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a>        <span class="cf">try</span>:</span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a>            dsets <span class="op">=</span> {}</span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a>            _collect_datasets(dsets, f[<span class="st">"matrix"</span>])</span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-24"><a href="#cb4-24" aria-hidden="true" tabindex="-1"></a>            <span class="im">from</span> scipy.sparse <span class="im">import</span> csr_matrix</span>
<span id="cb4-25"><a href="#cb4-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-26"><a href="#cb4-26" aria-hidden="true" tabindex="-1"></a>            M, N <span class="op">=</span> dsets[<span class="st">"shape"</span>]</span>
<span id="cb4-27"><a href="#cb4-27" aria-hidden="true" tabindex="-1"></a>            data <span class="op">=</span> dsets[<span class="st">"data"</span>]</span>
<span id="cb4-28"><a href="#cb4-28" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> dsets[<span class="st">"data"</span>].dtype <span class="op">==</span> np.dtype(<span class="st">"int32"</span>):</span>
<span id="cb4-29"><a href="#cb4-29" aria-hidden="true" tabindex="-1"></a>                data <span class="op">=</span> dsets[<span class="st">"data"</span>].view(<span class="st">"float32"</span>)</span>
<span id="cb4-30"><a href="#cb4-30" aria-hidden="true" tabindex="-1"></a>                data[:] <span class="op">=</span> dsets[<span class="st">"data"</span>]</span>
<span id="cb4-31"><a href="#cb4-31" aria-hidden="true" tabindex="-1"></a>            matrix <span class="op">=</span> csr_matrix(</span>
<span id="cb4-32"><a href="#cb4-32" aria-hidden="true" tabindex="-1"></a>                (data, dsets[<span class="st">"indices"</span>], dsets[<span class="st">"indptr"</span>]),</span>
<span id="cb4-33"><a href="#cb4-33" aria-hidden="true" tabindex="-1"></a>                shape<span class="op">=</span>(N, M),</span>
<span id="cb4-34"><a href="#cb4-34" aria-hidden="true" tabindex="-1"></a>            )</span>
<span id="cb4-35"><a href="#cb4-35" aria-hidden="true" tabindex="-1"></a>            obs_dict <span class="op">=</span> {<span class="st">"obs_names"</span>: dsets[<span class="st">"barcodes"</span>].astype(<span class="bu">str</span>)}</span>
<span id="cb4-36"><a href="#cb4-36" aria-hidden="true" tabindex="-1"></a>            var_dict <span class="op">=</span> {<span class="st">"var_names"</span>: dsets[<span class="st">"name"</span>].astype(<span class="bu">str</span>)}</span>
<span id="cb4-37"><a href="#cb4-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-38"><a href="#cb4-38" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> <span class="st">"gene_id"</span> <span class="kw">not</span> <span class="kw">in</span> dsets:</span>
<span id="cb4-39"><a href="#cb4-39" aria-hidden="true" tabindex="-1"></a>                <span class="co"># Read metadata specific to a feature-barcode matrix</span></span>
<span id="cb4-40"><a href="#cb4-40" aria-hidden="true" tabindex="-1"></a>                var_dict[<span class="st">"gene_ids"</span>] <span class="op">=</span> dsets[<span class="st">"id"</span>].astype(<span class="bu">str</span>)</span>
<span id="cb4-41"><a href="#cb4-41" aria-hidden="true" tabindex="-1"></a>            <span class="cf">else</span>:</span>
<span id="cb4-42"><a href="#cb4-42" aria-hidden="true" tabindex="-1"></a>                <span class="co"># Read metadata specific to a probe-barcode matrix</span></span>
<span id="cb4-43"><a href="#cb4-43" aria-hidden="true" tabindex="-1"></a>                var_dict.update(</span>
<span id="cb4-44"><a href="#cb4-44" aria-hidden="true" tabindex="-1"></a>                    {</span>
<span id="cb4-45"><a href="#cb4-45" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"gene_ids"</span>: dsets[<span class="st">"gene_id"</span>].astype(<span class="bu">str</span>),</span>
<span id="cb4-46"><a href="#cb4-46" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"probe_ids"</span>: dsets[<span class="st">"id"</span>].astype(<span class="bu">str</span>),</span>
<span id="cb4-47"><a href="#cb4-47" aria-hidden="true" tabindex="-1"></a>                    }</span>
<span id="cb4-48"><a href="#cb4-48" aria-hidden="true" tabindex="-1"></a>                )</span>
<span id="cb4-49"><a href="#cb4-49" aria-hidden="true" tabindex="-1"></a>            var_dict[<span class="st">"feature_types"</span>] <span class="op">=</span> dsets[<span class="st">"feature_type"</span>].astype(<span class="bu">str</span>)</span>
<span id="cb4-50"><a href="#cb4-50" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> <span class="st">"filtered_barcodes"</span> <span class="kw">in</span> f[<span class="st">"matrix"</span>]:</span>
<span id="cb4-51"><a href="#cb4-51" aria-hidden="true" tabindex="-1"></a>                obs_dict[<span class="st">"filtered_barcodes"</span>] <span class="op">=</span> dsets[<span class="st">"filtered_barcodes"</span>].astype(<span class="bu">bool</span>)</span>
<span id="cb4-52"><a href="#cb4-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-53"><a href="#cb4-53" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> <span class="st">"features"</span> <span class="kw">in</span> f[<span class="st">"matrix"</span>]:</span>
<span id="cb4-54"><a href="#cb4-54" aria-hidden="true" tabindex="-1"></a>                var_dict.update(</span>
<span id="cb4-55"><a href="#cb4-55" aria-hidden="true" tabindex="-1"></a>                    (</span>
<span id="cb4-56"><a href="#cb4-56" aria-hidden="true" tabindex="-1"></a>                        feature_metadata_name,</span>
<span id="cb4-57"><a href="#cb4-57" aria-hidden="true" tabindex="-1"></a>                        dsets[feature_metadata_name].astype(</span>
<span id="cb4-58"><a href="#cb4-58" aria-hidden="true" tabindex="-1"></a>                            <span class="bu">bool</span> <span class="cf">if</span> feature_metadata_item.dtype.kind <span class="op">==</span> <span class="st">"b"</span> <span class="cf">else</span> <span class="bu">str</span></span>
<span id="cb4-59"><a href="#cb4-59" aria-hidden="true" tabindex="-1"></a>                        ),</span>
<span id="cb4-60"><a href="#cb4-60" aria-hidden="true" tabindex="-1"></a>                    )</span>
<span id="cb4-61"><a href="#cb4-61" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">for</span> feature_metadata_name, feature_metadata_item <span class="kw">in</span> f[<span class="st">"matrix"</span>][</span>
<span id="cb4-62"><a href="#cb4-62" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"features"</span></span>
<span id="cb4-63"><a href="#cb4-63" aria-hidden="true" tabindex="-1"></a>                    ].items()</span>
<span id="cb4-64"><a href="#cb4-64" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">if</span> <span class="bu">isinstance</span>(feature_metadata_item, h5py.Dataset)</span>
<span id="cb4-65"><a href="#cb4-65" aria-hidden="true" tabindex="-1"></a>                    <span class="kw">and</span> feature_metadata_name</span>
<span id="cb4-66"><a href="#cb4-66" aria-hidden="true" tabindex="-1"></a>                    <span class="kw">not</span> <span class="kw">in</span> [</span>
<span id="cb4-67"><a href="#cb4-67" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"name"</span>,</span>
<span id="cb4-68"><a href="#cb4-68" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"feature_type"</span>,</span>
<span id="cb4-69"><a href="#cb4-69" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"id"</span>,</span>
<span id="cb4-70"><a href="#cb4-70" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"gene_id"</span>,</span>
<span id="cb4-71"><a href="#cb4-71" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"_all_tag_keys"</span>,</span>
<span id="cb4-72"><a href="#cb4-72" aria-hidden="true" tabindex="-1"></a>                    ]</span>
<span id="cb4-73"><a href="#cb4-73" aria-hidden="true" tabindex="-1"></a>                )</span>
<span id="cb4-74"><a href="#cb4-74" aria-hidden="true" tabindex="-1"></a>            <span class="cf">else</span>:</span>
<span id="cb4-75"><a href="#cb4-75" aria-hidden="true" tabindex="-1"></a>                msg <span class="op">=</span> <span class="st">"10x h5 has no features group"</span></span>
<span id="cb4-76"><a href="#cb4-76" aria-hidden="true" tabindex="-1"></a>                <span class="cf">raise</span> <span class="pp">ValueError</span>(msg)</span>
<span id="cb4-77"><a href="#cb4-77" aria-hidden="true" tabindex="-1"></a>            adata <span class="op">=</span> ad.AnnData(</span>
<span id="cb4-78"><a href="#cb4-78" aria-hidden="true" tabindex="-1"></a>                matrix,</span>
<span id="cb4-79"><a href="#cb4-79" aria-hidden="true" tabindex="-1"></a>                obs<span class="op">=</span>obs_dict,</span>
<span id="cb4-80"><a href="#cb4-80" aria-hidden="true" tabindex="-1"></a>                var<span class="op">=</span>var_dict,</span>
<span id="cb4-81"><a href="#cb4-81" aria-hidden="true" tabindex="-1"></a>            )</span>
<span id="cb4-82"><a href="#cb4-82" aria-hidden="true" tabindex="-1"></a>            <span class="co">#logg.info("", time=start)</span></span>
<span id="cb4-83"><a href="#cb4-83" aria-hidden="true" tabindex="-1"></a>            <span class="cf">return</span> adata</span>
<span id="cb4-84"><a href="#cb4-84" aria-hidden="true" tabindex="-1"></a>        <span class="cf">except</span> <span class="pp">KeyError</span>:</span>
<span id="cb4-85"><a href="#cb4-85" aria-hidden="true" tabindex="-1"></a>            msg <span class="op">=</span> <span class="st">"File is missing one or more required datasets."</span></span>
<span id="cb4-86"><a href="#cb4-86" aria-hidden="true" tabindex="-1"></a>            <span class="cf">raise</span> <span class="pp">Exception</span>(msg)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="load-data" class="cell" data-execution_count="5">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>alldata <span class="op">=</span> {}</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>emptydrops_genes <span class="op">=</span> {}</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> sample <span class="kw">in</span> settings[<span class="st">'samples'</span>].keys():</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(sample)</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>    rawfile <span class="op">=</span> settings[<span class="st">'samples'</span>][sample][<span class="st">'raw_file'</span>]</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>    f <span class="op">=</span> os.path.split(rawfile)</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>    read_cb_flag <span class="op">=</span> <span class="va">True</span> <span class="co"># with .mtx infiles the cellbender output is different cannot be read with read_10x-h5</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="st">"raw"</span> <span class="kw">in</span> f[<span class="dv">1</span>]:</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>        raw <span class="op">=</span> sc.read_10x_h5(rawfile)</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>        read_cb_flag <span class="op">=</span> <span class="va">True</span>    </span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>    <span class="cf">elif</span> f[<span class="dv">1</span>] <span class="op">==</span> <span class="st">"matrix.mtx.gz"</span>:</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>        raw <span class="op">=</span> sc.read_10x_mtx(f[<span class="dv">0</span>])</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>        read_cb_flag <span class="op">=</span> <span class="va">False</span> </span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>    <span class="cf">elif</span> <span class="st">"mtx.gz"</span> <span class="kw">in</span> f[<span class="dv">1</span>]: </span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a>        raw <span class="op">=</span> sc.read_10x_mtx(f[<span class="dv">0</span>],prefix <span class="op">=</span> fname<span class="op">+</span><span class="st">"_"</span>)</span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a>        read_cb_flag <span class="op">=</span> <span class="va">False</span> </span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a>    fname <span class="op">=</span> sample</span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a>    raw.var_names_make_unique()</span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a>    <span class="co"># remove zero barcodes</span></span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a>    raw <span class="op">=</span> raw[raw.X.<span class="bu">sum</span>(axis<span class="op">=</span><span class="dv">1</span>)<span class="op">&gt;</span><span class="dv">0</span>,:]</span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a>    <span class="co"># read in the cellbender file.</span></span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true" tabindex="-1"></a>    cb_file <span class="op">=</span> os.path.join(cbdir,sample,<span class="st">"cellbender"</span>,<span class="st">"cellbender_out_filtered.h5"</span>)</span>
<span id="cb5-25"><a href="#cb5-25" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> os.path.exists(cb_file):</span>
<span id="cb5-26"><a href="#cb5-26" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> read_cb_flag:</span>
<span id="cb5-27"><a href="#cb5-27" aria-hidden="true" tabindex="-1"></a>            cb <span class="op">=</span> sc.read_10x_h5(cb_file)</span>
<span id="cb5-28"><a href="#cb5-28" aria-hidden="true" tabindex="-1"></a>        <span class="cf">else</span>:   </span>
<span id="cb5-29"><a href="#cb5-29" aria-hidden="true" tabindex="-1"></a>            cb <span class="op">=</span> _read_v3_10x_h5(cb_file)</span>
<span id="cb5-30"><a href="#cb5-30" aria-hidden="true" tabindex="-1"></a>        cb.var_names_make_unique()</span>
<span id="cb5-31"><a href="#cb5-31" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:   </span>
<span id="cb5-32"><a href="#cb5-32" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="st">"No cellbender file:"</span> <span class="op">+</span> cb_file)</span>
<span id="cb5-33"><a href="#cb5-33" aria-hidden="true" tabindex="-1"></a>        <span class="cf">continue</span></span>
<span id="cb5-34"><a href="#cb5-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-35"><a href="#cb5-35" aria-hidden="true" tabindex="-1"></a>    <span class="co"># for multiseq data, cellbender also includes the HTOs!</span></span>
<span id="cb5-36"><a href="#cb5-36" aria-hidden="true" tabindex="-1"></a>    cb <span class="op">=</span> cb[:,raw.var_names]</span>
<span id="cb5-37"><a href="#cb5-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-38"><a href="#cb5-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-39"><a href="#cb5-39" aria-hidden="true" tabindex="-1"></a>    <span class="co"># add in biotype info</span></span>
<span id="cb5-40"><a href="#cb5-40" aria-hidden="true" tabindex="-1"></a>    t <span class="op">=</span> cb.var.merge(annot, left_on<span class="op">=</span><span class="st">"gene_ids"</span>, right_on<span class="op">=</span><span class="st">"ensembl_gene_id"</span>, how<span class="op">=</span><span class="st">'left'</span>)</span>
<span id="cb5-41"><a href="#cb5-41" aria-hidden="true" tabindex="-1"></a>    t.index <span class="op">=</span> cb.var.index</span>
<span id="cb5-42"><a href="#cb5-42" aria-hidden="true" tabindex="-1"></a>    <span class="co"># add in "unnass" for all nas in gene_class</span></span>
<span id="cb5-43"><a href="#cb5-43" aria-hidden="true" tabindex="-1"></a>    t.gene_class[t.gene_class.isna()] <span class="op">=</span> <span class="st">"other"</span></span>
<span id="cb5-44"><a href="#cb5-44" aria-hidden="true" tabindex="-1"></a>    cb.var <span class="op">=</span> t</span>
<span id="cb5-45"><a href="#cb5-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-46"><a href="#cb5-46" aria-hidden="true" tabindex="-1"></a>    <span class="co"># define ribo hb genes. not in biotypes</span></span>
<span id="cb5-47"><a href="#cb5-47" aria-hidden="true" tabindex="-1"></a>    cb.var[<span class="st">'ribo'</span>] <span class="op">=</span> cb.var_names.<span class="bu">str</span>.startswith((<span class="st">"Rps"</span>,<span class="st">"Rpl"</span>))</span>
<span id="cb5-48"><a href="#cb5-48" aria-hidden="true" tabindex="-1"></a>    cb.var[<span class="st">'hb'</span>] <span class="op">=</span> cb.var_names.<span class="bu">str</span>.contains((<span class="st">"^Hb[^(p|e|s)]"</span>))</span>
<span id="cb5-49"><a href="#cb5-49" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> bt <span class="kw">in</span> gene_classes:</span>
<span id="cb5-50"><a href="#cb5-50" aria-hidden="true" tabindex="-1"></a>        cb.var[bt] <span class="op">=</span> cb.var[<span class="st">"gene_class"</span>] <span class="op">==</span> bt</span>
<span id="cb5-51"><a href="#cb5-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-52"><a href="#cb5-52" aria-hidden="true" tabindex="-1"></a>    qc_vars <span class="op">=</span>  gene_classes <span class="op">+</span> [<span class="st">"ribo"</span>,<span class="st">"hb"</span>]           </span>
<span id="cb5-53"><a href="#cb5-53" aria-hidden="true" tabindex="-1"></a>    raw.var <span class="op">=</span> cb.var</span>
<span id="cb5-54"><a href="#cb5-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-55"><a href="#cb5-55" aria-hidden="true" tabindex="-1"></a>    <span class="co"># calculate stats for barcode rank plots from the full raw matrix.</span></span>
<span id="cb5-56"><a href="#cb5-56" aria-hidden="true" tabindex="-1"></a>    sc.pp.calculate_qc_metrics(raw, inplace<span class="op">=</span><span class="va">True</span>, log1p<span class="op">=</span><span class="va">False</span>, qc_vars<span class="op">=</span>qc_vars)</span>
<span id="cb5-57"><a href="#cb5-57" aria-hidden="true" tabindex="-1"></a>        <span class="co"># add the stats to the cb object</span></span>
<span id="cb5-58"><a href="#cb5-58" aria-hidden="true" tabindex="-1"></a>    cb.uns[<span class="st">"raw_obs"</span>] <span class="op">=</span> raw.obs.copy()</span>
<span id="cb5-59"><a href="#cb5-59" aria-hidden="true" tabindex="-1"></a>    cb.uns[<span class="st">"raw_var"</span>] <span class="op">=</span> raw.var.copy()</span>
<span id="cb5-60"><a href="#cb5-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-61"><a href="#cb5-61" aria-hidden="true" tabindex="-1"></a>    <span class="co">#filter for same cells in raw and recalculate the stats. </span></span>
<span id="cb5-62"><a href="#cb5-62" aria-hidden="true" tabindex="-1"></a>    <span class="co"># add in as a layer</span></span>
<span id="cb5-63"><a href="#cb5-63" aria-hidden="true" tabindex="-1"></a>    raw <span class="op">=</span> raw[cb.obs_names,:]</span>
<span id="cb5-64"><a href="#cb5-64" aria-hidden="true" tabindex="-1"></a>    cb.layers[<span class="st">"raw"</span>]<span class="op">=</span>raw.X.copy()</span>
<span id="cb5-65"><a href="#cb5-65" aria-hidden="true" tabindex="-1"></a>    cb.layers[<span class="st">"CB"</span>]<span class="op">=</span>cb.X.copy()</span>
<span id="cb5-66"><a href="#cb5-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-67"><a href="#cb5-67" aria-hidden="true" tabindex="-1"></a>    <span class="co"># read in the other methods. For soupX, 2 resolutions, 0.2 and 0.8</span></span>
<span id="cb5-68"><a href="#cb5-68" aria-hidden="true" tabindex="-1"></a>    <span class="co"># OBS! DX and SX does not have integers, round the values!!</span></span>
<span id="cb5-69"><a href="#cb5-69" aria-hidden="true" tabindex="-1"></a>    sx_file1 <span class="op">=</span> os.path.join(cbdir,fname,<span class="st">"soupX"</span>,<span class="st">"soupX_cormat_0.2.h5"</span>)</span>
<span id="cb5-70"><a href="#cb5-70" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> os.path.exists(sx_file1):</span>
<span id="cb5-71"><a href="#cb5-71" aria-hidden="true" tabindex="-1"></a>        tmp <span class="op">=</span> sc.read_10x_h5(sx_file1)</span>
<span id="cb5-72"><a href="#cb5-72" aria-hidden="true" tabindex="-1"></a>        cb.layers[<span class="st">"SX02"</span>] <span class="op">=</span> np.<span class="bu">round</span>(tmp.X.copy()) </span>
<span id="cb5-73"><a href="#cb5-73" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>: </span>
<span id="cb5-74"><a href="#cb5-74" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="st">"no file "</span> <span class="op">+</span> sx_file1)</span>
<span id="cb5-75"><a href="#cb5-75" aria-hidden="true" tabindex="-1"></a>        tmp <span class="op">=</span> cb.X.copy()</span>
<span id="cb5-76"><a href="#cb5-76" aria-hidden="true" tabindex="-1"></a>        tmp[:,:] <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb5-77"><a href="#cb5-77" aria-hidden="true" tabindex="-1"></a>        cb.layers[<span class="st">"SX02"</span>] <span class="op">=</span> tmp.copy()</span>
<span id="cb5-78"><a href="#cb5-78" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-79"><a href="#cb5-79" aria-hidden="true" tabindex="-1"></a>    sx_file2 <span class="op">=</span> os.path.join(cbdir,fname,<span class="st">"soupX"</span>,<span class="st">"soupX_cormat_0.8.h5"</span>)</span>
<span id="cb5-80"><a href="#cb5-80" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> os.path.exists(sx_file2):</span>
<span id="cb5-81"><a href="#cb5-81" aria-hidden="true" tabindex="-1"></a>        tmp <span class="op">=</span> sc.read_10x_h5(sx_file2)</span>
<span id="cb5-82"><a href="#cb5-82" aria-hidden="true" tabindex="-1"></a>        cb.layers[<span class="st">"SX08"</span>] <span class="op">=</span> np.<span class="bu">round</span>(tmp.X.copy())        </span>
<span id="cb5-83"><a href="#cb5-83" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>: </span>
<span id="cb5-84"><a href="#cb5-84" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="st">"no file "</span> <span class="op">+</span> sx_file2)    </span>
<span id="cb5-85"><a href="#cb5-85" aria-hidden="true" tabindex="-1"></a>        tmp <span class="op">=</span> cb.X.copy()</span>
<span id="cb5-86"><a href="#cb5-86" aria-hidden="true" tabindex="-1"></a>        tmp[:,:] <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb5-87"><a href="#cb5-87" aria-hidden="true" tabindex="-1"></a>        cb.layers[<span class="st">"SX08"</span>] <span class="op">=</span> tmp.copy()        </span>
<span id="cb5-88"><a href="#cb5-88" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb5-89"><a href="#cb5-89" aria-hidden="true" tabindex="-1"></a>    dx_file <span class="op">=</span> os.path.join(cbdir,fname,<span class="st">"decontX"</span>,<span class="st">"decontX_cormat.h5"</span>)</span>
<span id="cb5-90"><a href="#cb5-90" aria-hidden="true" tabindex="-1"></a>    tmp <span class="op">=</span> sc.read_10x_h5(dx_file)</span>
<span id="cb5-91"><a href="#cb5-91" aria-hidden="true" tabindex="-1"></a>    cb.layers[<span class="st">"DX"</span>] <span class="op">=</span> np.<span class="bu">round</span>(tmp.X.copy())        </span>
<span id="cb5-92"><a href="#cb5-92" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-93"><a href="#cb5-93" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> l <span class="kw">in</span> cb.layers.keys():</span>
<span id="cb5-94"><a href="#cb5-94" aria-hidden="true" tabindex="-1"></a>        cb.X <span class="op">=</span> cb.layers[l].copy()</span>
<span id="cb5-95"><a href="#cb5-95" aria-hidden="true" tabindex="-1"></a>        qc <span class="op">=</span> sc.pp.calculate_qc_metrics(cb, inplace<span class="op">=</span><span class="va">False</span>, log1p<span class="op">=</span><span class="va">False</span>, qc_vars<span class="op">=</span>qc_vars) <span class="co"># first is obs, second is var.</span></span>
<span id="cb5-96"><a href="#cb5-96" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> l <span class="op">!=</span> <span class="st">"raw"</span>:</span>
<span id="cb5-97"><a href="#cb5-97" aria-hidden="true" tabindex="-1"></a>            <span class="co"># proportions relative to raw data.</span></span>
<span id="cb5-98"><a href="#cb5-98" aria-hidden="true" tabindex="-1"></a>            qc[<span class="dv">0</span>][<span class="st">"nFilt"</span>] <span class="op">=</span> cb.obs[<span class="st">"raw_total_counts"</span>] <span class="op">-</span> qc[<span class="dv">0</span>][<span class="st">"total_counts"</span>]</span>
<span id="cb5-99"><a href="#cb5-99" aria-hidden="true" tabindex="-1"></a>            qc[<span class="dv">0</span>][<span class="st">"propFilt"</span>] <span class="op">=</span> qc[<span class="dv">0</span>][<span class="st">"nFilt"</span>] <span class="op">/</span> cb.obs[<span class="st">"raw_total_counts"</span>]</span>
<span id="cb5-100"><a href="#cb5-100" aria-hidden="true" tabindex="-1"></a>            qc[<span class="dv">1</span>][<span class="st">"nFilt"</span>] <span class="op">=</span> cb.var[<span class="st">"raw_total_counts"</span>] <span class="op">-</span> qc[<span class="dv">1</span>][<span class="st">"total_counts"</span>]</span>
<span id="cb5-101"><a href="#cb5-101" aria-hidden="true" tabindex="-1"></a>            qc[<span class="dv">1</span>][<span class="st">"propFilt"</span>] <span class="op">=</span> qc[<span class="dv">1</span>][<span class="st">"nFilt"</span>] <span class="op">/</span> cb.var[<span class="st">"raw_total_counts"</span>]</span>
<span id="cb5-102"><a href="#cb5-102" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-103"><a href="#cb5-103" aria-hidden="true" tabindex="-1"></a>        qc[<span class="dv">0</span>].columns <span class="op">=</span> l <span class="op">+</span> <span class="st">"_"</span> <span class="op">+</span> qc[<span class="dv">0</span>].columns</span>
<span id="cb5-104"><a href="#cb5-104" aria-hidden="true" tabindex="-1"></a>        cb.obs <span class="op">=</span> pd.concat([cb.obs,qc[<span class="dv">0</span>]], axis<span class="op">=</span><span class="dv">1</span>, join<span class="op">=</span><span class="st">'inner'</span>)</span>
<span id="cb5-105"><a href="#cb5-105" aria-hidden="true" tabindex="-1"></a>        qc[<span class="dv">1</span>].columns <span class="op">=</span> l <span class="op">+</span> <span class="st">"_"</span> <span class="op">+</span> qc[<span class="dv">1</span>].columns</span>
<span id="cb5-106"><a href="#cb5-106" aria-hidden="true" tabindex="-1"></a>        cb.var <span class="op">=</span> pd.concat([cb.var,qc[<span class="dv">1</span>]], axis<span class="op">=</span><span class="dv">1</span>, join<span class="op">=</span><span class="st">'inner'</span>)</span>
<span id="cb5-107"><a href="#cb5-107" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-108"><a href="#cb5-108" aria-hidden="true" tabindex="-1"></a>    <span class="co"># read emptyDrops predictions</span></span>
<span id="cb5-109"><a href="#cb5-109" aria-hidden="true" tabindex="-1"></a>    ed_dir <span class="op">=</span> os.path.join(pdir, sample, <span class="st">"emptydrops"</span>)</span>
<span id="cb5-110"><a href="#cb5-110" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="kw">not</span> os.path.exists(os.path.join(ed_dir, <span class="st">"cell_stats.csv"</span>))):</span>
<span id="cb5-111"><a href="#cb5-111" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="st">"no emptydrops in "</span> <span class="op">+</span> ed_dir)        </span>
<span id="cb5-112"><a href="#cb5-112" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:  </span>
<span id="cb5-113"><a href="#cb5-113" aria-hidden="true" tabindex="-1"></a>        ed_cells <span class="op">=</span> pd.read_csv(os.path.join(ed_dir, <span class="st">"cell_stats.csv"</span>),index_col <span class="op">=</span> <span class="dv">0</span>)</span>
<span id="cb5-114"><a href="#cb5-114" aria-hidden="true" tabindex="-1"></a>        ed_cells <span class="op">=</span> ed_cells.loc[ed_cells[<span class="st">'Total'</span>]<span class="op">&gt;</span><span class="dv">0</span>,:]</span>
<span id="cb5-115"><a href="#cb5-115" aria-hidden="true" tabindex="-1"></a>        cb.uns[<span class="st">'raw_obs'</span>][<span class="st">"isCellED"</span>] <span class="op">=</span> cb.uns[<span class="st">'raw_obs'</span>].index.isin(ed_cells.index[ed_cells.FDR <span class="op">&lt;=</span> <span class="fl">0.05</span>])</span>
<span id="cb5-116"><a href="#cb5-116" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-117"><a href="#cb5-117" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-118"><a href="#cb5-118" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-119"><a href="#cb5-119" aria-hidden="true" tabindex="-1"></a>        emptydrops_genes[fname] <span class="op">=</span> pd.read_csv(os.path.join(ed_dir, <span class="st">"gene_stats.csv"</span>),index_col <span class="op">=</span> <span class="dv">0</span>)</span>
<span id="cb5-120"><a href="#cb5-120" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-121"><a href="#cb5-121" aria-hidden="true" tabindex="-1"></a>    <span class="co"># read cellranger filtered file, if available.</span></span>
<span id="cb5-122"><a href="#cb5-122" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="st">'filt_file'</span> <span class="kw">in</span> settings[<span class="st">'samples'</span>][sample].keys() <span class="kw">and</span> settings[<span class="st">'samples'</span>][sample][<span class="st">'filt_file'</span>] <span class="op">!=</span> <span class="va">None</span> :</span>
<span id="cb5-123"><a href="#cb5-123" aria-hidden="true" tabindex="-1"></a>        ffile <span class="op">=</span> settings[<span class="st">'samples'</span>][sample][<span class="st">'filt_file'</span>]</span>
<span id="cb5-124"><a href="#cb5-124" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> os.path.exists(ffile):</span>
<span id="cb5-125"><a href="#cb5-125" aria-hidden="true" tabindex="-1"></a>            tmp <span class="op">=</span> sc.read_10x_h5(ffile)</span>
<span id="cb5-126"><a href="#cb5-126" aria-hidden="true" tabindex="-1"></a>            cb.uns[<span class="st">'raw_obs'</span>][<span class="st">"isCellCR"</span>] <span class="op">=</span> cb.uns[<span class="st">'raw_obs'</span>].index.isin(tmp.obs.index)</span>
<span id="cb5-127"><a href="#cb5-127" aria-hidden="true" tabindex="-1"></a>        <span class="cf">else</span>:</span>
<span id="cb5-128"><a href="#cb5-128" aria-hidden="true" tabindex="-1"></a>            cb.uns[<span class="st">'raw_obs'</span>][<span class="st">"isCellCR"</span>] <span class="op">=</span> <span class="va">False</span></span>
<span id="cb5-129"><a href="#cb5-129" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb5-130"><a href="#cb5-130" aria-hidden="true" tabindex="-1"></a>        cb.uns[<span class="st">'raw_obs'</span>][<span class="st">"isCellCR"</span>] <span class="op">=</span> <span class="va">False</span></span>
<span id="cb5-131"><a href="#cb5-131" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-132"><a href="#cb5-132" aria-hidden="true" tabindex="-1"></a>    alldata[fname]<span class="op">=</span>cb</span>
<span id="cb5-133"><a href="#cb5-133" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">"Done "</span> <span class="op">+</span> fname)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>parse1_all
no file /Users/asabjor/projects/sc-devop/scQC/data/human/output/parse1_all/soupX/soupX_cormat_0.2.h5
no file /Users/asabjor/projects/sc-devop/scQC/data/human/output/parse1_all/soupX/soupX_cormat_0.8.h5
Done parse1_all
parse1_5-Azacytidine
no file /Users/asabjor/projects/sc-devop/scQC/data/human/output/parse1_5-Azacytidine/soupX/soupX_cormat_0.2.h5
Done parse1_5-Azacytidine
parse1_Fludarabine
Done parse1_Fludarabine
parse1_XRP44X
no file /Users/asabjor/projects/sc-devop/scQC/data/human/output/parse1_XRP44X/soupX/soupX_cormat_0.8.h5
Done parse1_XRP44X
parse1_Control
no file /Users/asabjor/projects/sc-devop/scQC/data/human/output/parse1_Control/soupX/soupX_cormat_0.2.h5
no file /Users/asabjor/projects/sc-devop/scQC/data/human/output/parse1_Control/soupX/soupX_cormat_0.8.h5
Done parse1_Control
parse1_Dexamethasone
Done parse1_Dexamethasone
parse1_Imatinib
no file /Users/asabjor/projects/sc-devop/scQC/data/human/output/parse1_Imatinib/soupX/soupX_cormat_0.2.h5
no file /Users/asabjor/projects/sc-devop/scQC/data/human/output/parse1_Imatinib/soupX/soupX_cormat_0.8.h5
Done parse1_Imatinib
parse2_all
Done parse2_all
parse2_5-Azacytidine
no file /Users/asabjor/projects/sc-devop/scQC/data/human/output/parse2_5-Azacytidine/soupX/soupX_cormat_0.2.h5
Done parse2_5-Azacytidine
parse2_Fludarabine
Done parse2_Fludarabine
parse2_XRP44X
Done parse2_XRP44X
parse2_Control
Done parse2_Control
parse2_Dexamethasone
Done parse2_Dexamethasone
parse2_Imatinib
no file /Users/asabjor/projects/sc-devop/scQC/data/human/output/parse2_Imatinib/soupX/soupX_cormat_0.8.h5
Done parse2_Imatinib</code></pre>
</div>
</div>
</section>
<section id="test-tabset" class="level3">
<h3 class="anchored" data-anchor-id="test-tabset">Test tabset</h3>
<div class="tabset-margin-container"></div><div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-1-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-1" role="tab" aria-controls="tabset-1-1" aria-selected="true" href="">tab1</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-2" role="tab" aria-controls="tabset-1-2" aria-selected="false" href="">tab2</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-3" role="tab" aria-controls="tabset-1-3" aria-selected="false" href="">tab3</a></li></ul>
<div class="tab-content">
<div id="tabset-1-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-1-1-tab">
<p>Some text</p>
</div>
<div id="tabset-1-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-2-tab">
<p>other</p>
</div>
<div id="tabset-1-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-3-tab">

</div>
</div>
</div>
</section>
<section id="cell-filtering-cellbender" class="level3">
<h3 class="anchored">Cell filtering cellbender</h3>
<p>Plot barcode rank plot colored by cellbender/cellranger filtering. Calculate rank plot based on cellranger total counts.</p>
<div id="plot-func" class="cell" data-execution_count="6">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> matplotlib_venn <span class="im">import</span> venn3</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.patches <span class="im">as</span> mpatches</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>clegend <span class="op">=</span> [mpatches.Patch(color<span class="op">=</span><span class="st">'Purple'</span>, label<span class="op">=</span><span class="st">'Union 3'</span>),</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>        mpatches.Patch(color<span class="op">=</span><span class="st">'Blue'</span>, label<span class="op">=</span><span class="st">'EmptyDrops'</span>),</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>        mpatches.Patch(color<span class="op">=</span><span class="st">'Green'</span>, label<span class="op">=</span><span class="st">'CellRanger'</span>),</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>        mpatches.Patch(color<span class="op">=</span><span class="st">'Red'</span>, label<span class="op">=</span><span class="st">'Cellbender'</span>),</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>        mpatches.Patch(color<span class="op">=</span><span class="st">'Yellow'</span>, label<span class="op">=</span><span class="st">'2 methods'</span>),</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>        mpatches.Patch(color<span class="op">=</span><span class="st">'Grey'</span>, label<span class="op">=</span><span class="st">'None'</span>)</span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>        ]</span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> plot_general_stats(tmp,name):</span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a>    display(Markdown(<span class="ss">f"## </span><span class="sc">{</span>name<span class="sc">}</span><span class="ss">"</span>))</span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a>    tmp.uns[<span class="st">"raw_obs"</span>][<span class="st">"inCB"</span>] <span class="op">=</span> tmp.uns[<span class="st">"raw_obs"</span>].index.isin(tmp.obs_names)</span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a>    tmp.uns[<span class="st">"raw_obs"</span>][<span class="st">"rankRaw"</span>] <span class="op">=</span> tmp.uns[<span class="st">"raw_obs"</span>][<span class="st">'total_counts'</span>].rank(method<span class="op">=</span><span class="st">'first'</span>, ascending <span class="op">=</span> <span class="va">False</span>)</span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true" tabindex="-1"></a>    fig, axs <span class="op">=</span> plt.subplots(<span class="dv">2</span>, <span class="dv">2</span>, figsize<span class="op">=</span>(<span class="dv">10</span>,<span class="dv">8</span>),constrained_layout<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb7-20"><a href="#cb7-20" aria-hidden="true" tabindex="-1"></a>    s1 <span class="op">=</span> <span class="bu">set</span>(tmp.uns[<span class="st">'raw_obs'</span>].index[tmp.uns[<span class="st">'raw_obs'</span>][<span class="st">'inCB'</span>]])</span>
<span id="cb7-21"><a href="#cb7-21" aria-hidden="true" tabindex="-1"></a>    s2 <span class="op">=</span> <span class="bu">set</span>(tmp.uns[<span class="st">'raw_obs'</span>].index[tmp.uns[<span class="st">'raw_obs'</span>][<span class="st">'isCellCR'</span>]])</span>
<span id="cb7-22"><a href="#cb7-22" aria-hidden="true" tabindex="-1"></a>    s3 <span class="op">=</span> <span class="bu">set</span>(tmp.uns[<span class="st">'raw_obs'</span>].index[tmp.uns[<span class="st">'raw_obs'</span>][<span class="st">'isCellED'</span>]])</span>
<span id="cb7-23"><a href="#cb7-23" aria-hidden="true" tabindex="-1"></a>    venn3((s1,s2,s3), set_labels <span class="op">=</span> (<span class="st">'CB'</span>, <span class="st">'CR'</span>, <span class="st">'ED'</span>), ax<span class="op">=</span>axs[<span class="dv">1</span>,<span class="dv">0</span>])</span>
<span id="cb7-24"><a href="#cb7-24" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">1</span>,<span class="dv">0</span>].title.set_text(<span class="st">"Cells selection"</span>)</span>
<span id="cb7-25"><a href="#cb7-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-26"><a href="#cb7-26" aria-hidden="true" tabindex="-1"></a>    cc <span class="op">=</span> tmp.uns[<span class="st">'raw_obs'</span>].loc[:,[<span class="st">"isCellED"</span>,<span class="st">"isCellCR"</span>,<span class="st">"inCB"</span>]]</span>
<span id="cb7-27"><a href="#cb7-27" aria-hidden="true" tabindex="-1"></a>    tmp.uns[<span class="st">'raw_obs'</span>][<span class="st">'color'</span>] <span class="op">=</span> <span class="st">"Grey"</span></span>
<span id="cb7-28"><a href="#cb7-28" aria-hidden="true" tabindex="-1"></a>    tmp.uns[<span class="st">'raw_obs'</span>][<span class="st">'color'</span>][tmp.uns[<span class="st">'raw_obs'</span>][<span class="st">'inCB'</span>]] <span class="op">=</span> <span class="st">"Red"</span> </span>
<span id="cb7-29"><a href="#cb7-29" aria-hidden="true" tabindex="-1"></a>    tmp.uns[<span class="st">'raw_obs'</span>][<span class="st">'color'</span>][tmp.uns[<span class="st">'raw_obs'</span>][<span class="st">'isCellCR'</span>]] <span class="op">=</span> <span class="st">"Green"</span> </span>
<span id="cb7-30"><a href="#cb7-30" aria-hidden="true" tabindex="-1"></a>    tmp.uns[<span class="st">'raw_obs'</span>][<span class="st">'color'</span>][tmp.uns[<span class="st">'raw_obs'</span>][<span class="st">'isCellED'</span>]] <span class="op">=</span> <span class="st">"Blue"</span> </span>
<span id="cb7-31"><a href="#cb7-31" aria-hidden="true" tabindex="-1"></a>    tmp.uns[<span class="st">'raw_obs'</span>][<span class="st">'color'</span>][cc.<span class="bu">sum</span>(axis<span class="op">=</span><span class="dv">1</span>) <span class="op">==</span> <span class="dv">3</span>] <span class="op">=</span> <span class="st">"Purple"</span></span>
<span id="cb7-32"><a href="#cb7-32" aria-hidden="true" tabindex="-1"></a>    tmp.uns[<span class="st">'raw_obs'</span>][<span class="st">'color'</span>][cc.<span class="bu">sum</span>(axis<span class="op">=</span><span class="dv">1</span>) <span class="op">==</span> <span class="dv">2</span>] <span class="op">=</span> <span class="st">"Yellow"</span></span>
<span id="cb7-33"><a href="#cb7-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-34"><a href="#cb7-34" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">0</span>,<span class="dv">0</span>].scatter(tmp.uns[<span class="st">"raw_obs"</span>][<span class="st">'rankRaw'</span>], tmp.uns[<span class="st">"raw_obs"</span>][<span class="st">'total_counts'</span>], c<span class="op">=</span>tmp.uns[<span class="st">"raw_obs"</span>][<span class="st">"color"</span>])</span>
<span id="cb7-35"><a href="#cb7-35" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">0</span>,<span class="dv">0</span>].set_xscale(<span class="st">"log"</span>)</span>
<span id="cb7-36"><a href="#cb7-36" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">0</span>,<span class="dv">0</span>].set_yscale(<span class="st">"log"</span>)</span>
<span id="cb7-37"><a href="#cb7-37" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">0</span>,<span class="dv">0</span>].set_xlabel(<span class="st">"rankRaw"</span>)</span>
<span id="cb7-38"><a href="#cb7-38" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">0</span>,<span class="dv">0</span>].set_ylabel(<span class="st">"nUMI"</span>)</span>
<span id="cb7-39"><a href="#cb7-39" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">0</span>,<span class="dv">0</span>].title.set_text(<span class="st">"Barcode rank plot"</span>)</span>
<span id="cb7-40"><a href="#cb7-40" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">0</span>,<span class="dv">0</span>].legend(handles<span class="op">=</span>clegend,</span>
<span id="cb7-41"><a href="#cb7-41" aria-hidden="true" tabindex="-1"></a>          loc<span class="op">=</span><span class="st">'lower left'</span>)</span>
<span id="cb7-42"><a href="#cb7-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-43"><a href="#cb7-43" aria-hidden="true" tabindex="-1"></a>    <span class="co"># counts before after filt</span></span>
<span id="cb7-44"><a href="#cb7-44" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">0</span>,<span class="dv">1</span>].scatter(tmp.obs[<span class="st">'raw_total_counts'</span>], tmp.obs[<span class="st">'CB_total_counts'</span>], alpha<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="cb7-45"><a href="#cb7-45" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">0</span>,<span class="dv">1</span>].axline((<span class="dv">0</span>, <span class="dv">0</span>), slope<span class="op">=</span><span class="dv">1</span>, color<span class="op">=</span><span class="st">"black"</span>)</span>
<span id="cb7-46"><a href="#cb7-46" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">0</span>,<span class="dv">1</span>].set_xlabel(<span class="st">"Raw counts"</span>)</span>
<span id="cb7-47"><a href="#cb7-47" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">0</span>,<span class="dv">1</span>].set_ylabel(<span class="st">"CB counts"</span>)</span>
<span id="cb7-48"><a href="#cb7-48" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">0</span>,<span class="dv">1</span>].set_xscale(<span class="st">"log"</span>)</span>
<span id="cb7-49"><a href="#cb7-49" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">0</span>,<span class="dv">1</span>].set_yscale(<span class="st">"log"</span>)</span>
<span id="cb7-50"><a href="#cb7-50" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">0</span>,<span class="dv">1</span>].title.set_text(<span class="st">"Counts per cell before/after CB"</span>)</span>
<span id="cb7-51"><a href="#cb7-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-52"><a href="#cb7-52" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">1</span>,<span class="dv">1</span>].scatter(tmp.var[<span class="st">"raw_total_counts"</span>], tmp.var[<span class="st">'CB_total_counts'</span>], alpha<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="cb7-53"><a href="#cb7-53" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">1</span>,<span class="dv">1</span>].axline((<span class="dv">0</span>, <span class="dv">0</span>), slope<span class="op">=</span><span class="dv">1</span>, color<span class="op">=</span><span class="st">"black"</span>)</span>
<span id="cb7-54"><a href="#cb7-54" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">1</span>,<span class="dv">1</span>].set_xlabel(<span class="st">"Raw gene counts"</span>)</span>
<span id="cb7-55"><a href="#cb7-55" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">1</span>,<span class="dv">1</span>].set_ylabel(<span class="st">"CB gene counts"</span>)</span>
<span id="cb7-56"><a href="#cb7-56" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">1</span>,<span class="dv">1</span>].set_xscale(<span class="st">"log"</span>)</span>
<span id="cb7-57"><a href="#cb7-57" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">1</span>,<span class="dv">1</span>].set_yscale(<span class="st">"log"</span>)</span>
<span id="cb7-58"><a href="#cb7-58" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">1</span>,<span class="dv">1</span>].title.set_text(<span class="st">"Counts per gene before/after CB"</span>)</span>
<span id="cb7-59"><a href="#cb7-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-60"><a href="#cb7-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-61"><a href="#cb7-61" aria-hidden="true" tabindex="-1"></a>    fig.suptitle(name)</span>
<span id="cb7-62"><a href="#cb7-62" aria-hidden="true" tabindex="-1"></a>    display(plt.gcf())</span>
<span id="cb7-63"><a href="#cb7-63" aria-hidden="true" tabindex="-1"></a>    plt.close()</span>
<span id="cb7-64"><a href="#cb7-64" aria-hidden="true" tabindex="-1"></a>    display(Markdown(<span class="ss">f" "</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="barcode-plot" class="cell" data-execution_count="7">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> sname <span class="kw">in</span> alldata.keys():</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>    plot_general_stats(alldata[sname], sname)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<section id="parse1_all" class="level2">
<h2 class="anchored" data-anchor-id="parse1_all">parse1_all</h2>
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/barcode-plot-output-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-1"><img src="ambient_methods_overview_hsa_parse_files/figure-html/barcode-plot-output-2.png" id="barcode-plot-2" width="809" height="649"></a></p>
</section>
<section id="parse1_5-azacytidine" class="level2">
<h2 class="anchored" data-anchor-id="parse1_5-azacytidine">parse1_5-Azacytidine</h2>
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/barcode-plot-output-5.png" class="lightbox" data-gallery="quarto-lightbox-gallery-2"><img src="ambient_methods_overview_hsa_parse_files/figure-html/barcode-plot-output-5.png" id="barcode-plot-5" width="809" height="649"></a></p>
</section>
<section id="parse1_fludarabine" class="level2">
<h2 class="anchored" data-anchor-id="parse1_fludarabine">parse1_Fludarabine</h2>
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/barcode-plot-output-8.png" class="lightbox" data-gallery="quarto-lightbox-gallery-3"><img src="ambient_methods_overview_hsa_parse_files/figure-html/barcode-plot-output-8.png" id="barcode-plot-8" width="809" height="649"></a></p>
</section>
<section id="parse1_xrp44x" class="level2">
<h2 class="anchored" data-anchor-id="parse1_xrp44x">parse1_XRP44X</h2>
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/barcode-plot-output-11.png" class="lightbox" data-gallery="quarto-lightbox-gallery-4"><img src="ambient_methods_overview_hsa_parse_files/figure-html/barcode-plot-output-11.png" id="barcode-plot-11" width="809" height="649"></a></p>
</section>
<section id="parse1_control" class="level2">
<h2 class="anchored" data-anchor-id="parse1_control">parse1_Control</h2>
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/barcode-plot-output-14.png" class="lightbox" data-gallery="quarto-lightbox-gallery-5"><img src="ambient_methods_overview_hsa_parse_files/figure-html/barcode-plot-output-14.png" id="barcode-plot-14" width="809" height="649"></a></p>
</section>
<section id="parse1_dexamethasone" class="level2">
<h2 class="anchored" data-anchor-id="parse1_dexamethasone">parse1_Dexamethasone</h2>
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/barcode-plot-output-17.png" class="lightbox" data-gallery="quarto-lightbox-gallery-6"><img src="ambient_methods_overview_hsa_parse_files/figure-html/barcode-plot-output-17.png" id="barcode-plot-17" width="809" height="649"></a></p>
</section>
<section id="parse1_imatinib" class="level2">
<h2 class="anchored" data-anchor-id="parse1_imatinib">parse1_Imatinib</h2>
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/barcode-plot-output-20.png" class="lightbox" data-gallery="quarto-lightbox-gallery-7"><img src="ambient_methods_overview_hsa_parse_files/figure-html/barcode-plot-output-20.png" id="barcode-plot-20" width="809" height="649"></a></p>
</section>
<section id="parse2_all" class="level2">
<h2 class="anchored" data-anchor-id="parse2_all">parse2_all</h2>
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/barcode-plot-output-23.png" class="lightbox" data-gallery="quarto-lightbox-gallery-8"><img src="ambient_methods_overview_hsa_parse_files/figure-html/barcode-plot-output-23.png" id="barcode-plot-23" width="809" height="649"></a></p>
</section>
<section id="parse2_5-azacytidine" class="level2">
<h2 class="anchored" data-anchor-id="parse2_5-azacytidine">parse2_5-Azacytidine</h2>
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/barcode-plot-output-26.png" class="lightbox" data-gallery="quarto-lightbox-gallery-9"><img src="ambient_methods_overview_hsa_parse_files/figure-html/barcode-plot-output-26.png" id="barcode-plot-26" width="809" height="649"></a></p>
</section>
<section id="parse2_fludarabine" class="level2">
<h2 class="anchored" data-anchor-id="parse2_fludarabine">parse2_Fludarabine</h2>
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/barcode-plot-output-29.png" class="lightbox" data-gallery="quarto-lightbox-gallery-10"><img src="ambient_methods_overview_hsa_parse_files/figure-html/barcode-plot-output-29.png" id="barcode-plot-29" width="809" height="649"></a></p>
</section>
<section id="parse2_xrp44x" class="level2">
<h2 class="anchored" data-anchor-id="parse2_xrp44x">parse2_XRP44X</h2>
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/barcode-plot-output-32.png" class="lightbox" data-gallery="quarto-lightbox-gallery-11"><img src="ambient_methods_overview_hsa_parse_files/figure-html/barcode-plot-output-32.png" id="barcode-plot-32" width="809" height="649"></a></p>
</section>
<section id="parse2_control" class="level2">
<h2 class="anchored" data-anchor-id="parse2_control">parse2_Control</h2>
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/barcode-plot-output-35.png" class="lightbox" data-gallery="quarto-lightbox-gallery-12"><img src="ambient_methods_overview_hsa_parse_files/figure-html/barcode-plot-output-35.png" id="barcode-plot-35" width="809" height="649"></a></p>
</section>
<section id="parse2_dexamethasone" class="level2">
<h2 class="anchored" data-anchor-id="parse2_dexamethasone">parse2_Dexamethasone</h2>
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/barcode-plot-output-38.png" class="lightbox" data-gallery="quarto-lightbox-gallery-13"><img src="ambient_methods_overview_hsa_parse_files/figure-html/barcode-plot-output-38.png" id="barcode-plot-38" width="809" height="649"></a></p>
</section>
<section id="parse2_imatinib" class="level2">
<h2 class="anchored" data-anchor-id="parse2_imatinib">parse2_Imatinib</h2>
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/barcode-plot-output-41.png" class="lightbox" data-gallery="quarto-lightbox-gallery-14"><img src="ambient_methods_overview_hsa_parse_files/figure-html/barcode-plot-output-41.png" id="barcode-plot-41" width="809" height="649"></a></p>
</section>
</div>
<p>Weird histograms for some samples, especially the SAM samples.</p>
</section>
<section id="filtering-stats-all-methods" class="level3">
<h3 class="anchored" data-anchor-id="filtering-stats-all-methods">Filtering stats all methods</h3>
<p>Proportion filtered per cell</p>
<div id="filt-stat" class="cell" data-collapse="true" data-execution_count="8">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>all_methods <span class="op">=</span> [<span class="st">"CB"</span>,<span class="st">"SX02"</span>,<span class="st">"SX08"</span>,<span class="st">"DX"</span>]</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>pf <span class="op">=</span> [x <span class="op">+</span> <span class="st">"_propFilt"</span> <span class="cf">for</span> x <span class="kw">in</span> all_methods]</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>nPlot <span class="op">=</span> <span class="dv">9</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> r <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">0</span>,<span class="bu">len</span>(alldata),nPlot): </span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>    s <span class="op">=</span> <span class="bu">list</span>(alldata.keys())[r:(r<span class="op">+</span>nPlot)]</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>    fig, axs <span class="op">=</span> plt.subplots(<span class="dv">3</span>, <span class="dv">3</span>, figsize<span class="op">=</span>(<span class="dv">10</span>,<span class="dv">8</span>),constrained_layout<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> sname, ax <span class="kw">in</span> <span class="bu">zip</span>(s, axs.flatten()):</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>        sc.pl.violin(alldata[sname], pf, rotation<span class="op">=</span><span class="dv">90</span>, ax<span class="op">=</span>ax, show<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>        ax.tick_params(axis<span class="op">=</span><span class="st">'x'</span>, labelsize<span class="op">=</span><span class="dv">8</span>)</span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>        ax.title.set_text(sname)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/filt-stat-output-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-15"><img src="ambient_methods_overview_hsa_parse_files/figure-html/filt-stat-output-1.png" id="filt-stat-1" width="809" height="649" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/filt-stat-output-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-16"><img src="ambient_methods_overview_hsa_parse_files/figure-html/filt-stat-output-2.png" id="filt-stat-2" width="809" height="649" class="figure-img"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="gene-stats" class="level3">
<h3 class="anchored" data-anchor-id="gene-stats">Gene stats</h3>
<p>Plot fraction filtered per gene vs total counts before filtering. Color genes by:</p>
<ul>
<li>green = mt</li>
<li>red = pc</li>
<li>blue = lincRNA</li>
<li>purple = other (all other classes, plus ones not in latest biomart ensembl version)</li>
</ul>
<div id="plot-func2" class="cell" data-execution_count="9">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>colors<span class="op">=</span>sns.color_palette(<span class="st">'Set1'</span>)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>gene_colors<span class="op">=</span>{<span class="st">'pc'</span>: colors[<span class="dv">0</span>], <span class="st">'lncRNA'</span>: colors[<span class="dv">1</span>], <span class="st">'mt'</span>:colors[<span class="dv">2</span>], <span class="st">'other'</span>:colors[<span class="dv">3</span>]} </span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>hexcols <span class="op">=</span> colors.as_hex()</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>gene_colors_hex <span class="op">=</span> {<span class="st">'pc'</span>: hexcols[<span class="dv">0</span>], <span class="st">'lncRNA'</span>: hexcols[<span class="dv">1</span>], <span class="st">'mt'</span>:hexcols[<span class="dv">2</span>], <span class="st">'other'</span>:hexcols[<span class="dv">3</span>]} </span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> plot_gene_proportions(df,sname<span class="op">=</span><span class="st">" "</span>, nPlot<span class="op">=</span><span class="dv">10</span>):</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>    df <span class="op">=</span> df.loc[df.raw_total_counts <span class="op">&gt;</span> <span class="dv">0</span>,:]</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>    df[<span class="st">'Clog'</span>] <span class="op">=</span> np.log1p(df[<span class="st">'raw_total_counts'</span>])</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>    fig, axs <span class="op">=</span> plt.subplots(<span class="dv">1</span>, <span class="dv">5</span>, figsize<span class="op">=</span>(<span class="dv">10</span>,<span class="dv">4</span>),constrained_layout<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i,n <span class="kw">in</span> <span class="bu">enumerate</span>(all_methods):</span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> <span class="kw">not</span> df.columns.isin( [n<span class="op">+</span><span class="st">'_propFilt'</span>]).<span class="bu">any</span>(): <span class="cf">continue</span></span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a>        df[n<span class="op">+</span><span class="st">"_rank"</span>]<span class="op">=</span> df[n<span class="op">+</span><span class="st">'_propFilt'</span>] <span class="op">*</span> df[<span class="st">'Clog'</span>]</span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a>        df <span class="op">=</span> df.sort_values(by<span class="op">=</span>n<span class="op">+</span><span class="st">"_rank"</span>, axis<span class="op">=</span><span class="dv">0</span>, ascending<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a>        sel <span class="op">=</span> df.head(nPlot)</span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a>        sns.scatterplot(df, x<span class="op">=</span><span class="st">'Clog'</span>,y<span class="op">=</span>n<span class="op">+</span><span class="st">'_propFilt'</span>,hue<span class="op">=</span><span class="st">'gene_class'</span>, palette<span class="op">=</span>gene_colors,  legend<span class="op">=</span><span class="va">False</span>, ax<span class="op">=</span>axs[i])</span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true" tabindex="-1"></a>        axs[i].title.set_text(n)</span>
<span id="cb10-18"><a href="#cb10-18" aria-hidden="true" tabindex="-1"></a>        axs[i].set_xlabel(<span class="st">''</span>)</span>
<span id="cb10-19"><a href="#cb10-19" aria-hidden="true" tabindex="-1"></a>        axs[i].set_ylabel(<span class="st">''</span>)</span>
<span id="cb10-20"><a href="#cb10-20" aria-hidden="true" tabindex="-1"></a>        <span class="co"># annotate points in axis</span></span>
<span id="cb10-21"><a href="#cb10-21" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> idx, row <span class="kw">in</span> sel.iterrows():</span>
<span id="cb10-22"><a href="#cb10-22" aria-hidden="true" tabindex="-1"></a>            axs[i].annotate(idx, ( row[<span class="st">'Clog'</span>],row[n<span class="op">+</span><span class="st">'_propFilt'</span>]), rotation <span class="op">=</span> <span class="dv">45</span>, fontsize<span class="op">=</span><span class="dv">6</span>)</span>
<span id="cb10-23"><a href="#cb10-23" aria-hidden="true" tabindex="-1"></a>        fig.suptitle(sname)</span>
<span id="cb10-24"><a href="#cb10-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-25"><a href="#cb10-25" aria-hidden="true" tabindex="-1"></a>    gstats <span class="op">=</span> emptydrops_genes[sname]</span>
<span id="cb10-26"><a href="#cb10-26" aria-hidden="true" tabindex="-1"></a>    gstats <span class="op">=</span> gstats.loc[gstats.index.isin(df.index),:]</span>
<span id="cb10-27"><a href="#cb10-27" aria-hidden="true" tabindex="-1"></a>    topA <span class="op">=</span> gstats[<span class="st">'ambient'</span>].sort_values(ascending<span class="op">=</span><span class="va">False</span>).head(nPlot).index</span>
<span id="cb10-28"><a href="#cb10-28" aria-hidden="true" tabindex="-1"></a>    gstats <span class="op">=</span> gstats.loc[topA,:]</span>
<span id="cb10-29"><a href="#cb10-29" aria-hidden="true" tabindex="-1"></a>    gstats[<span class="st">'gene_class'</span>] <span class="op">=</span> df[<span class="st">'gene_class'</span>][gstats.index]</span>
<span id="cb10-30"><a href="#cb10-30" aria-hidden="true" tabindex="-1"></a>    gstats[<span class="st">'gene'</span>] <span class="op">=</span> gstats.index    </span>
<span id="cb10-31"><a href="#cb10-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-32"><a href="#cb10-32" aria-hidden="true" tabindex="-1"></a>    sns.barplot(gstats, x<span class="op">=</span><span class="st">'gene'</span>,y<span class="op">=</span><span class="st">'ambient'</span>,hue<span class="op">=</span><span class="st">'gene_class'</span>, palette<span class="op">=</span>gene_colors, ax<span class="op">=</span>axs[<span class="dv">4</span>])</span>
<span id="cb10-33"><a href="#cb10-33" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">4</span>].title.set_text(<span class="st">"EmptyDrops score"</span>)</span>
<span id="cb10-34"><a href="#cb10-34" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">4</span>].set_xlabel(<span class="st">''</span>)</span>
<span id="cb10-35"><a href="#cb10-35" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">4</span>].set_ylabel(<span class="st">''</span>)</span>
<span id="cb10-36"><a href="#cb10-36" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">4</span>].legend(fontsize<span class="op">=</span><span class="dv">8</span>)</span>
<span id="cb10-37"><a href="#cb10-37" aria-hidden="true" tabindex="-1"></a>    rotate_xticks(axs[<span class="dv">4</span>], size<span class="op">=</span><span class="dv">6</span>)    </span>
<span id="cb10-38"><a href="#cb10-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-39"><a href="#cb10-39" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> rotate_xticks(ax, size<span class="op">=</span><span class="dv">8</span>, rotation<span class="op">=</span><span class="dv">45</span>):</span>
<span id="cb10-40"><a href="#cb10-40" aria-hidden="true" tabindex="-1"></a>    ax.set_xticklabels(ax.get_xticklabels(), rotation <span class="op">=</span> rotation, size<span class="op">=</span>size, ha<span class="op">=</span><span class="st">'right'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="plot-genes" class="cell" data-collapse="true" data-execution_count="10">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> sname <span class="kw">in</span> alldata.keys():</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>    plot_gene_proportions(alldata[sname].var, sname)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/plot-genes-output-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-17"><img src="ambient_methods_overview_hsa_parse_files/figure-html/plot-genes-output-1.png" id="plot-genes-1" width="820" height="329" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/plot-genes-output-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-18"><img src="ambient_methods_overview_hsa_parse_files/figure-html/plot-genes-output-2.png" id="plot-genes-2" width="821" height="329" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/plot-genes-output-3.png" class="lightbox" data-gallery="quarto-lightbox-gallery-19"><img src="ambient_methods_overview_hsa_parse_files/figure-html/plot-genes-output-3.png" id="plot-genes-3" width="816" height="329" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/plot-genes-output-4.png" class="lightbox" data-gallery="quarto-lightbox-gallery-20"><img src="ambient_methods_overview_hsa_parse_files/figure-html/plot-genes-output-4.png" id="plot-genes-4" width="821" height="329" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/plot-genes-output-5.png" class="lightbox" data-gallery="quarto-lightbox-gallery-21"><img src="ambient_methods_overview_hsa_parse_files/figure-html/plot-genes-output-5.png" id="plot-genes-5" width="821" height="329" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/plot-genes-output-6.png" class="lightbox" data-gallery="quarto-lightbox-gallery-22"><img src="ambient_methods_overview_hsa_parse_files/figure-html/plot-genes-output-6.png" id="plot-genes-6" width="818" height="329" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/plot-genes-output-7.png" class="lightbox" data-gallery="quarto-lightbox-gallery-23"><img src="ambient_methods_overview_hsa_parse_files/figure-html/plot-genes-output-7.png" id="plot-genes-7" width="820" height="329" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/plot-genes-output-8.png" class="lightbox" data-gallery="quarto-lightbox-gallery-24"><img src="ambient_methods_overview_hsa_parse_files/figure-html/plot-genes-output-8.png" id="plot-genes-8" width="818" height="329" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/plot-genes-output-9.png" class="lightbox" data-gallery="quarto-lightbox-gallery-25"><img src="ambient_methods_overview_hsa_parse_files/figure-html/plot-genes-output-9.png" id="plot-genes-9" width="818" height="329" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/plot-genes-output-10.png" class="lightbox" data-gallery="quarto-lightbox-gallery-26"><img src="ambient_methods_overview_hsa_parse_files/figure-html/plot-genes-output-10.png" id="plot-genes-10" width="816" height="329" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/plot-genes-output-11.png" class="lightbox" data-gallery="quarto-lightbox-gallery-27"><img src="ambient_methods_overview_hsa_parse_files/figure-html/plot-genes-output-11.png" id="plot-genes-11" width="818" height="329" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/plot-genes-output-12.png" class="lightbox" data-gallery="quarto-lightbox-gallery-28"><img src="ambient_methods_overview_hsa_parse_files/figure-html/plot-genes-output-12.png" id="plot-genes-12" width="818" height="329" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/plot-genes-output-13.png" class="lightbox" data-gallery="quarto-lightbox-gallery-29"><img src="ambient_methods_overview_hsa_parse_files/figure-html/plot-genes-output-13.png" id="plot-genes-13" width="818" height="329" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/plot-genes-output-14.png" class="lightbox" data-gallery="quarto-lightbox-gallery-30"><img src="ambient_methods_overview_hsa_parse_files/figure-html/plot-genes-output-14.png" id="plot-genes-14" width="819" height="329" class="figure-img"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="mito-etc." class="level3">
<h3 class="anchored" data-anchor-id="mito-etc.">Mito, etc.</h3>
<p>Violin plots after filtering and raw.</p>
<div id="plot-vln" class="cell" data-collapse="true" data-execution_count="11">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>qc_plot <span class="op">=</span> [<span class="st">"mt"</span>,<span class="st">"hb"</span>,<span class="st">"pc"</span>]</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>plot_feats <span class="op">=</span> [m <span class="op">+</span> <span class="st">"_pct_counts_"</span> <span class="op">+</span> q <span class="cf">for</span> q <span class="kw">in</span> qc_plot <span class="cf">for</span> m <span class="kw">in</span> [<span class="st">"raw"</span>] <span class="op">+</span> all_methods ]</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>nPlot <span class="op">=</span> <span class="dv">9</span></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> r <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">0</span>,<span class="bu">len</span>(alldata),nPlot): </span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>    s <span class="op">=</span> <span class="bu">list</span>(alldata.keys())[r:(r<span class="op">+</span>nPlot)]</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>    fig, axs <span class="op">=</span> plt.subplots(<span class="dv">3</span>, <span class="dv">3</span>, figsize<span class="op">=</span>(<span class="dv">10</span>,<span class="dv">8</span>),constrained_layout<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> sname, ax <span class="kw">in</span> <span class="bu">zip</span>(s, axs.flatten()):</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>        sc.pl.violin(alldata[sname], plot_feats, rotation<span class="op">=</span><span class="dv">90</span>, ax<span class="op">=</span>ax, show<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>        ax.tick_params(axis<span class="op">=</span><span class="st">'x'</span>, labelsize<span class="op">=</span><span class="dv">8</span>)</span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a>        ax.title.set_text(sname)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/plot-vln-output-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-31"><img src="ambient_methods_overview_hsa_parse_files/figure-html/plot-vln-output-1.png" id="plot-vln-1" width="809" height="649" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/plot-vln-output-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-32"><img src="ambient_methods_overview_hsa_parse_files/figure-html/plot-vln-output-2.png" id="plot-vln-2" width="809" height="649" class="figure-img"></a></p>
</figure>
</div>
</div>
</div>
<p>Generally more high mito cells after CB filtering.</p>
</section>
<section id="biotype-proportions" class="level3">
<h3 class="anchored" data-anchor-id="biotype-proportions">Biotype proportions</h3>
<p>Calculate total proportions of all biotypes per layer.</p>
<p>Total is the proportions in the full raw matrix also containing empty droplets, while raw is the same stats taking the subset of cells selected by cellbender.</p>
<div id="plot-func3" class="cell" data-execution_count="12">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> plot_biotypes(df,title,ax):</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>    tmp <span class="op">=</span> df.loc[:,df.columns.<span class="bu">str</span>.contains(<span class="st">"total_counts|gene_class"</span>)]</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>    tmp <span class="op">=</span> tmp.groupby(<span class="st">"gene_class"</span>).<span class="bu">sum</span>()</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>    tmp.columns <span class="op">=</span> tmp.columns.<span class="bu">str</span>.replace(<span class="st">"_counts"</span>,<span class="st">""</span>)</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>    tmp.columns <span class="op">=</span> tmp.columns.<span class="bu">str</span>.replace(<span class="st">"_total"</span>,<span class="st">""</span>)</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>    tmp_perc <span class="op">=</span> tmp.iloc[:, :].<span class="bu">apply</span>(<span class="kw">lambda</span> x: x <span class="op">/</span> x.<span class="bu">sum</span>(), axis<span class="op">=</span><span class="dv">0</span>) <span class="op">*</span> <span class="dv">100</span></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>    tmp_perc.transpose().plot(kind<span class="op">=</span><span class="st">"bar"</span>,stacked<span class="op">=</span><span class="va">True</span>, color<span class="op">=</span>gene_colors, legend<span class="op">=</span><span class="va">False</span>, title<span class="op">=</span>title, ax<span class="op">=</span>ax )</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>    lim <span class="op">=</span> np.<span class="bu">round</span>(<span class="dv">100</span><span class="op">-</span>tmp_perc.loc[<span class="st">"pc"</span>,:].<span class="bu">min</span>()<span class="op">+</span><span class="dv">5</span>)</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>    ax.set_ylim([<span class="dv">0</span>,lim])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="plot-biotype" class="cell" data-collapse="true" data-execution_count="13">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>qc_plot <span class="op">=</span> [<span class="st">"total_counts"</span>,<span class="st">"n_genes_by_counts"</span>]</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>plot_feats <span class="op">=</span> [m <span class="op">+</span> <span class="st">"_"</span> <span class="op">+</span> q <span class="cf">for</span> q <span class="kw">in</span> qc_plot <span class="cf">for</span> m <span class="kw">in</span> [<span class="st">"raw"</span>] <span class="op">+</span> all_methods ]</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>nPlot <span class="op">=</span> <span class="dv">9</span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> r <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">0</span>,<span class="bu">len</span>(alldata),nPlot): </span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>    s <span class="op">=</span> <span class="bu">list</span>(alldata.keys())[r:(r<span class="op">+</span>nPlot)]</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>    fig, axs <span class="op">=</span> plt.subplots(<span class="dv">3</span>, <span class="dv">3</span>, figsize<span class="op">=</span>(<span class="dv">10</span>,<span class="dv">8</span>),constrained_layout<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> sname, ax <span class="kw">in</span> <span class="bu">zip</span>(s, axs.flatten()):</span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>        plot_biotypes(alldata[sname].var,sname,ax)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/plot-biotype-output-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-33"><img src="ambient_methods_overview_hsa_parse_files/figure-html/plot-biotype-output-1.png" id="plot-biotype-1" width="809" height="649" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/plot-biotype-output-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-34"><img src="ambient_methods_overview_hsa_parse_files/figure-html/plot-biotype-output-2.png" id="plot-biotype-2" width="809" height="649" class="figure-img"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="check-top-genes" class="level2">
<h2 class="anchored" data-anchor-id="check-top-genes">Check top genes</h2>
<div id="topgenes" class="cell" data-execution_count="14">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> sname <span class="kw">in</span> alldata.keys():</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>    display(Markdown(<span class="ss">f"## </span><span class="sc">{</span>sname<span class="sc">}</span><span class="ss">"</span>))</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>    sc.pl.highest_expr_genes(alldata[sname])</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>    display(Markdown(<span class="ss">f" "</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<section id="parse1_all-1" class="level2">
<h2 class="anchored" data-anchor-id="parse1_all-1">parse1_all</h2>
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/topgenes-output-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-35"><img src="ambient_methods_overview_hsa_parse_files/figure-html/topgenes-output-2.png" id="topgenes-2" width="422" height="513"></a></p>
</section>
<section id="parse1_5-azacytidine-1" class="level2">
<h2 class="anchored" data-anchor-id="parse1_5-azacytidine-1">parse1_5-Azacytidine</h2>
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/topgenes-output-5.png" class="lightbox" data-gallery="quarto-lightbox-gallery-36"><img src="ambient_methods_overview_hsa_parse_files/figure-html/topgenes-output-5.png" id="topgenes-5" width="422" height="513"></a></p>
</section>
<section id="parse1_fludarabine-1" class="level2">
<h2 class="anchored" data-anchor-id="parse1_fludarabine-1">parse1_Fludarabine</h2>
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/topgenes-output-8.png" class="lightbox" data-gallery="quarto-lightbox-gallery-37"><img src="ambient_methods_overview_hsa_parse_files/figure-html/topgenes-output-8.png" id="topgenes-8" width="422" height="513"></a></p>
</section>
<section id="parse1_xrp44x-1" class="level2">
<h2 class="anchored" data-anchor-id="parse1_xrp44x-1">parse1_XRP44X</h2>
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/topgenes-output-11.png" class="lightbox" data-gallery="quarto-lightbox-gallery-38"><img src="ambient_methods_overview_hsa_parse_files/figure-html/topgenes-output-11.png" id="topgenes-11" width="422" height="513"></a></p>
</section>
<section id="parse1_control-1" class="level2">
<h2 class="anchored" data-anchor-id="parse1_control-1">parse1_Control</h2>
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/topgenes-output-14.png" class="lightbox" data-gallery="quarto-lightbox-gallery-39"><img src="ambient_methods_overview_hsa_parse_files/figure-html/topgenes-output-14.png" id="topgenes-14" width="422" height="513"></a></p>
</section>
<section id="parse1_dexamethasone-1" class="level2">
<h2 class="anchored" data-anchor-id="parse1_dexamethasone-1">parse1_Dexamethasone</h2>
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/topgenes-output-17.png" class="lightbox" data-gallery="quarto-lightbox-gallery-40"><img src="ambient_methods_overview_hsa_parse_files/figure-html/topgenes-output-17.png" id="topgenes-17" width="422" height="513"></a></p>
</section>
<section id="parse1_imatinib-1" class="level2">
<h2 class="anchored" data-anchor-id="parse1_imatinib-1">parse1_Imatinib</h2>
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/topgenes-output-20.png" class="lightbox" data-gallery="quarto-lightbox-gallery-41"><img src="ambient_methods_overview_hsa_parse_files/figure-html/topgenes-output-20.png" id="topgenes-20" width="422" height="513"></a></p>
</section>
<section id="parse2_all-1" class="level2">
<h2 class="anchored" data-anchor-id="parse2_all-1">parse2_all</h2>
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/topgenes-output-23.png" class="lightbox" data-gallery="quarto-lightbox-gallery-42"><img src="ambient_methods_overview_hsa_parse_files/figure-html/topgenes-output-23.png" id="topgenes-23" width="422" height="513"></a></p>
</section>
<section id="parse2_5-azacytidine-1" class="level2">
<h2 class="anchored" data-anchor-id="parse2_5-azacytidine-1">parse2_5-Azacytidine</h2>
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/topgenes-output-26.png" class="lightbox" data-gallery="quarto-lightbox-gallery-43"><img src="ambient_methods_overview_hsa_parse_files/figure-html/topgenes-output-26.png" id="topgenes-26" width="422" height="513"></a></p>
</section>
<section id="parse2_fludarabine-1" class="level2">
<h2 class="anchored" data-anchor-id="parse2_fludarabine-1">parse2_Fludarabine</h2>
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/topgenes-output-29.png" class="lightbox" data-gallery="quarto-lightbox-gallery-44"><img src="ambient_methods_overview_hsa_parse_files/figure-html/topgenes-output-29.png" id="topgenes-29" width="422" height="513"></a></p>
</section>
<section id="parse2_xrp44x-1" class="level2">
<h2 class="anchored" data-anchor-id="parse2_xrp44x-1">parse2_XRP44X</h2>
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/topgenes-output-32.png" class="lightbox" data-gallery="quarto-lightbox-gallery-45"><img src="ambient_methods_overview_hsa_parse_files/figure-html/topgenes-output-32.png" id="topgenes-32" width="424" height="513"></a></p>
</section>
<section id="parse2_control-1" class="level2">
<h2 class="anchored" data-anchor-id="parse2_control-1">parse2_Control</h2>
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/topgenes-output-35.png" class="lightbox" data-gallery="quarto-lightbox-gallery-46"><img src="ambient_methods_overview_hsa_parse_files/figure-html/topgenes-output-35.png" id="topgenes-35" width="422" height="513"></a></p>
</section>
<section id="parse2_dexamethasone-1" class="level2">
<h2 class="anchored" data-anchor-id="parse2_dexamethasone-1">parse2_Dexamethasone</h2>
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/topgenes-output-38.png" class="lightbox" data-gallery="quarto-lightbox-gallery-47"><img src="ambient_methods_overview_hsa_parse_files/figure-html/topgenes-output-38.png" id="topgenes-38" width="422" height="513"></a></p>
</section>
<section id="parse2_imatinib-1" class="level2">
<h2 class="anchored" data-anchor-id="parse2_imatinib-1">parse2_Imatinib</h2>
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/topgenes-output-41.png" class="lightbox" data-gallery="quarto-lightbox-gallery-48"><img src="ambient_methods_overview_hsa_parse_files/figure-html/topgenes-output-41.png" id="topgenes-41" width="422" height="513"></a></p>
</section>
</div>
</section>
<section id="parse-data" class="level2">
<h2 class="anchored" data-anchor-id="parse-data">Parse data</h2>
<p>Closer comparison of the bg predictions with all vs each sample.</p>
<section id="cell-filtering" class="level3">
<h3 class="anchored" data-anchor-id="cell-filtering">Cell filtering</h3>
<p>Compare the predictions from cellbender using all vs each sample separately. Clearly have different barcode rank plots.</p>
<div id="cell-parse-filtering" class="cell" data-execution_count="15">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>ed <span class="op">=</span> alldata[<span class="st">'parse1_all'</span>].uns[<span class="st">'raw_obs'</span>][<span class="st">'isCellED'</span>]</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>ed_stats <span class="op">=</span> {}</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>cb <span class="op">=</span> alldata[<span class="st">'parse1_all'</span>].uns[<span class="st">'raw_obs'</span>][<span class="st">'inCB'</span>]</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>cb_stats <span class="op">=</span> {}</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> sample <span class="kw">in</span> alldata.keys():</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="st">"parse1"</span> <span class="kw">in</span> sample <span class="kw">and</span> sample <span class="op">!=</span> <span class="st">'parse1_all'</span>:</span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>        tmp <span class="op">=</span> alldata[sample].uns[<span class="st">'raw_obs'</span>][<span class="st">'isCellED'</span>]</span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>        tmp2 <span class="op">=</span> pd.concat([ed,tmp], axis<span class="op">=</span><span class="dv">1</span>, join<span class="op">=</span><span class="st">'inner'</span>)</span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>        tmp2.columns <span class="op">=</span> [<span class="st">'all'</span>,<span class="st">'s'</span>]</span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a>        tmp2 <span class="op">=</span> tmp2.value_counts().reset_index()</span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a>        tmp2.index <span class="op">=</span> tmp2[<span class="st">'all'</span>].astype(<span class="st">'str'</span>) <span class="op">+</span> tmp2[<span class="st">'s'</span>].astype(<span class="st">'str'</span>)</span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a>        ed_stats[sample] <span class="op">=</span> tmp2[<span class="dv">0</span>]</span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a>        tmp <span class="op">=</span> alldata[sample].uns[<span class="st">'raw_obs'</span>][<span class="st">'inCB'</span>]</span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a>        tmp2 <span class="op">=</span> pd.concat([cb,tmp], axis<span class="op">=</span><span class="dv">1</span>, join<span class="op">=</span><span class="st">'inner'</span>)</span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true" tabindex="-1"></a>        tmp2.columns <span class="op">=</span> [<span class="st">'all'</span>,<span class="st">'s'</span>]</span>
<span id="cb16-17"><a href="#cb16-17" aria-hidden="true" tabindex="-1"></a>        tmp2 <span class="op">=</span> tmp2.value_counts().reset_index()</span>
<span id="cb16-18"><a href="#cb16-18" aria-hidden="true" tabindex="-1"></a>        tmp2.index <span class="op">=</span> tmp2[<span class="st">'all'</span>].astype(<span class="st">'str'</span>) <span class="op">+</span> tmp2[<span class="st">'s'</span>].astype(<span class="st">'str'</span>)</span>
<span id="cb16-19"><a href="#cb16-19" aria-hidden="true" tabindex="-1"></a>        cb_stats[sample] <span class="op">=</span> tmp2[<span class="dv">0</span>]</span>
<span id="cb16-20"><a href="#cb16-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-21"><a href="#cb16-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-22"><a href="#cb16-22" aria-hidden="true" tabindex="-1"></a>fig, axs <span class="op">=</span> plt.subplots(<span class="dv">1</span>, <span class="dv">2</span>, figsize<span class="op">=</span>(<span class="dv">10</span>,<span class="dv">8</span>),constrained_layout<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb16-23"><a href="#cb16-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-24"><a href="#cb16-24" aria-hidden="true" tabindex="-1"></a>ed_stats <span class="op">=</span> pd.DataFrame(ed_stats)</span>
<span id="cb16-25"><a href="#cb16-25" aria-hidden="true" tabindex="-1"></a>ed_stats <span class="op">=</span> ed_stats.iloc[<span class="dv">1</span>:,:]</span>
<span id="cb16-26"><a href="#cb16-26" aria-hidden="true" tabindex="-1"></a>ed_stats.index <span class="op">=</span> [<span class="st">'Sample only'</span>, <span class="st">'All only'</span>,<span class="st">"Both"</span>]</span>
<span id="cb16-27"><a href="#cb16-27" aria-hidden="true" tabindex="-1"></a>ed_stats.T.plot(kind<span class="op">=</span><span class="st">"bar"</span>,stacked<span class="op">=</span><span class="va">True</span>, title<span class="op">=</span><span class="st">'parse1 samples EmptyDrops'</span>, ax <span class="op">=</span> axs[<span class="dv">0</span>])</span>
<span id="cb16-28"><a href="#cb16-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-29"><a href="#cb16-29" aria-hidden="true" tabindex="-1"></a>cb_stats <span class="op">=</span> pd.DataFrame(cb_stats)</span>
<span id="cb16-30"><a href="#cb16-30" aria-hidden="true" tabindex="-1"></a>cb_stats <span class="op">=</span> cb_stats.iloc[<span class="dv">1</span>:,:]</span>
<span id="cb16-31"><a href="#cb16-31" aria-hidden="true" tabindex="-1"></a>cb_stats.index <span class="op">=</span> [<span class="st">'Sample only'</span>, <span class="st">'All only'</span>,<span class="st">"Both"</span>]</span>
<span id="cb16-32"><a href="#cb16-32" aria-hidden="true" tabindex="-1"></a>cb_stats.T.plot(kind<span class="op">=</span><span class="st">"bar"</span>,stacked<span class="op">=</span><span class="va">True</span>, title<span class="op">=</span><span class="st">'parse1 samples CellBender'</span>, ax <span class="op">=</span> axs[<span class="dv">1</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/parse-filtering-output-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-49"><img src="ambient_methods_overview_hsa_parse_files/figure-html/parse-filtering-output-1.png" id="parse-filtering" width="809" height="649" class="figure-img"></a></p>
</figure>
</div>
</div>
</div>
<div id="cell-parse-filtering2" class="cell" data-execution_count="16">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>ed <span class="op">=</span> alldata[<span class="st">'parse2_all'</span>].uns[<span class="st">'raw_obs'</span>][<span class="st">'isCellED'</span>]</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>ed_stats <span class="op">=</span> {}</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>cb <span class="op">=</span> alldata[<span class="st">'parse2_all'</span>].uns[<span class="st">'raw_obs'</span>][<span class="st">'inCB'</span>]</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>cb_stats <span class="op">=</span> {}</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> sample <span class="kw">in</span> alldata.keys():</span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="st">"parse2"</span> <span class="kw">in</span> sample <span class="kw">and</span> sample <span class="op">!=</span> <span class="st">'parse2_all'</span>:</span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>        tmp <span class="op">=</span> alldata[sample].uns[<span class="st">'raw_obs'</span>][<span class="st">'isCellED'</span>]</span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>        tmp2 <span class="op">=</span> pd.concat([ed,tmp], axis<span class="op">=</span><span class="dv">1</span>, join<span class="op">=</span><span class="st">'inner'</span>)</span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>        tmp2.columns <span class="op">=</span> [<span class="st">'all'</span>,<span class="st">'s'</span>]</span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a>        tmp2 <span class="op">=</span> tmp2.value_counts().reset_index()</span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a>        tmp2.index <span class="op">=</span> tmp2[<span class="st">'all'</span>].astype(<span class="st">'str'</span>) <span class="op">+</span> tmp2[<span class="st">'s'</span>].astype(<span class="st">'str'</span>)</span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a>        ed_stats[sample] <span class="op">=</span> tmp2[<span class="dv">0</span>]</span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a>        tmp <span class="op">=</span> alldata[sample].uns[<span class="st">'raw_obs'</span>][<span class="st">'inCB'</span>]</span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true" tabindex="-1"></a>        tmp2 <span class="op">=</span> pd.concat([cb,tmp], axis<span class="op">=</span><span class="dv">1</span>, join<span class="op">=</span><span class="st">'inner'</span>)</span>
<span id="cb17-16"><a href="#cb17-16" aria-hidden="true" tabindex="-1"></a>        tmp2.columns <span class="op">=</span> [<span class="st">'all'</span>,<span class="st">'s'</span>]</span>
<span id="cb17-17"><a href="#cb17-17" aria-hidden="true" tabindex="-1"></a>        tmp2 <span class="op">=</span> tmp2.value_counts().reset_index()</span>
<span id="cb17-18"><a href="#cb17-18" aria-hidden="true" tabindex="-1"></a>        tmp2.index <span class="op">=</span> tmp2[<span class="st">'all'</span>].astype(<span class="st">'str'</span>) <span class="op">+</span> tmp2[<span class="st">'s'</span>].astype(<span class="st">'str'</span>)</span>
<span id="cb17-19"><a href="#cb17-19" aria-hidden="true" tabindex="-1"></a>        cb_stats[sample] <span class="op">=</span> tmp2[<span class="dv">0</span>]</span>
<span id="cb17-20"><a href="#cb17-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-21"><a href="#cb17-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-22"><a href="#cb17-22" aria-hidden="true" tabindex="-1"></a>fig, axs <span class="op">=</span> plt.subplots(<span class="dv">1</span>, <span class="dv">2</span>, figsize<span class="op">=</span>(<span class="dv">10</span>,<span class="dv">8</span>),constrained_layout<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb17-23"><a href="#cb17-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-24"><a href="#cb17-24" aria-hidden="true" tabindex="-1"></a>ed_stats <span class="op">=</span> pd.DataFrame(ed_stats)</span>
<span id="cb17-25"><a href="#cb17-25" aria-hidden="true" tabindex="-1"></a>ed_stats <span class="op">=</span> ed_stats.iloc[<span class="dv">1</span>:,:]</span>
<span id="cb17-26"><a href="#cb17-26" aria-hidden="true" tabindex="-1"></a>ed_stats.index <span class="op">=</span> [<span class="st">'Sample only'</span>, <span class="st">'All only'</span>,<span class="st">"Both"</span>]</span>
<span id="cb17-27"><a href="#cb17-27" aria-hidden="true" tabindex="-1"></a>ed_stats.T.plot(kind<span class="op">=</span><span class="st">"bar"</span>,stacked<span class="op">=</span><span class="va">True</span>, title<span class="op">=</span><span class="st">'parse2 samples EmptyDrops'</span>, ax <span class="op">=</span> axs[<span class="dv">0</span>])</span>
<span id="cb17-28"><a href="#cb17-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-29"><a href="#cb17-29" aria-hidden="true" tabindex="-1"></a>cb_stats <span class="op">=</span> pd.DataFrame(cb_stats)</span>
<span id="cb17-30"><a href="#cb17-30" aria-hidden="true" tabindex="-1"></a>cb_stats <span class="op">=</span> cb_stats.iloc[<span class="dv">1</span>:,:]</span>
<span id="cb17-31"><a href="#cb17-31" aria-hidden="true" tabindex="-1"></a>cb_stats.index <span class="op">=</span> [<span class="st">'Sample only'</span>, <span class="st">'All only'</span>,<span class="st">"Both"</span>]</span>
<span id="cb17-32"><a href="#cb17-32" aria-hidden="true" tabindex="-1"></a>cb_stats.T.plot(kind<span class="op">=</span><span class="st">"bar"</span>,stacked<span class="op">=</span><span class="va">True</span>, title<span class="op">=</span><span class="st">'parse2 samples CellBender'</span>, ax <span class="op">=</span> axs[<span class="dv">1</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/parse-filtering2-output-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-50"><img src="ambient_methods_overview_hsa_parse_files/figure-html/parse-filtering2-output-1.png" id="parse-filtering2" width="809" height="649" class="figure-img"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="ambient-profile" class="level3">
<h3 class="anchored" data-anchor-id="ambient-profile">Ambient profile</h3>
<div id="cell-ambient-overlap" class="cell" data-execution_count="17">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># remove hto genes from multiseq.</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>ed <span class="op">=</span> pd.DataFrame([emptydrops_genes[sample].ambient <span class="cf">for</span> sample <span class="kw">in</span> emptydrops_genes.keys()]  )</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>ed.index <span class="op">=</span> emptydrops_genes.keys()</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Top genes per row</span></span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a><span class="co">#topA = gstats['ambient'].sort_values(ascending=False).head(10).index</span></span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>ed.<span class="bu">apply</span>(<span class="kw">lambda</span> x: x.nlargest(<span class="dv">10</span>).index.tolist(), axis<span class="op">=</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div id="ambient-overlap" class="cell-output cell-output-display" data-execution_count="35">
<pre><code>parse1_all              [MALAT1, AC109466.1, AL163541.1, EBF1, AUTS2, ...
parse1_5-Azacytidine    [MALAT1, AC109466.1, AL163541.1, EBF1, AUTS2, ...
parse1_Fludarabine      [MALAT1, AC109466.1, AUTS2, EBF1, AL163541.1, ...
parse1_XRP44X           [MALAT1, AC109466.1, EBF1, AL163541.1, AUTS2, ...
parse1_Control          [MALAT1, AC109466.1, EBF1, AL163541.1, AUTS2, ...
parse1_Dexamethasone    [MALAT1, AC109466.1, EBF1, AL163541.1, AUTS2, ...
parse1_Imatinib         [MALAT1, AC109466.1, AL163541.1, EBF1, MSI2, A...
parse2_all              [MALAT1, AC109466.1, EBF1, AL163541.1, AUTS2, ...
parse2_5-Azacytidine    [MALAT1, AC109466.1, EBF1, AL163541.1, AUTS2, ...
parse2_Fludarabine      [MALAT1, AC109466.1, AUTS2, EBF1, AL163541.1, ...
parse2_XRP44X           [MALAT1, AC109466.1, EBF1, AL163541.1, AUTS2, ...
parse2_Control          [MALAT1, EBF1, AC109466.1, AL163541.1, AUTS2, ...
parse2_Dexamethasone    [MALAT1, AC109466.1, EBF1, AL163541.1, AUTS2, ...
parse2_Imatinib         [MALAT1, AC109466.1, AL163541.1, EBF1, AUTS2, ...
dtype: object</code></pre>
</div>
</div>
<p>Top 10 genes per sample in heatmap colored by emptyDrops score.</p>
<div id="cell-ambient-overlap2" class="cell" data-fig.height="8" data-fig.width="14" data-execution_count="18">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>tmp <span class="op">=</span> ed.<span class="bu">apply</span>(<span class="kw">lambda</span> x: x.nlargest(<span class="dv">10</span>).index.tolist(), axis<span class="op">=</span><span class="dv">1</span>).tolist()</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>l <span class="op">=</span> <span class="bu">list</span>(<span class="bu">set</span>(<span class="bu">sum</span>(tmp, [])))</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> ed.loc[:,l]</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>plt.pcolor(df)</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>plt.yticks(np.arange(<span class="fl">0.5</span>, <span class="bu">len</span>(df.index), <span class="dv">1</span>), df.index, size<span class="op">=</span><span class="dv">8</span>)</span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>plt.xticks(np.arange(<span class="fl">0.5</span>, <span class="bu">len</span>(df.columns), <span class="dv">1</span>), df.columns, rotation <span class="op">=</span> <span class="dv">45</span>, size <span class="op">=</span> <span class="dv">6</span>, ha<span class="op">=</span><span class="st">'right'</span>)</span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_hsa_parse_files/figure-html/ambient-overlap2-output-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-51"><img src="ambient_methods_overview_hsa_parse_files/figure-html/ambient-overlap2-output-1.png" id="ambient-overlap2" width="370" height="297" class="figure-img"></a></p>
</figure>
</div>
</div>
</div>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<script>var lightboxQuarto = GLightbox({"loop":false,"closeEffect":"zoom","openEffect":"zoom","selector":".lightbox","descPosition":"bottom"});
(function() {
  let previousOnload = window.onload;
  window.onload = () => {
    if (previousOnload) {
      previousOnload();
    }
    lightboxQuarto.on('slide_before_load', (data) => {
      const { slideIndex, slideNode, slideConfig, player, trigger } = data;
      const href = trigger.getAttribute('href');
      if (href !== null) {
        const imgEl = window.document.querySelector(`a[href="${href}"] img`);
        if (imgEl !== null) {
          const srcAttr = imgEl.getAttribute("src");
          if (srcAttr && srcAttr.startsWith("data:")) {
            slideConfig.href = srcAttr;
          }
        }
      } 
    });
  
    lightboxQuarto.on('slide_after_load', (data) => {
      const { slideIndex, slideNode, slideConfig, player, trigger } = data;
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(slideNode);
      }
    });
  
  };
  
})();
          </script>




</body></html>