<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="description" content="Visualizing cellbender, soupX and decontX output">

<title>Ambient RNA removal</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="ambient_methods_overview_files/libs/clipboard/clipboard.min.js"></script>
<script src="ambient_methods_overview_files/libs/quarto-html/quarto.js"></script>
<script src="ambient_methods_overview_files/libs/quarto-html/popper.min.js"></script>
<script src="ambient_methods_overview_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="ambient_methods_overview_files/libs/quarto-html/anchor.min.js"></script>
<link href="ambient_methods_overview_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="ambient_methods_overview_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="ambient_methods_overview_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="ambient_methods_overview_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="ambient_methods_overview_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script src="ambient_methods_overview_files/libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="ambient_methods_overview_files/libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="ambient_methods_overview_files/libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet">


</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active" data-toc-expanded="1">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#read-in-data." id="toc-read-in-data." class="nav-link active" data-scroll-target="#read-in-data.">Read in data.</a></li>
  <li><a href="#cell-filtering-cellbender" id="toc-cell-filtering-cellbender" class="nav-link" data-scroll-target="#cell-filtering-cellbender">Cell filtering cellbender</a></li>
  <li><a href="#filtering-stats-all-methods" id="toc-filtering-stats-all-methods" class="nav-link" data-scroll-target="#filtering-stats-all-methods">Filtering stats all methods</a></li>
  <li><a href="#gene-stats" id="toc-gene-stats" class="nav-link" data-scroll-target="#gene-stats">Gene stats</a></li>
  <li><a href="#mito-etc." id="toc-mito-etc." class="nav-link" data-scroll-target="#mito-etc.">Mito, etc.</a></li>
  <li><a href="#biotype-proportions" id="toc-biotype-proportions" class="nav-link" data-scroll-target="#biotype-proportions">Biotype proportions</a></li>
  </ul>
</nav>
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Ambient RNA removal</h1>
</div>

<div>
  <div class="description">
    Visualizing cellbender, soupX and decontX output
  </div>
</div>


<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>Mouse samples from GEO with raw cellranger matrices:</p>
<ul>
<li>GSE213825 - root ganglia scRNAseq, 16 samples, Names SAM24349928-SAM24374041</li>
<li>GSE218853 - mixed strain kidney samples, sc/sn, 5 samples: rep1, rep2, rep3, nuc1, nuc2</li>
<li>GSE229059 - intramuscular immunization, immune cells. 24 samples, Names GSM7150713-GSM7150739</li>
<li>GSE244142 - mouse tumor model, snRNAseq, 8 samples, Names GSM7807801_SITTA8 - GSM7807808_SITTH9</li>
<li>GSE261852 - mouse CAR-T cells, scRNAseq, 2 samples, spl1, spl2</li>
</ul>
<div id="2717f3bc" class="cell" data-execution_count="1">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> glob</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> warnings</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pyreadr</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> IPython.display <span class="im">import</span> display, Markdown</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>warnings.simplefilter(action<span class="op">=</span><span class="st">'ignore'</span>, category<span class="op">=</span><span class="pp">Warning</span>)</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="co"># verbosity: errors (0), warnings (1), info (2), hints (3)</span></span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a>sc.settings.verbosity <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a>sc.settings.set_figure_params(dpi<span class="op">=</span><span class="dv">80</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<section id="read-in-data." class="level3">
<h3 class="anchored" data-anchor-id="read-in-data.">Read in data.</h3>
<p>First, read in the annotation file for biotypes.</p>
<p>Make broader groupings:</p>
<ul>
<li>Mito genes - both PC and nc.</li>
<li>PC genes - includes also IG/TR genes</li>
<li>lncRNA</li>
<li>other - all other categories.</li>
</ul>
<div id="7f37eebd" class="cell" data-execution_count="2">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>indir <span class="op">=</span> <span class="st">"/Users/asabjor/projects/sc-devop/scQC/data/mouse"</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>cbdir <span class="op">=</span> os.path.join(indir,<span class="st">"output"</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>annot_file <span class="op">=</span> os.path.join(cbdir, <span class="st">"mouse_annotation.csv"</span>)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> os.path.exists(annot_file):</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>    annot <span class="op">=</span> pd.read_csv(annot_file, index_col<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span>:   </span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>    annot <span class="op">=</span> sc.queries.biomart_annotations(<span class="st">"mmusculus"</span>, [<span class="st">"ensembl_gene_id"</span>, <span class="st">"external_gene_name"</span>, <span class="st">"start_position"</span>, <span class="st">"end_position"</span>, <span class="st">"chromosome_name"</span>, <span class="st">"gene_biotype"</span>] )</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>    annot.to_csv(annot_file)</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a><span class="co"># make slim biotype</span></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>annot[<span class="st">"gene_class"</span>]<span class="op">=</span> <span class="st">"other"</span></span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>pc_genes <span class="op">=</span> [<span class="st">"protein_coding"</span>,<span class="st">"IG_V_gene"</span>,<span class="st">"IG_D_gene"</span>,<span class="st">"IG_J_gene"</span>,<span class="st">"IG_C_gene"</span>,<span class="st">"IG_LV_gene"</span>, <span class="st">"TR_V_gene"</span>,<span class="st">"TR_J_gene"</span>,<span class="st">"TR_C_gene"</span>,<span class="st">"TR_D_gene"</span> ]</span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a>annot[<span class="st">"gene_class"</span>][annot.gene_biotype.isin(pc_genes)] <span class="op">=</span> <span class="st">"pc"</span></span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>annot[<span class="st">"gene_class"</span>][annot.gene_biotype <span class="op">==</span> <span class="st">"lncRNA"</span>] <span class="op">=</span> <span class="st">"lncRNA"</span></span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>annot[<span class="st">"gene_class"</span>][annot.external_gene_name.<span class="bu">str</span>.startswith(<span class="st">'mt-'</span>).fillna(<span class="va">False</span>)] <span class="op">=</span> <span class="st">"mt"</span></span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a><span class="co"># define list of all classes.</span></span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a>gene_classes <span class="op">=</span> [<span class="st">"mt"</span>,<span class="st">"pc"</span>,<span class="st">"lncRNA"</span>,<span class="st">"other"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Then read in all the raw and cellbender, soupX, decontX files. For soupX select 2 different resolutions (0.2, 0.8).</p>
<p>For soupX and decontX output is not in integer counts, round the values!</p>
<div id="ccaef617" class="cell" data-execution_count="3">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># does not work to read in the cellbender files from GSE244142 with sc.read_10x_h5</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="co"># gives zero genes.</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="co"># minor modifications of the scanpy function, now works to load the data!</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> h5py</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> anndata <span class="im">as</span> ad</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> _collect_datasets(dsets: <span class="bu">dict</span>, group: h5py.Group):</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> k, v <span class="kw">in</span> group.items():</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> <span class="bu">isinstance</span>(v, h5py.Dataset):</span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>            dsets[k] <span class="op">=</span> v[()]</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>        <span class="cf">else</span>:</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>            _collect_datasets(dsets, v)</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> _read_v3_10x_h5(filename, <span class="op">*</span>, start<span class="op">=</span><span class="va">None</span>):</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""</span></span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a><span class="co">    Read hdf5 file from Cell Ranger v3 or later versions.</span></span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a>    <span class="cf">with</span> h5py.File(<span class="bu">str</span>(filename), <span class="st">"r"</span>) <span class="im">as</span> f:</span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a>        <span class="cf">try</span>:</span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a>            dsets <span class="op">=</span> {}</span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a>            _collect_datasets(dsets, f[<span class="st">"matrix"</span>])</span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a>            <span class="im">from</span> scipy.sparse <span class="im">import</span> csr_matrix</span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true" tabindex="-1"></a>            M, N <span class="op">=</span> dsets[<span class="st">"shape"</span>]</span>
<span id="cb3-27"><a href="#cb3-27" aria-hidden="true" tabindex="-1"></a>            data <span class="op">=</span> dsets[<span class="st">"data"</span>]</span>
<span id="cb3-28"><a href="#cb3-28" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> dsets[<span class="st">"data"</span>].dtype <span class="op">==</span> np.dtype(<span class="st">"int32"</span>):</span>
<span id="cb3-29"><a href="#cb3-29" aria-hidden="true" tabindex="-1"></a>                data <span class="op">=</span> dsets[<span class="st">"data"</span>].view(<span class="st">"float32"</span>)</span>
<span id="cb3-30"><a href="#cb3-30" aria-hidden="true" tabindex="-1"></a>                data[:] <span class="op">=</span> dsets[<span class="st">"data"</span>]</span>
<span id="cb3-31"><a href="#cb3-31" aria-hidden="true" tabindex="-1"></a>            matrix <span class="op">=</span> csr_matrix(</span>
<span id="cb3-32"><a href="#cb3-32" aria-hidden="true" tabindex="-1"></a>                (data, dsets[<span class="st">"indices"</span>], dsets[<span class="st">"indptr"</span>]),</span>
<span id="cb3-33"><a href="#cb3-33" aria-hidden="true" tabindex="-1"></a>                shape<span class="op">=</span>(N, M),</span>
<span id="cb3-34"><a href="#cb3-34" aria-hidden="true" tabindex="-1"></a>            )</span>
<span id="cb3-35"><a href="#cb3-35" aria-hidden="true" tabindex="-1"></a>            obs_dict <span class="op">=</span> {<span class="st">"obs_names"</span>: dsets[<span class="st">"barcodes"</span>].astype(<span class="bu">str</span>)}</span>
<span id="cb3-36"><a href="#cb3-36" aria-hidden="true" tabindex="-1"></a>            var_dict <span class="op">=</span> {<span class="st">"var_names"</span>: dsets[<span class="st">"name"</span>].astype(<span class="bu">str</span>)}</span>
<span id="cb3-37"><a href="#cb3-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-38"><a href="#cb3-38" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> <span class="st">"gene_id"</span> <span class="kw">not</span> <span class="kw">in</span> dsets:</span>
<span id="cb3-39"><a href="#cb3-39" aria-hidden="true" tabindex="-1"></a>                <span class="co"># Read metadata specific to a feature-barcode matrix</span></span>
<span id="cb3-40"><a href="#cb3-40" aria-hidden="true" tabindex="-1"></a>                var_dict[<span class="st">"gene_ids"</span>] <span class="op">=</span> dsets[<span class="st">"id"</span>].astype(<span class="bu">str</span>)</span>
<span id="cb3-41"><a href="#cb3-41" aria-hidden="true" tabindex="-1"></a>            <span class="cf">else</span>:</span>
<span id="cb3-42"><a href="#cb3-42" aria-hidden="true" tabindex="-1"></a>                <span class="co"># Read metadata specific to a probe-barcode matrix</span></span>
<span id="cb3-43"><a href="#cb3-43" aria-hidden="true" tabindex="-1"></a>                var_dict.update(</span>
<span id="cb3-44"><a href="#cb3-44" aria-hidden="true" tabindex="-1"></a>                    {</span>
<span id="cb3-45"><a href="#cb3-45" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"gene_ids"</span>: dsets[<span class="st">"gene_id"</span>].astype(<span class="bu">str</span>),</span>
<span id="cb3-46"><a href="#cb3-46" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"probe_ids"</span>: dsets[<span class="st">"id"</span>].astype(<span class="bu">str</span>),</span>
<span id="cb3-47"><a href="#cb3-47" aria-hidden="true" tabindex="-1"></a>                    }</span>
<span id="cb3-48"><a href="#cb3-48" aria-hidden="true" tabindex="-1"></a>                )</span>
<span id="cb3-49"><a href="#cb3-49" aria-hidden="true" tabindex="-1"></a>            var_dict[<span class="st">"feature_types"</span>] <span class="op">=</span> dsets[<span class="st">"feature_type"</span>].astype(<span class="bu">str</span>)</span>
<span id="cb3-50"><a href="#cb3-50" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> <span class="st">"filtered_barcodes"</span> <span class="kw">in</span> f[<span class="st">"matrix"</span>]:</span>
<span id="cb3-51"><a href="#cb3-51" aria-hidden="true" tabindex="-1"></a>                obs_dict[<span class="st">"filtered_barcodes"</span>] <span class="op">=</span> dsets[<span class="st">"filtered_barcodes"</span>].astype(<span class="bu">bool</span>)</span>
<span id="cb3-52"><a href="#cb3-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-53"><a href="#cb3-53" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> <span class="st">"features"</span> <span class="kw">in</span> f[<span class="st">"matrix"</span>]:</span>
<span id="cb3-54"><a href="#cb3-54" aria-hidden="true" tabindex="-1"></a>                var_dict.update(</span>
<span id="cb3-55"><a href="#cb3-55" aria-hidden="true" tabindex="-1"></a>                    (</span>
<span id="cb3-56"><a href="#cb3-56" aria-hidden="true" tabindex="-1"></a>                        feature_metadata_name,</span>
<span id="cb3-57"><a href="#cb3-57" aria-hidden="true" tabindex="-1"></a>                        dsets[feature_metadata_name].astype(</span>
<span id="cb3-58"><a href="#cb3-58" aria-hidden="true" tabindex="-1"></a>                            <span class="bu">bool</span> <span class="cf">if</span> feature_metadata_item.dtype.kind <span class="op">==</span> <span class="st">"b"</span> <span class="cf">else</span> <span class="bu">str</span></span>
<span id="cb3-59"><a href="#cb3-59" aria-hidden="true" tabindex="-1"></a>                        ),</span>
<span id="cb3-60"><a href="#cb3-60" aria-hidden="true" tabindex="-1"></a>                    )</span>
<span id="cb3-61"><a href="#cb3-61" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">for</span> feature_metadata_name, feature_metadata_item <span class="kw">in</span> f[<span class="st">"matrix"</span>][</span>
<span id="cb3-62"><a href="#cb3-62" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"features"</span></span>
<span id="cb3-63"><a href="#cb3-63" aria-hidden="true" tabindex="-1"></a>                    ].items()</span>
<span id="cb3-64"><a href="#cb3-64" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">if</span> <span class="bu">isinstance</span>(feature_metadata_item, h5py.Dataset)</span>
<span id="cb3-65"><a href="#cb3-65" aria-hidden="true" tabindex="-1"></a>                    <span class="kw">and</span> feature_metadata_name</span>
<span id="cb3-66"><a href="#cb3-66" aria-hidden="true" tabindex="-1"></a>                    <span class="kw">not</span> <span class="kw">in</span> [</span>
<span id="cb3-67"><a href="#cb3-67" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"name"</span>,</span>
<span id="cb3-68"><a href="#cb3-68" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"feature_type"</span>,</span>
<span id="cb3-69"><a href="#cb3-69" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"id"</span>,</span>
<span id="cb3-70"><a href="#cb3-70" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"gene_id"</span>,</span>
<span id="cb3-71"><a href="#cb3-71" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"_all_tag_keys"</span>,</span>
<span id="cb3-72"><a href="#cb3-72" aria-hidden="true" tabindex="-1"></a>                    ]</span>
<span id="cb3-73"><a href="#cb3-73" aria-hidden="true" tabindex="-1"></a>                )</span>
<span id="cb3-74"><a href="#cb3-74" aria-hidden="true" tabindex="-1"></a>            <span class="cf">else</span>:</span>
<span id="cb3-75"><a href="#cb3-75" aria-hidden="true" tabindex="-1"></a>                msg <span class="op">=</span> <span class="st">"10x h5 has no features group"</span></span>
<span id="cb3-76"><a href="#cb3-76" aria-hidden="true" tabindex="-1"></a>                <span class="cf">raise</span> <span class="pp">ValueError</span>(msg)</span>
<span id="cb3-77"><a href="#cb3-77" aria-hidden="true" tabindex="-1"></a>            adata <span class="op">=</span> ad.AnnData(</span>
<span id="cb3-78"><a href="#cb3-78" aria-hidden="true" tabindex="-1"></a>                matrix,</span>
<span id="cb3-79"><a href="#cb3-79" aria-hidden="true" tabindex="-1"></a>                obs<span class="op">=</span>obs_dict,</span>
<span id="cb3-80"><a href="#cb3-80" aria-hidden="true" tabindex="-1"></a>                var<span class="op">=</span>var_dict,</span>
<span id="cb3-81"><a href="#cb3-81" aria-hidden="true" tabindex="-1"></a>            )</span>
<span id="cb3-82"><a href="#cb3-82" aria-hidden="true" tabindex="-1"></a>            <span class="co">#logg.info("", time=start)</span></span>
<span id="cb3-83"><a href="#cb3-83" aria-hidden="true" tabindex="-1"></a>            <span class="cf">return</span> adata</span>
<span id="cb3-84"><a href="#cb3-84" aria-hidden="true" tabindex="-1"></a>        <span class="cf">except</span> <span class="pp">KeyError</span>:</span>
<span id="cb3-85"><a href="#cb3-85" aria-hidden="true" tabindex="-1"></a>            msg <span class="op">=</span> <span class="st">"File is missing one or more required datasets."</span></span>
<span id="cb3-86"><a href="#cb3-86" aria-hidden="true" tabindex="-1"></a>            <span class="cf">raise</span> <span class="pp">Exception</span>(msg)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="91280caf" class="cell" data-execution_count="4">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>pdirs <span class="op">=</span> glob.glob(indir <span class="op">+</span> <span class="st">"/GSE*"</span>)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>alldata <span class="op">=</span> {}</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>counter <span class="op">=</span> <span class="dv">0</span> </span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>per_dataset <span class="op">=</span> <span class="dv">5</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> pdir <span class="kw">in</span> pdirs:</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(pdir)</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>    counter <span class="op">=</span> <span class="dv">1</span></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> f <span class="kw">in</span> os.listdir(pdir):</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>        read_cb_flag <span class="op">=</span> <span class="va">True</span> <span class="co"># with .mtx infiles the cellbender output is different cannot be read with read_10x-h5</span></span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> <span class="st">"raw"</span> <span class="kw">in</span> f:</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>            raw <span class="op">=</span> sc.read_10x_h5(os.path.join(pdir,f))</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>            <span class="co"># convert to correct name for cellbender runs</span></span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> <span class="st">"GSE213825"</span> <span class="kw">in</span> pdir <span class="kw">or</span> <span class="st">"GSE218853"</span> <span class="kw">in</span> pdir:</span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a>                fname <span class="op">=</span> f.split(<span class="st">"_"</span>)[<span class="dv">1</span>]</span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a>            <span class="cf">elif</span> <span class="st">"GSE229059"</span> <span class="kw">in</span> pdir:</span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a>                fname <span class="op">=</span> f.split(<span class="st">"_"</span>)[<span class="dv">0</span>]</span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a>            <span class="cf">elif</span> <span class="st">"GSE261852"</span> <span class="kw">in</span> pdir:</span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a>                fname <span class="op">=</span> f.split(<span class="st">"_"</span>)[<span class="dv">2</span>]</span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a>            read_cb_flag <span class="op">=</span> <span class="va">True</span>    </span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a>        <span class="cf">elif</span> <span class="st">"mtx.gz"</span> <span class="kw">in</span> f:</span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true" tabindex="-1"></a>            <span class="co"># skip for now, for some reason reads in the cellbender file with no genes.</span></span>
<span id="cb4-24"><a href="#cb4-24" aria-hidden="true" tabindex="-1"></a>            fname <span class="op">=</span> f.replace(<span class="st">"_matrix.mtx.gz"</span>,<span class="st">""</span>)</span>
<span id="cb4-25"><a href="#cb4-25" aria-hidden="true" tabindex="-1"></a>            raw <span class="op">=</span> sc.read_10x_mtx(pdir,prefix <span class="op">=</span> fname<span class="op">+</span><span class="st">"_"</span>)</span>
<span id="cb4-26"><a href="#cb4-26" aria-hidden="true" tabindex="-1"></a>            read_cb_flag <span class="op">=</span> <span class="va">False</span> </span>
<span id="cb4-27"><a href="#cb4-27" aria-hidden="true" tabindex="-1"></a>        <span class="cf">else</span>:</span>
<span id="cb4-28"><a href="#cb4-28" aria-hidden="true" tabindex="-1"></a>            <span class="cf">continue</span></span>
<span id="cb4-29"><a href="#cb4-29" aria-hidden="true" tabindex="-1"></a>        raw.var_names_make_unique()</span>
<span id="cb4-30"><a href="#cb4-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-31"><a href="#cb4-31" aria-hidden="true" tabindex="-1"></a>        <span class="co"># remove zero barcodes</span></span>
<span id="cb4-32"><a href="#cb4-32" aria-hidden="true" tabindex="-1"></a>        raw <span class="op">=</span> raw[raw.X.<span class="bu">sum</span>(axis<span class="op">=</span><span class="dv">1</span>)<span class="op">&gt;</span><span class="dv">0</span>,:]</span>
<span id="cb4-33"><a href="#cb4-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-34"><a href="#cb4-34" aria-hidden="true" tabindex="-1"></a>        <span class="co"># read in the cellbender file.</span></span>
<span id="cb4-35"><a href="#cb4-35" aria-hidden="true" tabindex="-1"></a>        cb_file <span class="op">=</span> os.path.join(cbdir,fname,<span class="st">"cellbender"</span>,<span class="st">"cellbender_out_filtered.h5"</span>)</span>
<span id="cb4-36"><a href="#cb4-36" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> os.path.exists(cb_file):</span>
<span id="cb4-37"><a href="#cb4-37" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> read_cb_flag:</span>
<span id="cb4-38"><a href="#cb4-38" aria-hidden="true" tabindex="-1"></a>                cb <span class="op">=</span> sc.read_10x_h5(cb_file)</span>
<span id="cb4-39"><a href="#cb4-39" aria-hidden="true" tabindex="-1"></a>            <span class="cf">else</span>:   </span>
<span id="cb4-40"><a href="#cb4-40" aria-hidden="true" tabindex="-1"></a>                cb <span class="op">=</span> _read_v3_10x_h5(cb_file)</span>
<span id="cb4-41"><a href="#cb4-41" aria-hidden="true" tabindex="-1"></a>            cb.var_names_make_unique()</span>
<span id="cb4-42"><a href="#cb4-42" aria-hidden="true" tabindex="-1"></a>        <span class="cf">else</span>:   </span>
<span id="cb4-43"><a href="#cb4-43" aria-hidden="true" tabindex="-1"></a>            <span class="bu">print</span>(<span class="st">"No cellbender file:"</span> <span class="op">+</span> cb_file)</span>
<span id="cb4-44"><a href="#cb4-44" aria-hidden="true" tabindex="-1"></a>            <span class="cf">continue</span></span>
<span id="cb4-45"><a href="#cb4-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-46"><a href="#cb4-46" aria-hidden="true" tabindex="-1"></a>        <span class="co"># add in biotype info</span></span>
<span id="cb4-47"><a href="#cb4-47" aria-hidden="true" tabindex="-1"></a>        t <span class="op">=</span> cb.var.merge(annot, left_on<span class="op">=</span><span class="st">"gene_ids"</span>, right_on<span class="op">=</span><span class="st">"ensembl_gene_id"</span>, how<span class="op">=</span><span class="st">'left'</span>)</span>
<span id="cb4-48"><a href="#cb4-48" aria-hidden="true" tabindex="-1"></a>        t.index <span class="op">=</span> cb.var.index</span>
<span id="cb4-49"><a href="#cb4-49" aria-hidden="true" tabindex="-1"></a>        <span class="co"># add in "unnass" for all nas in gene_class</span></span>
<span id="cb4-50"><a href="#cb4-50" aria-hidden="true" tabindex="-1"></a>        t.gene_class[t.gene_class.isna()] <span class="op">=</span> <span class="st">"other"</span></span>
<span id="cb4-51"><a href="#cb4-51" aria-hidden="true" tabindex="-1"></a>        cb.var <span class="op">=</span> t</span>
<span id="cb4-52"><a href="#cb4-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-53"><a href="#cb4-53" aria-hidden="true" tabindex="-1"></a>        <span class="co"># define ribo hb genes. not in biotypes</span></span>
<span id="cb4-54"><a href="#cb4-54" aria-hidden="true" tabindex="-1"></a>        cb.var[<span class="st">'ribo'</span>] <span class="op">=</span> cb.var_names.<span class="bu">str</span>.startswith((<span class="st">"Rps"</span>,<span class="st">"Rpl"</span>))</span>
<span id="cb4-55"><a href="#cb4-55" aria-hidden="true" tabindex="-1"></a>        cb.var[<span class="st">'hb'</span>] <span class="op">=</span> cb.var_names.<span class="bu">str</span>.contains((<span class="st">"^Hb[^(p|e|s)]"</span>))</span>
<span id="cb4-56"><a href="#cb4-56" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> bt <span class="kw">in</span> gene_classes:</span>
<span id="cb4-57"><a href="#cb4-57" aria-hidden="true" tabindex="-1"></a>            cb.var[bt] <span class="op">=</span> cb.var[<span class="st">"gene_class"</span>] <span class="op">==</span> bt</span>
<span id="cb4-58"><a href="#cb4-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-59"><a href="#cb4-59" aria-hidden="true" tabindex="-1"></a>        qc_vars <span class="op">=</span>  gene_classes <span class="op">+</span> [<span class="st">"ribo"</span>,<span class="st">"hb"</span>]           </span>
<span id="cb4-60"><a href="#cb4-60" aria-hidden="true" tabindex="-1"></a>        raw.var <span class="op">=</span> cb.var</span>
<span id="cb4-61"><a href="#cb4-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-62"><a href="#cb4-62" aria-hidden="true" tabindex="-1"></a>        <span class="co"># calculate stats for barcode rank plots from the full raw matrix.</span></span>
<span id="cb4-63"><a href="#cb4-63" aria-hidden="true" tabindex="-1"></a>        sc.pp.calculate_qc_metrics(raw, inplace<span class="op">=</span><span class="va">True</span>, log1p<span class="op">=</span><span class="va">False</span>, qc_vars<span class="op">=</span>qc_vars)</span>
<span id="cb4-64"><a href="#cb4-64" aria-hidden="true" tabindex="-1"></a>        <span class="co"># add the stats to the cb object</span></span>
<span id="cb4-65"><a href="#cb4-65" aria-hidden="true" tabindex="-1"></a>        cb.uns[<span class="st">"raw_obs"</span>] <span class="op">=</span> raw.obs.copy()</span>
<span id="cb4-66"><a href="#cb4-66" aria-hidden="true" tabindex="-1"></a>        cb.uns[<span class="st">"raw_var"</span>] <span class="op">=</span> raw.var.copy()</span>
<span id="cb4-67"><a href="#cb4-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-68"><a href="#cb4-68" aria-hidden="true" tabindex="-1"></a>        <span class="co">#filter for same cells in raw and recalculate the stats. </span></span>
<span id="cb4-69"><a href="#cb4-69" aria-hidden="true" tabindex="-1"></a>        <span class="co"># add in as a layer</span></span>
<span id="cb4-70"><a href="#cb4-70" aria-hidden="true" tabindex="-1"></a>        raw <span class="op">=</span> raw[cb.obs_names,:]</span>
<span id="cb4-71"><a href="#cb4-71" aria-hidden="true" tabindex="-1"></a>        cb.layers[<span class="st">"raw"</span>]<span class="op">=</span>raw.X.copy()</span>
<span id="cb4-72"><a href="#cb4-72" aria-hidden="true" tabindex="-1"></a>        cb.layers[<span class="st">"CB"</span>]<span class="op">=</span>cb.X.copy()</span>
<span id="cb4-73"><a href="#cb4-73" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-74"><a href="#cb4-74" aria-hidden="true" tabindex="-1"></a>        <span class="co"># read in the other methods. For soupX, 2 resolutions, 0.2 and 0.8</span></span>
<span id="cb4-75"><a href="#cb4-75" aria-hidden="true" tabindex="-1"></a>        <span class="co"># OBS! DX and SX does not have integers, round the values!!</span></span>
<span id="cb4-76"><a href="#cb4-76" aria-hidden="true" tabindex="-1"></a>        sx_file1 <span class="op">=</span> os.path.join(cbdir,fname,<span class="st">"soupX"</span>,<span class="st">"soupX_cormat_0.2.h5"</span>)</span>
<span id="cb4-77"><a href="#cb4-77" aria-hidden="true" tabindex="-1"></a>        tmp <span class="op">=</span> sc.read_10x_h5(sx_file1)</span>
<span id="cb4-78"><a href="#cb4-78" aria-hidden="true" tabindex="-1"></a>        cb.layers[<span class="st">"SX02"</span>] <span class="op">=</span> np.<span class="bu">round</span>(tmp.X.copy()) </span>
<span id="cb4-79"><a href="#cb4-79" aria-hidden="true" tabindex="-1"></a>        sx_file2 <span class="op">=</span> os.path.join(cbdir,fname,<span class="st">"soupX"</span>,<span class="st">"soupX_cormat_0.8.h5"</span>)</span>
<span id="cb4-80"><a href="#cb4-80" aria-hidden="true" tabindex="-1"></a>        tmp <span class="op">=</span> sc.read_10x_h5(sx_file1)</span>
<span id="cb4-81"><a href="#cb4-81" aria-hidden="true" tabindex="-1"></a>        cb.layers[<span class="st">"SX08"</span>] <span class="op">=</span> np.<span class="bu">round</span>(tmp.X.copy()) </span>
<span id="cb4-82"><a href="#cb4-82" aria-hidden="true" tabindex="-1"></a>        dx_file <span class="op">=</span> os.path.join(cbdir,fname,<span class="st">"decontX"</span>,<span class="st">"decontX_cormat.h5"</span>)</span>
<span id="cb4-83"><a href="#cb4-83" aria-hidden="true" tabindex="-1"></a>        tmp <span class="op">=</span> sc.read_10x_h5(dx_file)</span>
<span id="cb4-84"><a href="#cb4-84" aria-hidden="true" tabindex="-1"></a>        cb.layers[<span class="st">"DX"</span>] <span class="op">=</span> np.<span class="bu">round</span>(tmp.X.copy())        </span>
<span id="cb4-85"><a href="#cb4-85" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-86"><a href="#cb4-86" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> l <span class="kw">in</span> cb.layers.keys():</span>
<span id="cb4-87"><a href="#cb4-87" aria-hidden="true" tabindex="-1"></a>            cb.X <span class="op">=</span> cb.layers[l].copy()</span>
<span id="cb4-88"><a href="#cb4-88" aria-hidden="true" tabindex="-1"></a>            qc <span class="op">=</span> sc.pp.calculate_qc_metrics(cb, inplace<span class="op">=</span><span class="va">False</span>, log1p<span class="op">=</span><span class="va">False</span>, qc_vars<span class="op">=</span>qc_vars) <span class="co"># first is obs, second is var.</span></span>
<span id="cb4-89"><a href="#cb4-89" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> l <span class="op">!=</span> <span class="st">"raw"</span>:</span>
<span id="cb4-90"><a href="#cb4-90" aria-hidden="true" tabindex="-1"></a>                <span class="co"># proportions relative to raw data.</span></span>
<span id="cb4-91"><a href="#cb4-91" aria-hidden="true" tabindex="-1"></a>                qc[<span class="dv">0</span>][<span class="st">"nFilt"</span>] <span class="op">=</span> cb.obs[<span class="st">"raw_total_counts"</span>] <span class="op">-</span> qc[<span class="dv">0</span>][<span class="st">"total_counts"</span>]</span>
<span id="cb4-92"><a href="#cb4-92" aria-hidden="true" tabindex="-1"></a>                qc[<span class="dv">0</span>][<span class="st">"propFilt"</span>] <span class="op">=</span> qc[<span class="dv">0</span>][<span class="st">"nFilt"</span>] <span class="op">/</span> cb.obs[<span class="st">"raw_total_counts"</span>]</span>
<span id="cb4-93"><a href="#cb4-93" aria-hidden="true" tabindex="-1"></a>                qc[<span class="dv">1</span>][<span class="st">"nFilt"</span>] <span class="op">=</span> cb.var[<span class="st">"raw_total_counts"</span>] <span class="op">-</span> qc[<span class="dv">1</span>][<span class="st">"total_counts"</span>]</span>
<span id="cb4-94"><a href="#cb4-94" aria-hidden="true" tabindex="-1"></a>                qc[<span class="dv">1</span>][<span class="st">"propFilt"</span>] <span class="op">=</span> qc[<span class="dv">1</span>][<span class="st">"nFilt"</span>] <span class="op">/</span> cb.var[<span class="st">"raw_total_counts"</span>]</span>
<span id="cb4-95"><a href="#cb4-95" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-96"><a href="#cb4-96" aria-hidden="true" tabindex="-1"></a>            qc[<span class="dv">0</span>].columns <span class="op">=</span> l <span class="op">+</span> <span class="st">"_"</span> <span class="op">+</span> qc[<span class="dv">0</span>].columns</span>
<span id="cb4-97"><a href="#cb4-97" aria-hidden="true" tabindex="-1"></a>            cb.obs <span class="op">=</span> pd.concat([cb.obs,qc[<span class="dv">0</span>]], axis<span class="op">=</span><span class="dv">1</span>, join<span class="op">=</span><span class="st">'inner'</span>)</span>
<span id="cb4-98"><a href="#cb4-98" aria-hidden="true" tabindex="-1"></a>            qc[<span class="dv">1</span>].columns <span class="op">=</span> l <span class="op">+</span> <span class="st">"_"</span> <span class="op">+</span> qc[<span class="dv">1</span>].columns</span>
<span id="cb4-99"><a href="#cb4-99" aria-hidden="true" tabindex="-1"></a>            cb.var <span class="op">=</span> pd.concat([cb.var,qc[<span class="dv">1</span>]], axis<span class="op">=</span><span class="dv">1</span>, join<span class="op">=</span><span class="st">'inner'</span>)</span>
<span id="cb4-100"><a href="#cb4-100" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-101"><a href="#cb4-101" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-102"><a href="#cb4-102" aria-hidden="true" tabindex="-1"></a>        alldata[fname]<span class="op">=</span>cb</span>
<span id="cb4-103"><a href="#cb4-103" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="st">"Done "</span> <span class="op">+</span> fname)</span>
<span id="cb4-104"><a href="#cb4-104" aria-hidden="true" tabindex="-1"></a>        counter <span class="op">+=</span> <span class="dv">1</span></span>
<span id="cb4-105"><a href="#cb4-105" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> counter <span class="op">&gt;</span> per_dataset:</span>
<span id="cb4-106"><a href="#cb4-106" aria-hidden="true" tabindex="-1"></a>            <span class="cf">break</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>/Users/asabjor/projects/sc-devop/scQC/data/mouse/GSE229059
Done GSM7150735
Done GSM7150737
Done GSM7150718
Done GSM7150730
Done GSM7150720
/Users/asabjor/projects/sc-devop/scQC/data/mouse/GSE244142
Done GSM7807805_SITTE9
Done GSM7807804_SITTD9
Done GSM7807806_SITTF9
Done GSM7807807_SITTH4
Done GSM7807801_SITTA8
/Users/asabjor/projects/sc-devop/scQC/data/mouse/GSE218853
Done rep1
Done rep2
Done nuc3
Done nuc2
Done rep3
/Users/asabjor/projects/sc-devop/scQC/data/mouse/GSE213825
Done SAM24374037
Done SAM24352445
Done SAM24353297
Done SAM24352443
Done SAM24374041
/Users/asabjor/projects/sc-devop/scQC/data/mouse/GSE261852
Done spl2
Done spl1</code></pre>
</div>
</div>
</section>
<section id="cell-filtering-cellbender" class="level3">
<h3 class="anchored" data-anchor-id="cell-filtering-cellbender">Cell filtering cellbender</h3>
<p>Plot barcode rank plot colored by cellbender/cellranger filtering. Calculate rank plot based on cellranger total counts.</p>
<div id="f5870be2" class="cell" data-execution_count="5">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> plot_general_stats(tmp,name):</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>    display(Markdown(<span class="ss">f"## </span><span class="sc">{</span>name<span class="sc">}</span><span class="ss">"</span>))</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>    tmp.uns[<span class="st">"raw_obs"</span>][<span class="st">"inCB"</span>] <span class="op">=</span> tmp.uns[<span class="st">"raw_obs"</span>].index.isin(tmp.obs_names)</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>    tmp.uns[<span class="st">"raw_obs"</span>][<span class="st">"rankRaw"</span>] <span class="op">=</span> tmp.uns[<span class="st">"raw_obs"</span>][<span class="st">'total_counts'</span>].rank(method<span class="op">=</span><span class="st">'first'</span>, ascending <span class="op">=</span> <span class="va">False</span>)</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>    fig, axs <span class="op">=</span> plt.subplots(<span class="dv">2</span>, <span class="dv">2</span>, figsize<span class="op">=</span>(<span class="dv">10</span>,<span class="dv">8</span>),constrained_layout<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">0</span>,<span class="dv">0</span>].scatter(tmp.uns[<span class="st">"raw_obs"</span>][<span class="st">'rankRaw'</span>], tmp.uns[<span class="st">"raw_obs"</span>][<span class="st">'total_counts'</span>], c<span class="op">=</span>tmp.uns[<span class="st">"raw_obs"</span>][<span class="st">"inCB"</span>])</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">0</span>,<span class="dv">0</span>].set_xscale(<span class="st">"log"</span>)</span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">0</span>,<span class="dv">0</span>].set_yscale(<span class="st">"log"</span>)</span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">0</span>,<span class="dv">0</span>].set_xlabel(<span class="st">"rankRaw"</span>)</span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">0</span>,<span class="dv">0</span>].set_ylabel(<span class="st">"nUMI"</span>)</span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">0</span>,<span class="dv">0</span>].title.set_text(<span class="st">"CB cells selection"</span>)</span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a>    sns.histplot(data<span class="op">=</span>tmp.uns[<span class="st">"raw_obs"</span>], x<span class="op">=</span><span class="st">'rankRaw'</span>, hue<span class="op">=</span><span class="st">'inCB'</span>,  kde<span class="op">=</span><span class="va">False</span>, stat<span class="op">=</span><span class="st">'count'</span>, edgecolor<span class="op">=</span><span class="va">None</span>, multiple<span class="op">=</span><span class="st">"stack"</span>, ax<span class="op">=</span>axs[<span class="dv">1</span>,<span class="dv">0</span>])</span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">1</span>,<span class="dv">0</span>].set_xscale(<span class="st">"log"</span>)</span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">1</span>,<span class="dv">0</span>].title.set_text(<span class="st">"CB cells selection, proportion"</span>)</span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true" tabindex="-1"></a>    <span class="co"># counts before after filt</span></span>
<span id="cb6-21"><a href="#cb6-21" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">0</span>,<span class="dv">1</span>].scatter(tmp.obs[<span class="st">'raw_total_counts'</span>], tmp.obs[<span class="st">'CB_total_counts'</span>], alpha<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="cb6-22"><a href="#cb6-22" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">0</span>,<span class="dv">1</span>].axline((<span class="dv">0</span>, <span class="dv">0</span>), slope<span class="op">=</span><span class="dv">1</span>, color<span class="op">=</span><span class="st">"black"</span>)</span>
<span id="cb6-23"><a href="#cb6-23" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">0</span>,<span class="dv">1</span>].set_xlabel(<span class="st">"Raw counts"</span>)</span>
<span id="cb6-24"><a href="#cb6-24" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">0</span>,<span class="dv">1</span>].set_ylabel(<span class="st">"CB counts"</span>)</span>
<span id="cb6-25"><a href="#cb6-25" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">0</span>,<span class="dv">1</span>].set_xscale(<span class="st">"log"</span>)</span>
<span id="cb6-26"><a href="#cb6-26" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">0</span>,<span class="dv">1</span>].set_yscale(<span class="st">"log"</span>)</span>
<span id="cb6-27"><a href="#cb6-27" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">0</span>,<span class="dv">1</span>].title.set_text(<span class="st">"Counts per cell before/after CB"</span>)</span>
<span id="cb6-28"><a href="#cb6-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-29"><a href="#cb6-29" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">1</span>,<span class="dv">1</span>].scatter(tmp.var[<span class="st">"raw_total_counts"</span>], tmp.var[<span class="st">'CB_total_counts'</span>], alpha<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="cb6-30"><a href="#cb6-30" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">1</span>,<span class="dv">1</span>].axline((<span class="dv">0</span>, <span class="dv">0</span>), slope<span class="op">=</span><span class="dv">1</span>, color<span class="op">=</span><span class="st">"black"</span>)</span>
<span id="cb6-31"><a href="#cb6-31" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">1</span>,<span class="dv">1</span>].set_xlabel(<span class="st">"Raw gene counts"</span>)</span>
<span id="cb6-32"><a href="#cb6-32" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">1</span>,<span class="dv">1</span>].set_ylabel(<span class="st">"CB gene counts"</span>)</span>
<span id="cb6-33"><a href="#cb6-33" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">1</span>,<span class="dv">1</span>].set_xscale(<span class="st">"log"</span>)</span>
<span id="cb6-34"><a href="#cb6-34" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">1</span>,<span class="dv">1</span>].set_yscale(<span class="st">"log"</span>)</span>
<span id="cb6-35"><a href="#cb6-35" aria-hidden="true" tabindex="-1"></a>    axs[<span class="dv">1</span>,<span class="dv">1</span>].title.set_text(<span class="st">"Counts per gene before/after CB"</span>)</span>
<span id="cb6-36"><a href="#cb6-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-37"><a href="#cb6-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-38"><a href="#cb6-38" aria-hidden="true" tabindex="-1"></a>    fig.suptitle(name)</span>
<span id="cb6-39"><a href="#cb6-39" aria-hidden="true" tabindex="-1"></a>    display(plt.gcf())</span>
<span id="cb6-40"><a href="#cb6-40" aria-hidden="true" tabindex="-1"></a>    plt.close()</span>
<span id="cb6-41"><a href="#cb6-41" aria-hidden="true" tabindex="-1"></a>    display(Markdown(<span class="ss">f" "</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="tabset-margin-container"></div><div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-1-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-1" role="tab" aria-controls="tabset-1-1" aria-selected="true" href="">GSM7150735</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-2" role="tab" aria-controls="tabset-1-2" aria-selected="false" href="">GSM7150737</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-3" role="tab" aria-controls="tabset-1-3" aria-selected="false" href="">GSM7150718</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-4-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-4" role="tab" aria-controls="tabset-1-4" aria-selected="false" href="">GSM7150730</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-5-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-5" role="tab" aria-controls="tabset-1-5" aria-selected="false" href="">GSM7150720</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-6-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-6" role="tab" aria-controls="tabset-1-6" aria-selected="false" href="">GSM7807805_SITTE9</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-7-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-7" role="tab" aria-controls="tabset-1-7" aria-selected="false" href="">GSM7807804_SITTD9</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-8-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-8" role="tab" aria-controls="tabset-1-8" aria-selected="false" href="">GSM7807806_SITTF9</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-9-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-9" role="tab" aria-controls="tabset-1-9" aria-selected="false" href="">GSM7807807_SITTH4</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-10-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-10" role="tab" aria-controls="tabset-1-10" aria-selected="false" href="">GSM7807801_SITTA8</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-11-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-11" role="tab" aria-controls="tabset-1-11" aria-selected="false" href="">rep1</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-12-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-12" role="tab" aria-controls="tabset-1-12" aria-selected="false" href="">rep2</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-13-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-13" role="tab" aria-controls="tabset-1-13" aria-selected="false" href="">nuc3</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-14-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-14" role="tab" aria-controls="tabset-1-14" aria-selected="false" href="">nuc2</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-15-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-15" role="tab" aria-controls="tabset-1-15" aria-selected="false" href="">rep3</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-16-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-16" role="tab" aria-controls="tabset-1-16" aria-selected="false" href="">SAM24374037</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-17-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-17" role="tab" aria-controls="tabset-1-17" aria-selected="false" href="">SAM24352445</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-18-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-18" role="tab" aria-controls="tabset-1-18" aria-selected="false" href="">SAM24353297</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-19-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-19" role="tab" aria-controls="tabset-1-19" aria-selected="false" href="">SAM24352443</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-20-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-20" role="tab" aria-controls="tabset-1-20" aria-selected="false" href="">SAM24374041</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-21-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-21" role="tab" aria-controls="tabset-1-21" aria-selected="false" href="">spl2</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-22-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-22" role="tab" aria-controls="tabset-1-22" aria-selected="false" href="">spl1</a></li></ul>
<div class="tab-content">
<div id="tabset-1-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-1-1-tab">
<p><a href="ambient_methods_overview_files/figure-html/cell-7-output-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-1"><img src="ambient_methods_overview_files/figure-html/cell-7-output-2.png" width="809" height="649"></a></p>
</div>
<div id="tabset-1-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-2-tab">
<p><a href="ambient_methods_overview_files/figure-html/cell-7-output-5.png" class="lightbox" data-gallery="quarto-lightbox-gallery-2"><img src="ambient_methods_overview_files/figure-html/cell-7-output-5.png" width="809" height="649"></a></p>
</div>
<div id="tabset-1-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-3-tab">
<p><a href="ambient_methods_overview_files/figure-html/cell-7-output-8.png" class="lightbox" data-gallery="quarto-lightbox-gallery-3"><img src="ambient_methods_overview_files/figure-html/cell-7-output-8.png" width="809" height="649"></a></p>
</div>
<div id="tabset-1-4" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-4-tab">
<p><a href="ambient_methods_overview_files/figure-html/cell-7-output-11.png" class="lightbox" data-gallery="quarto-lightbox-gallery-4"><img src="ambient_methods_overview_files/figure-html/cell-7-output-11.png" width="809" height="649"></a></p>
</div>
<div id="tabset-1-5" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-5-tab">
<p><a href="ambient_methods_overview_files/figure-html/cell-7-output-14.png" class="lightbox" data-gallery="quarto-lightbox-gallery-5"><img src="ambient_methods_overview_files/figure-html/cell-7-output-14.png" width="809" height="649"></a></p>
</div>
<div id="tabset-1-6" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-6-tab">
<p><a href="ambient_methods_overview_files/figure-html/cell-7-output-17.png" class="lightbox" data-gallery="quarto-lightbox-gallery-6"><img src="ambient_methods_overview_files/figure-html/cell-7-output-17.png" width="809" height="649"></a></p>
</div>
<div id="tabset-1-7" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-7-tab">
<p><a href="ambient_methods_overview_files/figure-html/cell-7-output-20.png" class="lightbox" data-gallery="quarto-lightbox-gallery-7"><img src="ambient_methods_overview_files/figure-html/cell-7-output-20.png" width="809" height="649"></a></p>
</div>
<div id="tabset-1-8" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-8-tab">
<p><a href="ambient_methods_overview_files/figure-html/cell-7-output-23.png" class="lightbox" data-gallery="quarto-lightbox-gallery-8"><img src="ambient_methods_overview_files/figure-html/cell-7-output-23.png" width="809" height="649"></a></p>
</div>
<div id="tabset-1-9" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-9-tab">
<p><a href="ambient_methods_overview_files/figure-html/cell-7-output-26.png" class="lightbox" data-gallery="quarto-lightbox-gallery-9"><img src="ambient_methods_overview_files/figure-html/cell-7-output-26.png" width="809" height="649"></a></p>
</div>
<div id="tabset-1-10" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-10-tab">
<p><a href="ambient_methods_overview_files/figure-html/cell-7-output-29.png" class="lightbox" data-gallery="quarto-lightbox-gallery-10"><img src="ambient_methods_overview_files/figure-html/cell-7-output-29.png" width="809" height="649"></a></p>
</div>
<div id="tabset-1-11" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-11-tab">
<p><a href="ambient_methods_overview_files/figure-html/cell-7-output-32.png" class="lightbox" data-gallery="quarto-lightbox-gallery-11"><img src="ambient_methods_overview_files/figure-html/cell-7-output-32.png" width="809" height="649"></a></p>
</div>
<div id="tabset-1-12" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-12-tab">
<p><a href="ambient_methods_overview_files/figure-html/cell-7-output-35.png" class="lightbox" data-gallery="quarto-lightbox-gallery-12"><img src="ambient_methods_overview_files/figure-html/cell-7-output-35.png" width="809" height="649"></a></p>
</div>
<div id="tabset-1-13" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-13-tab">
<p><a href="ambient_methods_overview_files/figure-html/cell-7-output-38.png" class="lightbox" data-gallery="quarto-lightbox-gallery-13"><img src="ambient_methods_overview_files/figure-html/cell-7-output-38.png" width="809" height="649"></a></p>
</div>
<div id="tabset-1-14" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-14-tab">
<p><a href="ambient_methods_overview_files/figure-html/cell-7-output-41.png" class="lightbox" data-gallery="quarto-lightbox-gallery-14"><img src="ambient_methods_overview_files/figure-html/cell-7-output-41.png" width="809" height="649"></a></p>
</div>
<div id="tabset-1-15" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-15-tab">
<p><a href="ambient_methods_overview_files/figure-html/cell-7-output-44.png" class="lightbox" data-gallery="quarto-lightbox-gallery-15"><img src="ambient_methods_overview_files/figure-html/cell-7-output-44.png" width="809" height="649"></a></p>
</div>
<div id="tabset-1-16" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-16-tab">
<p><a href="ambient_methods_overview_files/figure-html/cell-7-output-47.png" class="lightbox" data-gallery="quarto-lightbox-gallery-16"><img src="ambient_methods_overview_files/figure-html/cell-7-output-47.png" width="809" height="649"></a></p>
</div>
<div id="tabset-1-17" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-17-tab">
<p><a href="ambient_methods_overview_files/figure-html/cell-7-output-50.png" class="lightbox" data-gallery="quarto-lightbox-gallery-17"><img src="ambient_methods_overview_files/figure-html/cell-7-output-50.png" width="809" height="649"></a></p>
</div>
<div id="tabset-1-18" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-18-tab">
<p><a href="ambient_methods_overview_files/figure-html/cell-7-output-53.png" class="lightbox" data-gallery="quarto-lightbox-gallery-18"><img src="ambient_methods_overview_files/figure-html/cell-7-output-53.png" width="809" height="649"></a></p>
</div>
<div id="tabset-1-19" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-19-tab">
<p><a href="ambient_methods_overview_files/figure-html/cell-7-output-56.png" class="lightbox" data-gallery="quarto-lightbox-gallery-19"><img src="ambient_methods_overview_files/figure-html/cell-7-output-56.png" width="809" height="649"></a></p>
</div>
<div id="tabset-1-20" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-20-tab">
<p><a href="ambient_methods_overview_files/figure-html/cell-7-output-59.png" class="lightbox" data-gallery="quarto-lightbox-gallery-20"><img src="ambient_methods_overview_files/figure-html/cell-7-output-59.png" width="809" height="649"></a></p>
</div>
<div id="tabset-1-21" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-21-tab">
<p><a href="ambient_methods_overview_files/figure-html/cell-7-output-62.png" class="lightbox" data-gallery="quarto-lightbox-gallery-21"><img src="ambient_methods_overview_files/figure-html/cell-7-output-62.png" width="809" height="649"></a></p>
</div>
<div id="tabset-1-22" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-22-tab">
<p><a href="ambient_methods_overview_files/figure-html/cell-7-output-65.png" class="lightbox" data-gallery="quarto-lightbox-gallery-22"><img src="ambient_methods_overview_files/figure-html/cell-7-output-65.png" width="809" height="649"></a></p>
</div>
</div>
</div>
<p>Weird histograms for some samples, especially the SAM samples.</p>
</section>
<section id="filtering-stats-all-methods" class="level3">
<h3 class="anchored" data-anchor-id="filtering-stats-all-methods">Filtering stats all methods</h3>
<p>Proportion filtered per cell</p>
<div id="a75ee098" class="cell" data-collapse="true" data-execution_count="7">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>all_methods <span class="op">=</span> [<span class="st">"CB"</span>,<span class="st">"SX02"</span>,<span class="st">"SX08"</span>,<span class="st">"DX"</span>]</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>pf <span class="op">=</span> [x <span class="op">+</span> <span class="st">"_propFilt"</span> <span class="cf">for</span> x <span class="kw">in</span> all_methods]</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>nPlot <span class="op">=</span> <span class="dv">9</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> r <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">0</span>,<span class="bu">len</span>(alldata),nPlot): </span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>    s <span class="op">=</span> <span class="bu">list</span>(alldata.keys())[r:(r<span class="op">+</span>nPlot)]</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>    fig, axs <span class="op">=</span> plt.subplots(<span class="dv">3</span>, <span class="dv">3</span>, figsize<span class="op">=</span>(<span class="dv">10</span>,<span class="dv">8</span>),constrained_layout<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> sname, ax <span class="kw">in</span> <span class="bu">zip</span>(s, axs.flatten()):</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>        sc.pl.violin(alldata[sname], pf, rotation<span class="op">=</span><span class="dv">90</span>, ax<span class="op">=</span>ax, show<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>        ax.tick_params(axis<span class="op">=</span><span class="st">'x'</span>, labelsize<span class="op">=</span><span class="dv">8</span>)</span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>        ax.title.set_text(sname)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_files/figure-html/cell-8-output-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-23"><img src="ambient_methods_overview_files/figure-html/cell-8-output-1.png" width="809" height="649" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_files/figure-html/cell-8-output-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-24"><img src="ambient_methods_overview_files/figure-html/cell-8-output-2.png" width="809" height="649" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_files/figure-html/cell-8-output-3.png" class="lightbox" data-gallery="quarto-lightbox-gallery-25"><img src="ambient_methods_overview_files/figure-html/cell-8-output-3.png" width="809" height="649" class="figure-img"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="gene-stats" class="level3">
<h3 class="anchored" data-anchor-id="gene-stats">Gene stats</h3>
<p>Plot fraction filtered per gene vs total counts before filtering. Color genes by:</p>
<ul>
<li>green = mt</li>
<li>red = pc</li>
<li>blue = lincRNA</li>
<li>purple = other (all other classes, plus ones not in latest biomart ensembl version)</li>
</ul>
<div id="62d6e08d" class="cell" data-execution_count="8">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>colors<span class="op">=</span>sns.color_palette(<span class="st">'Set1'</span>)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>gene_colors<span class="op">=</span>{<span class="st">'pc'</span>: colors[<span class="dv">0</span>], <span class="st">'lncRNA'</span>: colors[<span class="dv">1</span>], <span class="st">'mt'</span>:colors[<span class="dv">2</span>], <span class="st">'other'</span>:colors[<span class="dv">3</span>]} </span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>hexcols <span class="op">=</span> colors.as_hex()</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>gene_colors_hex <span class="op">=</span> {<span class="st">'pc'</span>: hexcols[<span class="dv">0</span>], <span class="st">'lncRNA'</span>: hexcols[<span class="dv">1</span>], <span class="st">'mt'</span>:hexcols[<span class="dv">2</span>], <span class="st">'other'</span>:hexcols[<span class="dv">3</span>]} </span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> plot_gene_proportions(df,title<span class="op">=</span><span class="st">" "</span>, nPlot<span class="op">=</span><span class="dv">20</span>):</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>    df <span class="op">=</span> df.loc[df.raw_total_counts <span class="op">&gt;</span> <span class="dv">0</span>,:]</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>    df[<span class="st">'Clog'</span>] <span class="op">=</span> np.log1p(df[<span class="st">'raw_total_counts'</span>])</span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>    fig, axs <span class="op">=</span> plt.subplots(<span class="dv">1</span>, <span class="dv">4</span>, figsize<span class="op">=</span>(<span class="dv">10</span>,<span class="dv">4</span>),constrained_layout<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i,n <span class="kw">in</span> <span class="bu">enumerate</span>(all_methods):</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>        <span class="co"># select sels to annotate</span></span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>        df[n<span class="op">+</span><span class="st">"_rank"</span>]<span class="op">=</span> df[n<span class="op">+</span><span class="st">'_propFilt'</span>] <span class="op">*</span> df[<span class="st">'Clog'</span>]</span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>        df <span class="op">=</span> df.sort_values(by<span class="op">=</span>n<span class="op">+</span><span class="st">"_rank"</span>, axis<span class="op">=</span><span class="dv">0</span>, ascending<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>        sel <span class="op">=</span> df.head(nPlot)</span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a>        sns.scatterplot(df, x<span class="op">=</span><span class="st">'Clog'</span>,y<span class="op">=</span>n<span class="op">+</span><span class="st">'_propFilt'</span>,hue<span class="op">=</span><span class="st">'gene_class'</span>, palette<span class="op">=</span>gene_colors,  legend<span class="op">=</span><span class="va">False</span>, ax<span class="op">=</span>axs[i])</span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a>        axs[i].title.set_text(n)</span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true" tabindex="-1"></a>        <span class="co"># annotate points in axis</span></span>
<span id="cb8-21"><a href="#cb8-21" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> idx, row <span class="kw">in</span> sel.iterrows():</span>
<span id="cb8-22"><a href="#cb8-22" aria-hidden="true" tabindex="-1"></a>            axs[i].annotate(idx, ( row[<span class="st">'Clog'</span>],row[n<span class="op">+</span><span class="st">'_propFilt'</span>]), rotation <span class="op">=</span> <span class="dv">45</span>, size<span class="op">=</span><span class="st">'x-small'</span>)</span>
<span id="cb8-23"><a href="#cb8-23" aria-hidden="true" tabindex="-1"></a>        fig.suptitle(title)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="b061a002" class="cell" data-collapse="true" data-execution_count="9">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> sname <span class="kw">in</span> alldata.keys():</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>    plot_gene_proportions(alldata[sname].var, title <span class="op">=</span> sname)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_files/figure-html/cell-10-output-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-26"><img src="ambient_methods_overview_files/figure-html/cell-10-output-1.png" width="809" height="329" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_files/figure-html/cell-10-output-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-27"><img src="ambient_methods_overview_files/figure-html/cell-10-output-2.png" width="809" height="329" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_files/figure-html/cell-10-output-3.png" class="lightbox" data-gallery="quarto-lightbox-gallery-28"><img src="ambient_methods_overview_files/figure-html/cell-10-output-3.png" width="809" height="329" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_files/figure-html/cell-10-output-4.png" class="lightbox" data-gallery="quarto-lightbox-gallery-29"><img src="ambient_methods_overview_files/figure-html/cell-10-output-4.png" width="809" height="329" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_files/figure-html/cell-10-output-5.png" class="lightbox" data-gallery="quarto-lightbox-gallery-30"><img src="ambient_methods_overview_files/figure-html/cell-10-output-5.png" width="809" height="329" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_files/figure-html/cell-10-output-6.png" class="lightbox" data-gallery="quarto-lightbox-gallery-31"><img src="ambient_methods_overview_files/figure-html/cell-10-output-6.png" width="809" height="329" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_files/figure-html/cell-10-output-7.png" class="lightbox" data-gallery="quarto-lightbox-gallery-32"><img src="ambient_methods_overview_files/figure-html/cell-10-output-7.png" width="809" height="329" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_files/figure-html/cell-10-output-8.png" class="lightbox" data-gallery="quarto-lightbox-gallery-33"><img src="ambient_methods_overview_files/figure-html/cell-10-output-8.png" width="809" height="329" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_files/figure-html/cell-10-output-9.png" class="lightbox" data-gallery="quarto-lightbox-gallery-34"><img src="ambient_methods_overview_files/figure-html/cell-10-output-9.png" width="809" height="329" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_files/figure-html/cell-10-output-10.png" class="lightbox" data-gallery="quarto-lightbox-gallery-35"><img src="ambient_methods_overview_files/figure-html/cell-10-output-10.png" width="809" height="329" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_files/figure-html/cell-10-output-11.png" class="lightbox" data-gallery="quarto-lightbox-gallery-36"><img src="ambient_methods_overview_files/figure-html/cell-10-output-11.png" width="809" height="329" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_files/figure-html/cell-10-output-12.png" class="lightbox" data-gallery="quarto-lightbox-gallery-37"><img src="ambient_methods_overview_files/figure-html/cell-10-output-12.png" width="809" height="329" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_files/figure-html/cell-10-output-13.png" class="lightbox" data-gallery="quarto-lightbox-gallery-38"><img src="ambient_methods_overview_files/figure-html/cell-10-output-13.png" width="809" height="329" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_files/figure-html/cell-10-output-14.png" class="lightbox" data-gallery="quarto-lightbox-gallery-39"><img src="ambient_methods_overview_files/figure-html/cell-10-output-14.png" width="809" height="329" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_files/figure-html/cell-10-output-15.png" class="lightbox" data-gallery="quarto-lightbox-gallery-40"><img src="ambient_methods_overview_files/figure-html/cell-10-output-15.png" width="809" height="329" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_files/figure-html/cell-10-output-16.png" class="lightbox" data-gallery="quarto-lightbox-gallery-41"><img src="ambient_methods_overview_files/figure-html/cell-10-output-16.png" width="809" height="329" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_files/figure-html/cell-10-output-17.png" class="lightbox" data-gallery="quarto-lightbox-gallery-42"><img src="ambient_methods_overview_files/figure-html/cell-10-output-17.png" width="809" height="329" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_files/figure-html/cell-10-output-18.png" class="lightbox" data-gallery="quarto-lightbox-gallery-43"><img src="ambient_methods_overview_files/figure-html/cell-10-output-18.png" width="809" height="329" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_files/figure-html/cell-10-output-19.png" class="lightbox" data-gallery="quarto-lightbox-gallery-44"><img src="ambient_methods_overview_files/figure-html/cell-10-output-19.png" width="809" height="329" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_files/figure-html/cell-10-output-20.png" class="lightbox" data-gallery="quarto-lightbox-gallery-45"><img src="ambient_methods_overview_files/figure-html/cell-10-output-20.png" width="809" height="329" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_files/figure-html/cell-10-output-21.png" class="lightbox" data-gallery="quarto-lightbox-gallery-46"><img src="ambient_methods_overview_files/figure-html/cell-10-output-21.png" width="809" height="329" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_files/figure-html/cell-10-output-22.png" class="lightbox" data-gallery="quarto-lightbox-gallery-47"><img src="ambient_methods_overview_files/figure-html/cell-10-output-22.png" width="809" height="329" class="figure-img"></a></p>
</figure>
</div>
</div>
</div>
</section>
<section id="mito-etc." class="level3">
<h3 class="anchored" data-anchor-id="mito-etc.">Mito, etc.</h3>
<p>Violin plots after filtering and raw.</p>
<div id="e814ac3e" class="cell" data-collapse="true" data-execution_count="10">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>qc_plot <span class="op">=</span> [<span class="st">"mt"</span>,<span class="st">"hb"</span>,<span class="st">"pc"</span>]</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>plot_feats <span class="op">=</span> [m <span class="op">+</span> <span class="st">"_pct_counts_"</span> <span class="op">+</span> q <span class="cf">for</span> q <span class="kw">in</span> qc_plot <span class="cf">for</span> m <span class="kw">in</span> [<span class="st">"raw"</span>] <span class="op">+</span> all_methods ]</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>nPlot <span class="op">=</span> <span class="dv">9</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> r <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">0</span>,<span class="bu">len</span>(alldata),nPlot): </span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>    s <span class="op">=</span> <span class="bu">list</span>(alldata.keys())[r:(r<span class="op">+</span>nPlot)]</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>    fig, axs <span class="op">=</span> plt.subplots(<span class="dv">3</span>, <span class="dv">3</span>, figsize<span class="op">=</span>(<span class="dv">10</span>,<span class="dv">8</span>),constrained_layout<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> sname, ax <span class="kw">in</span> <span class="bu">zip</span>(s, axs.flatten()):</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>        sc.pl.violin(alldata[sname], plot_feats, rotation<span class="op">=</span><span class="dv">90</span>, ax<span class="op">=</span>ax, show<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>        ax.tick_params(axis<span class="op">=</span><span class="st">'x'</span>, labelsize<span class="op">=</span><span class="dv">8</span>)</span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>        ax.title.set_text(sname)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_files/figure-html/cell-11-output-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-48"><img src="ambient_methods_overview_files/figure-html/cell-11-output-1.png" width="809" height="649" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_files/figure-html/cell-11-output-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-49"><img src="ambient_methods_overview_files/figure-html/cell-11-output-2.png" width="809" height="649" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_files/figure-html/cell-11-output-3.png" class="lightbox" data-gallery="quarto-lightbox-gallery-50"><img src="ambient_methods_overview_files/figure-html/cell-11-output-3.png" width="809" height="649" class="figure-img"></a></p>
</figure>
</div>
</div>
</div>
<p>Generally more high mito cells after CB filtering.</p>
</section>
<section id="biotype-proportions" class="level3">
<h3 class="anchored" data-anchor-id="biotype-proportions">Biotype proportions</h3>
<p>Calculate total proportions of all biotypes per layer.</p>
<div id="561c7738" class="cell" data-execution_count="11">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> plot_biotypes(df,title,ax):</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>    tmp <span class="op">=</span> df.loc[:,df.columns.<span class="bu">str</span>.contains(<span class="st">"total_counts|gene_class"</span>)]</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>    tmp <span class="op">=</span> tmp.groupby(<span class="st">"gene_class"</span>).<span class="bu">sum</span>()</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>    tmp.columns <span class="op">=</span> tmp.columns.<span class="bu">str</span>.replace(<span class="st">"_counts"</span>,<span class="st">""</span>)</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>    tmp.columns <span class="op">=</span> tmp.columns.<span class="bu">str</span>.replace(<span class="st">"_total"</span>,<span class="st">""</span>)</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>    tmp_perc <span class="op">=</span> tmp.iloc[:, :].<span class="bu">apply</span>(<span class="kw">lambda</span> x: x <span class="op">/</span> x.<span class="bu">sum</span>(), axis<span class="op">=</span><span class="dv">0</span>) <span class="op">*</span> <span class="dv">100</span></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>    tmp_perc.transpose().plot(kind<span class="op">=</span><span class="st">"bar"</span>,stacked<span class="op">=</span><span class="va">True</span>, color<span class="op">=</span>gene_colors, legend<span class="op">=</span><span class="va">False</span>, title<span class="op">=</span>title, ax<span class="op">=</span>ax )</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>    lim <span class="op">=</span> np.<span class="bu">round</span>(<span class="dv">100</span><span class="op">-</span>tmp_perc.loc[<span class="st">"pc"</span>,:].<span class="bu">min</span>()<span class="op">+</span><span class="dv">5</span>)</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>    ax.set_ylim([<span class="dv">0</span>,lim])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="d8ca54cd" class="cell" data-collapse="true" data-execution_count="12">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>qc_plot <span class="op">=</span> [<span class="st">"total_counts"</span>,<span class="st">"n_genes_by_counts"</span>]</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>plot_feats <span class="op">=</span> [m <span class="op">+</span> <span class="st">"_"</span> <span class="op">+</span> q <span class="cf">for</span> q <span class="kw">in</span> qc_plot <span class="cf">for</span> m <span class="kw">in</span> [<span class="st">"raw"</span>] <span class="op">+</span> all_methods ]</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>nPlot <span class="op">=</span> <span class="dv">9</span></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> r <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">0</span>,<span class="bu">len</span>(alldata),nPlot): </span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>    s <span class="op">=</span> <span class="bu">list</span>(alldata.keys())[r:(r<span class="op">+</span>nPlot)]</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>    fig, axs <span class="op">=</span> plt.subplots(<span class="dv">3</span>, <span class="dv">3</span>, figsize<span class="op">=</span>(<span class="dv">10</span>,<span class="dv">8</span>),constrained_layout<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> sname, ax <span class="kw">in</span> <span class="bu">zip</span>(s, axs.flatten()):</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>        plot_biotypes(alldata[sname].var,sname,ax)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_files/figure-html/cell-13-output-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-51"><img src="ambient_methods_overview_files/figure-html/cell-13-output-1.png" width="809" height="649" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_files/figure-html/cell-13-output-2.png" class="lightbox" data-gallery="quarto-lightbox-gallery-52"><img src="ambient_methods_overview_files/figure-html/cell-13-output-2.png" width="809" height="649" class="figure-img"></a></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><a href="ambient_methods_overview_files/figure-html/cell-13-output-3.png" class="lightbox" data-gallery="quarto-lightbox-gallery-53"><img src="ambient_methods_overview_files/figure-html/cell-13-output-3.png" width="809" height="649" class="figure-img"></a></p>
</figure>
</div>
</div>
</div>
<div id="049b82f1" class="cell" data-execution_count="13">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>qc_plot <span class="op">=</span> [<span class="st">"total_counts"</span>,<span class="st">"n_genes_by_counts"</span>]</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>plot_feats <span class="op">=</span> [m <span class="op">+</span> <span class="st">"_"</span> <span class="op">+</span> q <span class="cf">for</span> q <span class="kw">in</span> qc_plot <span class="cf">for</span> m <span class="kw">in</span> [<span class="st">"raw"</span>] <span class="op">+</span> all_methods ]</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>nPlot <span class="op">=</span> <span class="dv">9</span></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> r <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">0</span>,<span class="bu">len</span>(alldata),nPlot): </span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>    s <span class="op">=</span> <span class="bu">list</span>(alldata.keys())[r:(r<span class="op">+</span>nPlot)]</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>    fig, axs <span class="op">=</span> plt.subplots(<span class="dv">3</span>, <span class="dv">3</span>, figsize<span class="op">=</span>(<span class="dv">10</span>,<span class="dv">8</span>),constrained_layout<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> sname, ax <span class="kw">in</span> <span class="bu">zip</span>(s, axs.flatten()):</span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>        sc.pl.violin(alldata[sname], plot_feats, rotation<span class="op">=</span><span class="dv">90</span>, ax<span class="op">=</span>ax, show<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>        ax.tick_params(axis<span class="op">=</span><span class="st">'x'</span>, labelsize<span class="op">=</span><span class="dv">8</span>)</span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>        ax.title.set_text(sname)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="f219ef60" class="cell" data-execution_count="14">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>nPlot <span class="op">=</span> <span class="dv">9</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> r <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">0</span>,<span class="bu">len</span>(alldata),nPlot): </span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>    s <span class="op">=</span> <span class="bu">list</span>(alldata.keys())[r:(r<span class="op">+</span>nPlot)]</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>    fig, axs <span class="op">=</span> plt.subplots(<span class="dv">3</span>, <span class="dv">3</span>, figsize<span class="op">=</span>(<span class="dv">10</span>,<span class="dv">8</span>),constrained_layout<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> sname, ax <span class="kw">in</span> <span class="bu">zip</span>(s, axs.flatten()):</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>        sc.pl.scatter(alldata[sname], <span class="st">"raw_pct_counts_mt"</span>,<span class="st">"CB_pct_counts_mt"</span>, ax<span class="op">=</span>ax, show<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>        ax.tick_params(axis<span class="op">=</span><span class="st">'x'</span>, labelsize<span class="op">=</span><span class="dv">8</span>)</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>        ax.title.set_text(sname)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<script>var lightboxQuarto = GLightbox({"loop":false,"closeEffect":"zoom","selector":".lightbox","descPosition":"bottom","openEffect":"zoom"});
(function() {
  let previousOnload = window.onload;
  window.onload = () => {
    if (previousOnload) {
      previousOnload();
    }
    lightboxQuarto.on('slide_before_load', (data) => {
      const { slideIndex, slideNode, slideConfig, player, trigger } = data;
      const href = trigger.getAttribute('href');
      if (href !== null) {
        const imgEl = window.document.querySelector(`a[href="${href}"] img`);
        if (imgEl !== null) {
          const srcAttr = imgEl.getAttribute("src");
          if (srcAttr && srcAttr.startsWith("data:")) {
            slideConfig.href = srcAttr;
          }
        }
      } 
    });
  
    lightboxQuarto.on('slide_after_load', (data) => {
      const { slideIndex, slideNode, slideConfig, player, trigger } = data;
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(slideNode);
      }
    });
  
  };
  
})();
          </script>




</body></html>